<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>

<meta charset="utf-8">
<meta name="generator" content="quarto-1.4.547">

<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">


<title>Analysis v5: Detecting Covariant Sites</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
  vertical-align: middle;
}
/* CSS for syntax highlighting */
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
  }
pre.numberSource { margin-left: 3em;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
</style>


<script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.5.1/jquery.min.js" integrity="sha512-bLT0Qm9VnAYZDflyKcBaQ2gg0hSYNQrJ8RilYldYQ1FxQYoCLtUjuuRuZo+fjqhx/qtq/1itJ0C2ejDxltZVFg==" crossorigin="anonymous"></script><script src="variant-freq-analysis-v5_files/libs/clipboard/clipboard.min.js"></script>
<script src="variant-freq-analysis-v5_files/libs/quarto-html/quarto.js"></script>
<script src="variant-freq-analysis-v5_files/libs/quarto-html/popper.min.js"></script>
<script src="variant-freq-analysis-v5_files/libs/quarto-html/tippy.umd.min.js"></script>
<script src="variant-freq-analysis-v5_files/libs/quarto-html/anchor.min.js"></script>
<link href="variant-freq-analysis-v5_files/libs/quarto-html/tippy.css" rel="stylesheet">
<link href="variant-freq-analysis-v5_files/libs/quarto-html/quarto-syntax-highlighting.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="variant-freq-analysis-v5_files/libs/bootstrap/bootstrap.min.js"></script>
<link href="variant-freq-analysis-v5_files/libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="variant-freq-analysis-v5_files/libs/bootstrap/bootstrap.min.css" rel="stylesheet" id="quarto-bootstrap" data-mode="light">
<script src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.6/require.min.js" integrity="sha512-c3Nl8+7g4LMSTdrm621y7kf9v3SDPnhxLNhcjFJbKECVnmZHTdo+IRO05sNLTH/D3vA6u1X32ehoLC7WFVdheg==" crossorigin="anonymous"></script>

<script type="application/javascript">define('jquery', [],function() {return window.jQuery;})</script>


</head>

<body class="fullcontent">

<div id="quarto-content" class="page-columns page-rows-contents page-layout-article">

<main class="content" id="quarto-document-content">

<header id="title-block-header" class="quarto-title-block default">
<div class="quarto-title">
<h1 class="title">Analysis v5: Detecting Covariant Sites</h1>
</div>



<div class="quarto-title-meta">

    
  
    
  </div>
  


</header>


<section id="analysis-v5-detecting-covariant-sites" class="level2">
<h2 class="anchored" data-anchor-id="analysis-v5-detecting-covariant-sites">Analysis v5: Detecting Covariant Sites</h2>
<div id="cell-1" class="cell" data-execution_count="1">
<div class="sourceCode cell-code" id="cb1"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb1-1"><a href="#cb1-1" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> os</span>
<span id="cb1-2"><a href="#cb1-2" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> numpy <span class="im">as</span> np</span>
<span id="cb1-3"><a href="#cb1-3" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> pandas <span class="im">as</span> pd</span>
<span id="cb1-4"><a href="#cb1-4" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> seaborn <span class="im">as</span> sns</span>
<span id="cb1-5"><a href="#cb1-5" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> statsmodels.api <span class="im">as</span> sm</span>
<span id="cb1-6"><a href="#cb1-6" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> matplotlib.pyplot <span class="im">as</span> plt</span>
<span id="cb1-7"><a href="#cb1-7" aria-hidden="true" tabindex="-1"></a><span class="im">from</span> natsort <span class="im">import</span> index_natsorted</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div id="cell-2" class="cell" data-execution_count="2">
<div class="sourceCode cell-code" id="cb2"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb2-1"><a href="#cb2-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Set color values</span></span>
<span id="cb2-2"><a href="#cb2-2" aria-hidden="true" tabindex="-1"></a>dutchfield <span class="op">=</span> [<span class="st">"#e60049"</span>, <span class="st">"#0bb4ff"</span>, <span class="st">"#87bc45"</span>, <span class="st">"#ef9b20"</span>, <span class="st">"#b33dc6"</span>]</span>
<span id="cb2-3"><a href="#cb2-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb2-4"><a href="#cb2-4" aria-hidden="true" tabindex="-1"></a>springpastel <span class="op">=</span> [<span class="st">"#fd7f6f"</span>, <span class="st">"#7eb0d5"</span>, <span class="st">"#b2e061"</span>, <span class="st">"#bd7ebe"</span>, <span class="st">"#ffb55a"</span>] </span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<section id="calculating-the-variance-of-the-af-and-fps-values" class="level3">
<h3 class="anchored" data-anchor-id="calculating-the-variance-of-the-af-and-fps-values">Calculating the Variance of the AF and FPS values</h3>
<p>It would be quite interesting to see if for each variant site per motif, the variance of the AF and FPS values are correlated.</p>
<p>I.e. if a variant site has a high variance in AF values, does it also have a high variance in FPS values?</p>
<p>First, load the data file as dataframe and then truncate the columns to include on AF and FPS data. Next, convert the dataframe into long format.</p>
<div id="cell-5" class="cell" data-execution_count="5">
<div class="sourceCode cell-code" id="cb3"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb3-1"><a href="#cb3-1" aria-hidden="true" tabindex="-1"></a><span class="co"># import the data</span></span>
<span id="cb3-2"><a href="#cb3-2" aria-hidden="true" tabindex="-1"></a><span class="co"># filepath = '/home/users/ntu/suffiazi/scratch/outputs/tmp/input_data/E2F5_E2F5_HUMAN.H11MO.0.B_fpscore-af-varsites-combined-matrix-wide.tsv'</span></span>
<span id="cb3-3"><a href="#cb3-3" aria-hidden="true" tabindex="-1"></a>filepath <span class="op">=</span> <span class="st">'../output-data/combined_matrix/TFDP1_M08108_2.00_fpscore-af-varsites-combined-matrix-wide.tsv'</span></span>
<span id="cb3-4"><a href="#cb3-4" aria-hidden="true" tabindex="-1"></a>dt_afps <span class="op">=</span> pd.read_csv(filepath, sep<span class="op">=</span><span class="st">'</span><span class="ch">\t</span><span class="st">'</span>)</span>
<span id="cb3-5"><a href="#cb3-5" aria-hidden="true" tabindex="-1"></a><span class="co"># extract motif id from filename</span></span>
<span id="cb3-6"><a href="#cb3-6" aria-hidden="true" tabindex="-1"></a>motif_id <span class="op">=</span> os.path.basename(filepath).replace(<span class="st">'_fpscore-af-varsites-combined-matrix-wide.tsv'</span>, <span class="st">''</span>)</span>
<span id="cb3-7"><a href="#cb3-7" aria-hidden="true" tabindex="-1"></a>motif_id</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display" data-execution_count="5">
<pre><code>'TFDP1_M08108_2.00'</code></pre>
</div>
</div>
<div id="cell-6" class="cell" data-execution_count="6">
<div class="sourceCode cell-code" id="cb5"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb5-1"><a href="#cb5-1" aria-hidden="true" tabindex="-1"></a>dt_afps.head()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display" data-execution_count="6">
<div>


<table class="dataframe table table-sm table-striped small" data-quarto-postprocess="true" data-border="1">
<thead>
<tr class="header">
<th data-quarto-table-cell-role="th"></th>
<th data-quarto-table-cell-role="th">Chromosome</th>
<th data-quarto-table-cell-role="th">Start</th>
<th data-quarto-table-cell-role="th">End</th>
<th data-quarto-table-cell-role="th">98JKPD8_lumA_fps</th>
<th data-quarto-table-cell-role="th">ANAB5F7_basal_fps</th>
<th data-quarto-table-cell-role="th">S6R691V_her2_fps</th>
<th data-quarto-table-cell-role="th">PU24GB8_lumB_fps</th>
<th data-quarto-table-cell-role="th">2GAMBDQ_norm_fps</th>
<th data-quarto-table-cell-role="th">2GAMBDQ_norm_varsite_pos</th>
<th data-quarto-table-cell-role="th">2GAMBDQ_norm_ref_allele</th>
<th data-quarto-table-cell-role="th">...</th>
<th data-quarto-table-cell-role="th">ANAB5F7_basal_AF</th>
<th data-quarto-table-cell-role="th">PU24GB8_lumB_varsite_pos</th>
<th data-quarto-table-cell-role="th">PU24GB8_lumB_ref_allele</th>
<th data-quarto-table-cell-role="th">PU24GB8_lumB_alt_allele</th>
<th data-quarto-table-cell-role="th">PU24GB8_lumB_AF</th>
<th data-quarto-table-cell-role="th">S6R691V_her2_varsite_pos</th>
<th data-quarto-table-cell-role="th">S6R691V_her2_ref_allele</th>
<th data-quarto-table-cell-role="th">S6R691V_her2_alt_allele</th>
<th data-quarto-table-cell-role="th">S6R691V_her2_AF</th>
<th data-quarto-table-cell-role="th">region_id</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td data-quarto-table-cell-role="th">0</td>
<td>chr1</td>
<td>191257</td>
<td>191268</td>
<td>0.00000</td>
<td>0.00000</td>
<td>0.00000</td>
<td>0.00000</td>
<td>0.00000</td>
<td>NaN</td>
<td>NaN</td>
<td>...</td>
<td>0.000000</td>
<td>191258.0</td>
<td>G</td>
<td>C</td>
<td>0.026316</td>
<td>NaN</td>
<td>NaN</td>
<td>NaN</td>
<td>0.000000</td>
<td>chr1:191257-191268</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">1</td>
<td>chr1</td>
<td>818021</td>
<td>818032</td>
<td>0.02966</td>
<td>0.04399</td>
<td>0.02131</td>
<td>0.03470</td>
<td>0.03443</td>
<td>818025.0</td>
<td>C</td>
<td>...</td>
<td>0.812500</td>
<td>818025.0</td>
<td>C</td>
<td>A</td>
<td>0.833333</td>
<td>818025.0</td>
<td>C</td>
<td>A</td>
<td>0.900000</td>
<td>chr1:818021-818032</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">2</td>
<td>chr1</td>
<td>818780</td>
<td>818791</td>
<td>0.02572</td>
<td>0.02376</td>
<td>0.02271</td>
<td>0.02814</td>
<td>0.03391</td>
<td>NaN</td>
<td>NaN</td>
<td>...</td>
<td>0.033333</td>
<td>818783.0</td>
<td>T</td>
<td>C</td>
<td>0.055556</td>
<td>818783.0</td>
<td>T</td>
<td>C</td>
<td>0.062500</td>
<td>chr1:818780-818791</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">3</td>
<td>chr1</td>
<td>960570</td>
<td>960581</td>
<td>0.28046</td>
<td>0.33500</td>
<td>0.28495</td>
<td>0.27618</td>
<td>0.22986</td>
<td>NaN</td>
<td>NaN</td>
<td>...</td>
<td>0.000000</td>
<td>NaN</td>
<td>NaN</td>
<td>NaN</td>
<td>0.000000</td>
<td>NaN</td>
<td>NaN</td>
<td>NaN</td>
<td>0.000000</td>
<td>chr1:960570-960581</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">4</td>
<td>chr1</td>
<td>977015</td>
<td>977026</td>
<td>1.10047</td>
<td>0.50514</td>
<td>0.71035</td>
<td>1.21305</td>
<td>1.30155</td>
<td>977023.0</td>
<td>G</td>
<td>...</td>
<td>0.941176</td>
<td>977023.0</td>
<td>G</td>
<td>A</td>
<td>0.947368</td>
<td>977023.0</td>
<td>G</td>
<td>A</td>
<td>0.911765</td>
<td>chr1:977015-977026</td>
</tr>
</tbody>
</table>

<p>5 rows × 29 columns</p>
</div>
</div>
</div>
<p>Let’s filter the loaded table to include only the <code>_AF</code> and <code>_fps</code> columns, as well as the <code>region_id</code> column.</p>
<div id="cell-8" class="cell" data-execution_count="7">
<div class="sourceCode cell-code" id="cb6"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb6-1"><a href="#cb6-1" aria-hidden="true" tabindex="-1"></a>afps_df <span class="op">=</span> dt_afps.<span class="bu">filter</span>(regex<span class="op">=</span><span class="st">'_AF$|_fps$|_id$'</span>).copy()</span>
<span id="cb6-2"><a href="#cb6-2" aria-hidden="true" tabindex="-1"></a>afps_df.head(<span class="dv">5</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display" data-execution_count="7">
<div>


<table class="dataframe table table-sm table-striped small" data-quarto-postprocess="true" data-border="1">
<thead>
<tr class="header">
<th data-quarto-table-cell-role="th"></th>
<th data-quarto-table-cell-role="th">98JKPD8_lumA_fps</th>
<th data-quarto-table-cell-role="th">ANAB5F7_basal_fps</th>
<th data-quarto-table-cell-role="th">S6R691V_her2_fps</th>
<th data-quarto-table-cell-role="th">PU24GB8_lumB_fps</th>
<th data-quarto-table-cell-role="th">2GAMBDQ_norm_fps</th>
<th data-quarto-table-cell-role="th">2GAMBDQ_norm_AF</th>
<th data-quarto-table-cell-role="th">98JKPD8_lumA_AF</th>
<th data-quarto-table-cell-role="th">ANAB5F7_basal_AF</th>
<th data-quarto-table-cell-role="th">PU24GB8_lumB_AF</th>
<th data-quarto-table-cell-role="th">S6R691V_her2_AF</th>
<th data-quarto-table-cell-role="th">region_id</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td data-quarto-table-cell-role="th">0</td>
<td>0.00000</td>
<td>0.00000</td>
<td>0.00000</td>
<td>0.00000</td>
<td>0.00000</td>
<td>0.000</td>
<td>0.000000</td>
<td>0.000000</td>
<td>0.026316</td>
<td>0.000000</td>
<td>chr1:191257-191268</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">1</td>
<td>0.02966</td>
<td>0.04399</td>
<td>0.02131</td>
<td>0.03470</td>
<td>0.03443</td>
<td>0.875</td>
<td>0.900000</td>
<td>0.812500</td>
<td>0.833333</td>
<td>0.900000</td>
<td>chr1:818021-818032</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">2</td>
<td>0.02572</td>
<td>0.02376</td>
<td>0.02271</td>
<td>0.02814</td>
<td>0.03391</td>
<td>0.000</td>
<td>0.125000</td>
<td>0.033333</td>
<td>0.055556</td>
<td>0.062500</td>
<td>chr1:818780-818791</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">3</td>
<td>0.28046</td>
<td>0.33500</td>
<td>0.28495</td>
<td>0.27618</td>
<td>0.22986</td>
<td>0.000</td>
<td>0.041667</td>
<td>0.000000</td>
<td>0.000000</td>
<td>0.000000</td>
<td>chr1:960570-960581</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">4</td>
<td>1.10047</td>
<td>0.50514</td>
<td>0.71035</td>
<td>1.21305</td>
<td>1.30155</td>
<td>1.000</td>
<td>1.000000</td>
<td>0.941176</td>
<td>0.947368</td>
<td>0.911765</td>
<td>chr1:977015-977026</td>
</tr>
</tbody>
</table>

</div>
</div>
</div>
<div id="cell-9" class="cell" data-execution_count="8">
<div class="sourceCode cell-code" id="cb7"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb7-1"><a href="#cb7-1" aria-hidden="true" tabindex="-1"></a><span class="co"># convert to long format</span></span>
<span id="cb7-2"><a href="#cb7-2" aria-hidden="true" tabindex="-1"></a>afps_df_long <span class="op">=</span> afps_df.melt(id_vars<span class="op">=</span>[<span class="st">"region_id"</span>], var_name<span class="op">=</span><span class="st">"variable"</span>, value_name<span class="op">=</span><span class="st">"value"</span>)</span>
<span id="cb7-3"><a href="#cb7-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb7-4"><a href="#cb7-4" aria-hidden="true" tabindex="-1"></a><span class="co"># split the variable column into sample_id and type columns using reverse split string method, which returns a dataframe of columns based on the number of splits (n=x); this can directly be assigned to new columns in the original dataframe</span></span>
<span id="cb7-5"><a href="#cb7-5" aria-hidden="true" tabindex="-1"></a>afps_df_long[[<span class="st">'sample_id'</span>, <span class="st">'type'</span>]] <span class="op">=</span> afps_df_long[<span class="st">'variable'</span>].<span class="bu">str</span>.rsplit(<span class="st">'_'</span>, n<span class="op">=</span><span class="dv">1</span>, expand<span class="op">=</span><span class="va">True</span>)</span>
<span id="cb7-6"><a href="#cb7-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb7-7"><a href="#cb7-7" aria-hidden="true" tabindex="-1"></a><span class="co"># drop the redundant 'variable' column</span></span>
<span id="cb7-8"><a href="#cb7-8" aria-hidden="true" tabindex="-1"></a>afps_df_long <span class="op">=</span> afps_df_long.drop(columns<span class="op">=</span>[<span class="st">"variable"</span>])</span>
<span id="cb7-9"><a href="#cb7-9" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb7-10"><a href="#cb7-10" aria-hidden="true" tabindex="-1"></a><span class="co"># now pivot the dataframe to create new columns based on the type column</span></span>
<span id="cb7-11"><a href="#cb7-11" aria-hidden="true" tabindex="-1"></a>afps_df_lpv <span class="op">=</span> afps_df_long.pivot(index<span class="op">=</span>[<span class="st">'region_id'</span>, <span class="st">'sample_id'</span>], columns<span class="op">=</span><span class="st">'type'</span>, values<span class="op">=</span><span class="st">'value'</span>).reset_index()</span>
<span id="cb7-12"><a href="#cb7-12" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb7-13"><a href="#cb7-13" aria-hidden="true" tabindex="-1"></a><span class="co"># remove the index name and rename the columns to match the type values</span></span>
<span id="cb7-14"><a href="#cb7-14" aria-hidden="true" tabindex="-1"></a>afps_df_lpv <span class="op">=</span> afps_df_lpv.rename_axis(<span class="va">None</span>, axis<span class="op">=</span><span class="dv">1</span>).rename(columns<span class="op">=</span>{<span class="st">'fps'</span>: <span class="st">'FPS'</span>})</span>
<span id="cb7-15"><a href="#cb7-15" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb7-16"><a href="#cb7-16" aria-hidden="true" tabindex="-1"></a><span class="co"># sort the dataframe by region_id naturally</span></span>
<span id="cb7-17"><a href="#cb7-17" aria-hidden="true" tabindex="-1"></a>afps_df_lpv <span class="op">=</span> afps_df_lpv.reindex(index<span class="op">=</span>index_natsorted(afps_df_lpv[<span class="st">'region_id'</span>]))</span>
<span id="cb7-18"><a href="#cb7-18" aria-hidden="true" tabindex="-1"></a>afps_df_lpv <span class="op">=</span> afps_df_lpv.reset_index(drop<span class="op">=</span><span class="va">True</span>)</span>
<span id="cb7-19"><a href="#cb7-19" aria-hidden="true" tabindex="-1"></a>afps_df_lpv</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display" data-execution_count="8">
<div>


<table class="dataframe table table-sm table-striped small" data-quarto-postprocess="true" data-border="1">
<thead>
<tr class="header">
<th data-quarto-table-cell-role="th"></th>
<th data-quarto-table-cell-role="th">region_id</th>
<th data-quarto-table-cell-role="th">sample_id</th>
<th data-quarto-table-cell-role="th">AF</th>
<th data-quarto-table-cell-role="th">FPS</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td data-quarto-table-cell-role="th">0</td>
<td>chr1:191257-191268</td>
<td>2GAMBDQ_norm</td>
<td>0.000000</td>
<td>0.0</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">1</td>
<td>chr1:191257-191268</td>
<td>98JKPD8_lumA</td>
<td>0.000000</td>
<td>0.0</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">2</td>
<td>chr1:191257-191268</td>
<td>ANAB5F7_basal</td>
<td>0.000000</td>
<td>0.0</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">3</td>
<td>chr1:191257-191268</td>
<td>PU24GB8_lumB</td>
<td>0.026316</td>
<td>0.0</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">4</td>
<td>chr1:191257-191268</td>
<td>S6R691V_her2</td>
<td>0.000000</td>
<td>0.0</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">...</td>
<td>...</td>
<td>...</td>
<td>...</td>
<td>...</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">47795</td>
<td>chrY:11108019-11108030</td>
<td>2GAMBDQ_norm</td>
<td>0.437500</td>
<td>0.0</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">47796</td>
<td>chrY:11108019-11108030</td>
<td>98JKPD8_lumA</td>
<td>0.100000</td>
<td>0.0</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">47797</td>
<td>chrY:11108019-11108030</td>
<td>ANAB5F7_basal</td>
<td>0.382353</td>
<td>0.0</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">47798</td>
<td>chrY:11108019-11108030</td>
<td>PU24GB8_lumB</td>
<td>0.264706</td>
<td>0.0</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">47799</td>
<td>chrY:11108019-11108030</td>
<td>S6R691V_her2</td>
<td>0.437500</td>
<td>0.0</td>
</tr>
</tbody>
</table>

<p>47800 rows × 4 columns</p>
</div>
</div>
</div>
</section>
<section id="affps-matrix" class="level3">
<h3 class="anchored" data-anchor-id="affps-matrix">AF–FPS Matrix</h3>
<p>At this point, we now have both a matrix of AF and FPS values in wide form (<code>afps_df</code>) and the same dataset in long form (<code>afps_df_lpv</code>). We can scale the FPS values to be between 0 and 1 now.</p>
<div id="cell-11" class="cell" data-execution_count="9">
<div class="sourceCode cell-code" id="cb8"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb8-1"><a href="#cb8-1" aria-hidden="true" tabindex="-1"></a><span class="co"># use MinMaxScaler to scale the raw fps values to range between 0 and 1</span></span>
<span id="cb8-2"><a href="#cb8-2" aria-hidden="true" tabindex="-1"></a><span class="im">from</span> sklearn.preprocessing <span class="im">import</span> MinMaxScaler</span>
<span id="cb8-3"><a href="#cb8-3" aria-hidden="true" tabindex="-1"></a><span class="co"># scale the FPS values to a range of 0-1</span></span>
<span id="cb8-4"><a href="#cb8-4" aria-hidden="true" tabindex="-1"></a><span class="co"># Initialize a MinMaxScaler</span></span>
<span id="cb8-5"><a href="#cb8-5" aria-hidden="true" tabindex="-1"></a>scaler <span class="op">=</span> MinMaxScaler()</span>
<span id="cb8-6"><a href="#cb8-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb8-7"><a href="#cb8-7" aria-hidden="true" tabindex="-1"></a><span class="co"># copy df</span></span>
<span id="cb8-8"><a href="#cb8-8" aria-hidden="true" tabindex="-1"></a>fps_df_scaled <span class="op">=</span> dt_afps.<span class="bu">filter</span>(regex<span class="op">=</span><span class="st">'_fps$|_id$'</span>).copy()</span>
<span id="cb8-9"><a href="#cb8-9" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb8-10"><a href="#cb8-10" aria-hidden="true" tabindex="-1"></a><span class="co"># set the index to 'region_id'</span></span>
<span id="cb8-11"><a href="#cb8-11" aria-hidden="true" tabindex="-1"></a>fps_df_scaled <span class="op">=</span> fps_df_scaled.set_index(<span class="st">'region_id'</span>)</span>
<span id="cb8-12"><a href="#cb8-12" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb8-13"><a href="#cb8-13" aria-hidden="true" tabindex="-1"></a><span class="co"># Fit the MinMaxScaler to the 'FPS' column and transform it</span></span>
<span id="cb8-14"><a href="#cb8-14" aria-hidden="true" tabindex="-1"></a>fps_df_scaled <span class="op">=</span> pd.DataFrame(scaler.fit_transform(fps_df_scaled), columns<span class="op">=</span>fps_df_scaled.columns, index<span class="op">=</span>fps_df_scaled.index)</span>
<span id="cb8-15"><a href="#cb8-15" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb8-16"><a href="#cb8-16" aria-hidden="true" tabindex="-1"></a><span class="co"># rename columns by adding '_scaled' to the column names</span></span>
<span id="cb8-17"><a href="#cb8-17" aria-hidden="true" tabindex="-1"></a>fps_df_scaled <span class="op">=</span> fps_df_scaled.add_suffix(<span class="st">'_scaled'</span>)</span>
<span id="cb8-18"><a href="#cb8-18" aria-hidden="true" tabindex="-1"></a>fps_df_scaled</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display" data-execution_count="9">
<div>


<table class="dataframe table table-sm table-striped small" data-quarto-postprocess="true" data-border="1">
<thead>
<tr class="header">
<th data-quarto-table-cell-role="th"></th>
<th data-quarto-table-cell-role="th">98JKPD8_lumA_fps_scaled</th>
<th data-quarto-table-cell-role="th">ANAB5F7_basal_fps_scaled</th>
<th data-quarto-table-cell-role="th">S6R691V_her2_fps_scaled</th>
<th data-quarto-table-cell-role="th">PU24GB8_lumB_fps_scaled</th>
<th data-quarto-table-cell-role="th">2GAMBDQ_norm_fps_scaled</th>
</tr>
<tr class="odd">
<th data-quarto-table-cell-role="th">region_id</th>
<th data-quarto-table-cell-role="th"></th>
<th data-quarto-table-cell-role="th"></th>
<th data-quarto-table-cell-role="th"></th>
<th data-quarto-table-cell-role="th"></th>
<th data-quarto-table-cell-role="th"></th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td data-quarto-table-cell-role="th">chr1:191257-191268</td>
<td>0.000000</td>
<td>0.000000</td>
<td>0.000000</td>
<td>0.000000</td>
<td>0.000000</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">chr1:818021-818032</td>
<td>0.007823</td>
<td>0.011176</td>
<td>0.002785</td>
<td>0.009443</td>
<td>0.009147</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">chr1:818780-818791</td>
<td>0.006783</td>
<td>0.006036</td>
<td>0.002968</td>
<td>0.007658</td>
<td>0.009009</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">chr1:960570-960581</td>
<td>0.073969</td>
<td>0.085110</td>
<td>0.037244</td>
<td>0.075158</td>
<td>0.061066</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">chr1:977015-977026</td>
<td>0.290240</td>
<td>0.128336</td>
<td>0.092845</td>
<td>0.330112</td>
<td>0.345779</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">...</td>
<td>...</td>
<td>...</td>
<td>...</td>
<td>...</td>
<td>...</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">chrX:153934861-153934872</td>
<td>0.051878</td>
<td>0.042657</td>
<td>0.022497</td>
<td>0.051831</td>
<td>0.042934</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">chrX:154444629-154444640</td>
<td>0.012923</td>
<td>0.018879</td>
<td>0.009550</td>
<td>0.024527</td>
<td>0.011246</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">chrX:155010397-155010408</td>
<td>0.003104</td>
<td>0.001263</td>
<td>0.001144</td>
<td>0.003532</td>
<td>0.005279</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">chrX:155641610-155641621</td>
<td>0.003294</td>
<td>0.003158</td>
<td>0.001737</td>
<td>0.004060</td>
<td>0.004623</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">chrY:11108019-11108030</td>
<td>0.000000</td>
<td>0.000000</td>
<td>0.000000</td>
<td>0.000000</td>
<td>0.000000</td>
</tr>
</tbody>
</table>

<p>9560 rows × 5 columns</p>
</div>
</div>
</div>
<p>Now we should convert the FPS scaled dataframe into long form.</p>
<div id="cell-13" class="cell" data-execution_count="10">
<div class="sourceCode cell-code" id="cb9"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb9-1"><a href="#cb9-1" aria-hidden="true" tabindex="-1"></a><span class="co"># reset index</span></span>
<span id="cb9-2"><a href="#cb9-2" aria-hidden="true" tabindex="-1"></a>fps_df_scaled_long <span class="op">=</span> fps_df_scaled.reset_index()</span>
<span id="cb9-3"><a href="#cb9-3" aria-hidden="true" tabindex="-1"></a><span class="co"># convert to long format</span></span>
<span id="cb9-4"><a href="#cb9-4" aria-hidden="true" tabindex="-1"></a>fps_df_scaled_long <span class="op">=</span> fps_df_scaled_long.melt(id_vars<span class="op">=</span>[<span class="st">"region_id"</span>], var_name<span class="op">=</span><span class="st">"variable"</span>, value_name<span class="op">=</span><span class="st">"value"</span>)</span>
<span id="cb9-5"><a href="#cb9-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb9-6"><a href="#cb9-6" aria-hidden="true" tabindex="-1"></a><span class="co"># split the variable column into sample_id and type columns using reverse split string method, which returns a dataframe of columns based on the number of splits (n=x); this can directly be assigned to new columns in the original dataframe</span></span>
<span id="cb9-7"><a href="#cb9-7" aria-hidden="true" tabindex="-1"></a><span class="co"># Split the 'variable' column into three parts</span></span>
<span id="cb9-8"><a href="#cb9-8" aria-hidden="true" tabindex="-1"></a>fps_df_scaled_long[[<span class="st">'part1'</span>, <span class="st">'part2'</span>, <span class="st">'part3'</span>]] <span class="op">=</span> fps_df_scaled_long[<span class="st">'variable'</span>].<span class="bu">str</span>.rsplit(<span class="st">'_'</span>, n<span class="op">=</span><span class="dv">2</span>, expand<span class="op">=</span><span class="va">True</span>)</span>
<span id="cb9-9"><a href="#cb9-9" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb9-10"><a href="#cb9-10" aria-hidden="true" tabindex="-1"></a><span class="co"># Assign part1 to 'sample_id' and concatenate the other parts to form 'type'</span></span>
<span id="cb9-11"><a href="#cb9-11" aria-hidden="true" tabindex="-1"></a>fps_df_scaled_long[<span class="st">'sample_id'</span>] <span class="op">=</span> fps_df_scaled_long[<span class="st">'part1'</span>]</span>
<span id="cb9-12"><a href="#cb9-12" aria-hidden="true" tabindex="-1"></a>fps_df_scaled_long[<span class="st">'type'</span>] <span class="op">=</span> fps_df_scaled_long[<span class="st">'part2'</span>].<span class="bu">str</span>.upper() <span class="op">+</span> <span class="st">'_'</span> <span class="op">+</span> fps_df_scaled_long[<span class="st">'part3'</span>]</span>
<span id="cb9-13"><a href="#cb9-13" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb9-14"><a href="#cb9-14" aria-hidden="true" tabindex="-1"></a><span class="co"># Drop the unnecessary columns</span></span>
<span id="cb9-15"><a href="#cb9-15" aria-hidden="true" tabindex="-1"></a>fps_df_scaled_long <span class="op">=</span> fps_df_scaled_long.drop([<span class="st">'variable'</span>, <span class="st">'part1'</span>, <span class="st">'part2'</span>, <span class="st">'part3'</span>], axis<span class="op">=</span><span class="dv">1</span>)</span>
<span id="cb9-16"><a href="#cb9-16" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb9-17"><a href="#cb9-17" aria-hidden="true" tabindex="-1"></a><span class="co"># now pivot the dataframe to create new columns based on the type column</span></span>
<span id="cb9-18"><a href="#cb9-18" aria-hidden="true" tabindex="-1"></a>fps_df_scaled_lpv <span class="op">=</span> fps_df_scaled_long.pivot(index<span class="op">=</span>[<span class="st">'region_id'</span>, <span class="st">'sample_id'</span>], columns<span class="op">=</span><span class="st">'type'</span>, values<span class="op">=</span><span class="st">'value'</span>).reset_index()</span>
<span id="cb9-19"><a href="#cb9-19" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb9-20"><a href="#cb9-20" aria-hidden="true" tabindex="-1"></a><span class="co"># remove the index name and rename the columns to match the type values</span></span>
<span id="cb9-21"><a href="#cb9-21" aria-hidden="true" tabindex="-1"></a>fps_df_scaled_lpv <span class="op">=</span> fps_df_scaled_lpv.rename_axis(<span class="va">None</span>, axis<span class="op">=</span><span class="dv">1</span>)</span>
<span id="cb9-22"><a href="#cb9-22" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb9-23"><a href="#cb9-23" aria-hidden="true" tabindex="-1"></a><span class="co"># sort the dataframe by region_id naturally</span></span>
<span id="cb9-24"><a href="#cb9-24" aria-hidden="true" tabindex="-1"></a>fps_df_scaled_lpv <span class="op">=</span> fps_df_scaled_lpv.reindex(index<span class="op">=</span>index_natsorted(fps_df_scaled_lpv[<span class="st">'region_id'</span>]))</span>
<span id="cb9-25"><a href="#cb9-25" aria-hidden="true" tabindex="-1"></a>fps_df_scaled_lpv <span class="op">=</span> fps_df_scaled_lpv.reset_index(drop<span class="op">=</span><span class="va">True</span>)</span>
<span id="cb9-26"><a href="#cb9-26" aria-hidden="true" tabindex="-1"></a>fps_df_scaled_lpv</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display" data-execution_count="10">
<div>


<table class="dataframe table table-sm table-striped small" data-quarto-postprocess="true" data-border="1">
<thead>
<tr class="header">
<th data-quarto-table-cell-role="th"></th>
<th data-quarto-table-cell-role="th">region_id</th>
<th data-quarto-table-cell-role="th">sample_id</th>
<th data-quarto-table-cell-role="th">FPS_scaled</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td data-quarto-table-cell-role="th">0</td>
<td>chr1:191257-191268</td>
<td>2GAMBDQ_norm</td>
<td>0.0</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">1</td>
<td>chr1:191257-191268</td>
<td>98JKPD8_lumA</td>
<td>0.0</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">2</td>
<td>chr1:191257-191268</td>
<td>ANAB5F7_basal</td>
<td>0.0</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">3</td>
<td>chr1:191257-191268</td>
<td>PU24GB8_lumB</td>
<td>0.0</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">4</td>
<td>chr1:191257-191268</td>
<td>S6R691V_her2</td>
<td>0.0</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">...</td>
<td>...</td>
<td>...</td>
<td>...</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">47795</td>
<td>chrY:11108019-11108030</td>
<td>2GAMBDQ_norm</td>
<td>0.0</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">47796</td>
<td>chrY:11108019-11108030</td>
<td>98JKPD8_lumA</td>
<td>0.0</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">47797</td>
<td>chrY:11108019-11108030</td>
<td>ANAB5F7_basal</td>
<td>0.0</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">47798</td>
<td>chrY:11108019-11108030</td>
<td>PU24GB8_lumB</td>
<td>0.0</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">47799</td>
<td>chrY:11108019-11108030</td>
<td>S6R691V_her2</td>
<td>0.0</td>
</tr>
</tbody>
</table>

<p>47800 rows × 3 columns</p>
</div>
</div>
</div>
<div id="cell-14" class="cell" data-execution_count="9">
<div class="sourceCode cell-code" id="cb10"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb10-1"><a href="#cb10-1" aria-hidden="true" tabindex="-1"></a><span class="co"># plot fps distribution</span></span>
<span id="cb10-2"><a href="#cb10-2" aria-hidden="true" tabindex="-1"></a>plt.figure(figsize<span class="op">=</span>(<span class="dv">10</span>, <span class="dv">5</span>), dpi<span class="op">=</span><span class="dv">300</span>)</span>
<span id="cb10-3"><a href="#cb10-3" aria-hidden="true" tabindex="-1"></a>sns.kdeplot(data<span class="op">=</span>fps_df_scaled_lpv, x<span class="op">=</span><span class="st">"FPS_scaled"</span>, hue<span class="op">=</span><span class="st">"sample_id"</span>, fill<span class="op">=</span><span class="va">False</span>, common_norm<span class="op">=</span><span class="va">False</span>, alpha<span class="op">=</span><span class="fl">.5</span>, linewidth<span class="op">=</span><span class="dv">1</span>)</span>
<span id="cb10-4"><a href="#cb10-4" aria-hidden="true" tabindex="-1"></a>plt.show()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><img src="variant-freq-analysis-v5_files/figure-html/cell-10-output-1.png" class="img-fluid figure-img"></p>
</figure>
</div>
</div>
</div>
</section>
<section id="merging-afps_df_lpv-and-fps_df_scaled_lpv" class="level3">
<h3 class="anchored" data-anchor-id="merging-afps_df_lpv-and-fps_df_scaled_lpv">Merging <code>afps_df_lpv</code> and <code>fps_df_scaled_lpv</code></h3>
<p>Now we can merge the two long dataframes together to get both raw FPS and scaled FPS values in one dataframe.</p>
<div id="cell-16" class="cell" data-execution_count="11">
<div class="sourceCode cell-code" id="cb11"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb11-1"><a href="#cb11-1" aria-hidden="true" tabindex="-1"></a>afps_df_lpv</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display" data-execution_count="11">
<div>


<table class="dataframe table table-sm table-striped small" data-quarto-postprocess="true" data-border="1">
<thead>
<tr class="header">
<th data-quarto-table-cell-role="th"></th>
<th data-quarto-table-cell-role="th">region_id</th>
<th data-quarto-table-cell-role="th">sample_id</th>
<th data-quarto-table-cell-role="th">AF</th>
<th data-quarto-table-cell-role="th">FPS</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td data-quarto-table-cell-role="th">0</td>
<td>chr1:191257-191268</td>
<td>2GAMBDQ_norm</td>
<td>0.000000</td>
<td>0.0</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">1</td>
<td>chr1:191257-191268</td>
<td>98JKPD8_lumA</td>
<td>0.000000</td>
<td>0.0</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">2</td>
<td>chr1:191257-191268</td>
<td>ANAB5F7_basal</td>
<td>0.000000</td>
<td>0.0</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">3</td>
<td>chr1:191257-191268</td>
<td>PU24GB8_lumB</td>
<td>0.026316</td>
<td>0.0</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">4</td>
<td>chr1:191257-191268</td>
<td>S6R691V_her2</td>
<td>0.000000</td>
<td>0.0</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">...</td>
<td>...</td>
<td>...</td>
<td>...</td>
<td>...</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">47795</td>
<td>chrY:11108019-11108030</td>
<td>2GAMBDQ_norm</td>
<td>0.437500</td>
<td>0.0</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">47796</td>
<td>chrY:11108019-11108030</td>
<td>98JKPD8_lumA</td>
<td>0.100000</td>
<td>0.0</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">47797</td>
<td>chrY:11108019-11108030</td>
<td>ANAB5F7_basal</td>
<td>0.382353</td>
<td>0.0</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">47798</td>
<td>chrY:11108019-11108030</td>
<td>PU24GB8_lumB</td>
<td>0.264706</td>
<td>0.0</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">47799</td>
<td>chrY:11108019-11108030</td>
<td>S6R691V_her2</td>
<td>0.437500</td>
<td>0.0</td>
</tr>
</tbody>
</table>

<p>47800 rows × 4 columns</p>
</div>
</div>
</div>
<div id="cell-17" class="cell" data-execution_count="13">
<div class="sourceCode cell-code" id="cb12"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb12-1"><a href="#cb12-1" aria-hidden="true" tabindex="-1"></a><span class="co"># merge two dataframes on region_id and sample_id</span></span>
<span id="cb12-2"><a href="#cb12-2" aria-hidden="true" tabindex="-1"></a>afps_full_dfl <span class="op">=</span> afps_df_lpv.merge(fps_df_scaled_lpv, on<span class="op">=</span>[<span class="st">'region_id'</span>, <span class="st">'sample_id'</span>])</span>
<span id="cb12-3"><a href="#cb12-3" aria-hidden="true" tabindex="-1"></a>afps_full_dfl</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display" data-execution_count="13">
<div>


<table class="dataframe table table-sm table-striped small" data-quarto-postprocess="true" data-border="1">
<thead>
<tr class="header">
<th data-quarto-table-cell-role="th"></th>
<th data-quarto-table-cell-role="th">region_id</th>
<th data-quarto-table-cell-role="th">sample_id</th>
<th data-quarto-table-cell-role="th">AF</th>
<th data-quarto-table-cell-role="th">FPS</th>
<th data-quarto-table-cell-role="th">FPS_scaled</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td data-quarto-table-cell-role="th">0</td>
<td>chr1:191257-191268</td>
<td>2GAMBDQ_norm</td>
<td>0.000000</td>
<td>0.0</td>
<td>0.0</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">1</td>
<td>chr1:191257-191268</td>
<td>98JKPD8_lumA</td>
<td>0.000000</td>
<td>0.0</td>
<td>0.0</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">2</td>
<td>chr1:191257-191268</td>
<td>ANAB5F7_basal</td>
<td>0.000000</td>
<td>0.0</td>
<td>0.0</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">3</td>
<td>chr1:191257-191268</td>
<td>PU24GB8_lumB</td>
<td>0.026316</td>
<td>0.0</td>
<td>0.0</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">4</td>
<td>chr1:191257-191268</td>
<td>S6R691V_her2</td>
<td>0.000000</td>
<td>0.0</td>
<td>0.0</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">...</td>
<td>...</td>
<td>...</td>
<td>...</td>
<td>...</td>
<td>...</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">47795</td>
<td>chrY:11108019-11108030</td>
<td>2GAMBDQ_norm</td>
<td>0.437500</td>
<td>0.0</td>
<td>0.0</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">47796</td>
<td>chrY:11108019-11108030</td>
<td>98JKPD8_lumA</td>
<td>0.100000</td>
<td>0.0</td>
<td>0.0</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">47797</td>
<td>chrY:11108019-11108030</td>
<td>ANAB5F7_basal</td>
<td>0.382353</td>
<td>0.0</td>
<td>0.0</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">47798</td>
<td>chrY:11108019-11108030</td>
<td>PU24GB8_lumB</td>
<td>0.264706</td>
<td>0.0</td>
<td>0.0</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">47799</td>
<td>chrY:11108019-11108030</td>
<td>S6R691V_her2</td>
<td>0.437500</td>
<td>0.0</td>
<td>0.0</td>
</tr>
</tbody>
</table>

<p>47800 rows × 5 columns</p>
</div>
</div>
</div>
<div id="cell-18" class="cell" data-execution_count="15">
<div class="sourceCode cell-code" id="cb13"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb13-1"><a href="#cb13-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Create a figure and a 2x2 grid of subplots</span></span>
<span id="cb13-2"><a href="#cb13-2" aria-hidden="true" tabindex="-1"></a>fig, axs <span class="op">=</span> plt.subplots(<span class="dv">2</span>, <span class="dv">1</span>, figsize<span class="op">=</span>(<span class="dv">10</span>, <span class="dv">10</span>), dpi<span class="op">=</span><span class="dv">300</span>)</span>
<span id="cb13-3"><a href="#cb13-3" aria-hidden="true" tabindex="-1"></a>sns.histplot(data<span class="op">=</span>afps_full_dfl, x<span class="op">=</span><span class="st">"FPS"</span>, hue<span class="op">=</span><span class="st">"sample_id"</span>, bins<span class="op">=</span><span class="dv">200</span>, kde<span class="op">=</span><span class="va">True</span>, palette<span class="op">=</span><span class="st">"deep"</span>, ax<span class="op">=</span>axs[<span class="dv">0</span>])</span>
<span id="cb13-4"><a href="#cb13-4" aria-hidden="true" tabindex="-1"></a>sns.histplot(data<span class="op">=</span>afps_full_dfl, x<span class="op">=</span><span class="st">"FPS_scaled"</span>, hue<span class="op">=</span><span class="st">"sample_id"</span>, bins<span class="op">=</span><span class="dv">200</span>, kde<span class="op">=</span><span class="va">True</span>, palette<span class="op">=</span><span class="st">"deep"</span>, ax<span class="op">=</span>axs[<span class="dv">1</span>])</span>
<span id="cb13-5"><a href="#cb13-5" aria-hidden="true" tabindex="-1"></a><span class="co"># # now draw a subset plot where x is above a certain threshold</span></span>
<span id="cb13-6"><a href="#cb13-6" aria-hidden="true" tabindex="-1"></a><span class="co"># sns.histplot(data=afps_full_dfl[afps_full_dfl['FPS'] &gt; 0.5], x="FPS", hue="sample_id", bins=200, kde=True, palette="deep", ax=axs[2])</span></span>
<span id="cb13-7"><a href="#cb13-7" aria-hidden="true" tabindex="-1"></a><span class="co"># plt.title(f"FPS distribution of {motif_id} in all samples")</span></span>
<span id="cb13-8"><a href="#cb13-8" aria-hidden="true" tabindex="-1"></a>plt.show()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><img src="variant-freq-analysis-v5_files/figure-html/cell-13-output-1.png" class="img-fluid figure-img"></p>
</figure>
</div>
</div>
</div>
<div id="cell-19" class="cell" data-execution_count="14">
<div class="sourceCode cell-code" id="cb14"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb14-1"><a href="#cb14-1" aria-hidden="true" tabindex="-1"></a>g <span class="op">=</span> sns.jointplot(data<span class="op">=</span>afps_full_dfl, x<span class="op">=</span><span class="st">'AF'</span>, y<span class="op">=</span><span class="st">'FPS_scaled'</span>, kind<span class="op">=</span><span class="st">'scatter'</span>, hue<span class="op">=</span><span class="st">'sample_id'</span>, height<span class="op">=</span><span class="dv">12</span>)</span>
<span id="cb14-2"><a href="#cb14-2" aria-hidden="true" tabindex="-1"></a>plt.xlim(<span class="op">-</span><span class="fl">0.1</span>, <span class="fl">1.1</span>)</span>
<span id="cb14-3"><a href="#cb14-3" aria-hidden="true" tabindex="-1"></a>g.figure.suptitle(<span class="ss">f"Jointpot of AF and scaled FPS values of </span><span class="sc">{</span>motif_id<span class="sc">}</span><span class="ss">"</span>, fontsize<span class="op">=</span><span class="dv">14</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display" data-execution_count="14">
<pre><code>Text(0.5, 0.98, 'Jointpot of AF and scaled FPS values of TFDP1_M08108_2.00')</code></pre>
</div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><img src="variant-freq-analysis-v5_files/figure-html/cell-14-output-2.png" class="img-fluid figure-img"></p>
</figure>
</div>
</div>
</div>
</section>
<section id="filtering-zero-sites" class="level3">
<h3 class="anchored" data-anchor-id="filtering-zero-sites">Filtering Zero Sites</h3>
<p>We should filter out the sites that have zero AF values in ALL subtypes, as these are not informative for our analysis. We will also filter out the sites that have zero FPS values in ALL subtypes.</p>
<div id="cell-21" class="cell" data-execution_count="15">
<div class="sourceCode cell-code" id="cb16"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb16-1"><a href="#cb16-1" aria-hidden="true" tabindex="-1"></a>afps_full_dfl</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display" data-execution_count="15">
<div>


<table class="dataframe table table-sm table-striped small" data-quarto-postprocess="true" data-border="1">
<thead>
<tr class="header">
<th data-quarto-table-cell-role="th"></th>
<th data-quarto-table-cell-role="th">region_id</th>
<th data-quarto-table-cell-role="th">sample_id</th>
<th data-quarto-table-cell-role="th">AF</th>
<th data-quarto-table-cell-role="th">FPS</th>
<th data-quarto-table-cell-role="th">FPS_scaled</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td data-quarto-table-cell-role="th">0</td>
<td>chr1:191257-191268</td>
<td>2GAMBDQ_norm</td>
<td>0.000000</td>
<td>0.0</td>
<td>0.0</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">1</td>
<td>chr1:191257-191268</td>
<td>98JKPD8_lumA</td>
<td>0.000000</td>
<td>0.0</td>
<td>0.0</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">2</td>
<td>chr1:191257-191268</td>
<td>ANAB5F7_basal</td>
<td>0.000000</td>
<td>0.0</td>
<td>0.0</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">3</td>
<td>chr1:191257-191268</td>
<td>PU24GB8_lumB</td>
<td>0.026316</td>
<td>0.0</td>
<td>0.0</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">4</td>
<td>chr1:191257-191268</td>
<td>S6R691V_her2</td>
<td>0.000000</td>
<td>0.0</td>
<td>0.0</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">...</td>
<td>...</td>
<td>...</td>
<td>...</td>
<td>...</td>
<td>...</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">47795</td>
<td>chrY:11108019-11108030</td>
<td>2GAMBDQ_norm</td>
<td>0.437500</td>
<td>0.0</td>
<td>0.0</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">47796</td>
<td>chrY:11108019-11108030</td>
<td>98JKPD8_lumA</td>
<td>0.100000</td>
<td>0.0</td>
<td>0.0</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">47797</td>
<td>chrY:11108019-11108030</td>
<td>ANAB5F7_basal</td>
<td>0.382353</td>
<td>0.0</td>
<td>0.0</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">47798</td>
<td>chrY:11108019-11108030</td>
<td>PU24GB8_lumB</td>
<td>0.264706</td>
<td>0.0</td>
<td>0.0</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">47799</td>
<td>chrY:11108019-11108030</td>
<td>S6R691V_her2</td>
<td>0.437500</td>
<td>0.0</td>
<td>0.0</td>
</tr>
</tbody>
</table>

<p>47800 rows × 5 columns</p>
</div>
</div>
</div>
<p><code>FPS</code> value <code>== 0</code> means that the raw FPS values across subtypes are zero, i.e.&nbsp;there is no evidence of footprinting at all at these sites despite a TFBS matching a motif being present (assigned by TOBIAS). AF value <code>== 0</code> means that there is no evidence of a variant genotype at these sites across the samples making up a subtype.</p>
<div id="cell-23" class="cell" data-execution_count="16">
<div class="sourceCode cell-code" id="cb17"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb17-1"><a href="#cb17-1" aria-hidden="true" tabindex="-1"></a><span class="co"># filter out unique region_id rows that have fps == 0 across the sample_ids and AF == 0</span></span>
<span id="cb17-2"><a href="#cb17-2" aria-hidden="true" tabindex="-1"></a><span class="co"># group by 'region_id' first </span></span>
<span id="cb17-3"><a href="#cb17-3" aria-hidden="true" tabindex="-1"></a>merged_filt_dfl <span class="op">=</span> afps_full_dfl.groupby(<span class="st">'region_id'</span>).<span class="bu">filter</span>(<span class="kw">lambda</span> x: x[<span class="st">'FPS'</span>].<span class="bu">sum</span>() <span class="op">&gt;</span> <span class="dv">0</span> <span class="kw">and</span> x[<span class="st">'AF'</span>].<span class="bu">sum</span>() <span class="op">&gt;</span> <span class="dv">0</span>)</span>
<span id="cb17-4"><a href="#cb17-4" aria-hidden="true" tabindex="-1"></a>merged_filt_dfl</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display" data-execution_count="16">
<div>


<table class="dataframe table table-sm table-striped small" data-quarto-postprocess="true" data-border="1">
<thead>
<tr class="header">
<th data-quarto-table-cell-role="th"></th>
<th data-quarto-table-cell-role="th">region_id</th>
<th data-quarto-table-cell-role="th">sample_id</th>
<th data-quarto-table-cell-role="th">AF</th>
<th data-quarto-table-cell-role="th">FPS</th>
<th data-quarto-table-cell-role="th">FPS_scaled</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td data-quarto-table-cell-role="th">5</td>
<td>chr1:818021-818032</td>
<td>2GAMBDQ_norm</td>
<td>0.875000</td>
<td>0.03443</td>
<td>0.009147</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">6</td>
<td>chr1:818021-818032</td>
<td>98JKPD8_lumA</td>
<td>0.900000</td>
<td>0.02966</td>
<td>0.007823</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">7</td>
<td>chr1:818021-818032</td>
<td>ANAB5F7_basal</td>
<td>0.812500</td>
<td>0.04399</td>
<td>0.011176</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">8</td>
<td>chr1:818021-818032</td>
<td>PU24GB8_lumB</td>
<td>0.833333</td>
<td>0.03470</td>
<td>0.009443</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">9</td>
<td>chr1:818021-818032</td>
<td>S6R691V_her2</td>
<td>0.900000</td>
<td>0.02131</td>
<td>0.002785</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">...</td>
<td>...</td>
<td>...</td>
<td>...</td>
<td>...</td>
<td>...</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">47790</td>
<td>chrX:155641610-155641621</td>
<td>2GAMBDQ_norm</td>
<td>0.000000</td>
<td>0.01740</td>
<td>0.004623</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">47791</td>
<td>chrX:155641610-155641621</td>
<td>98JKPD8_lumA</td>
<td>0.000000</td>
<td>0.01249</td>
<td>0.003294</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">47792</td>
<td>chrX:155641610-155641621</td>
<td>ANAB5F7_basal</td>
<td>0.000000</td>
<td>0.01243</td>
<td>0.003158</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">47793</td>
<td>chrX:155641610-155641621</td>
<td>PU24GB8_lumB</td>
<td>0.029412</td>
<td>0.01492</td>
<td>0.004060</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">47794</td>
<td>chrX:155641610-155641621</td>
<td>S6R691V_her2</td>
<td>0.000000</td>
<td>0.01329</td>
<td>0.001737</td>
</tr>
</tbody>
</table>

<p>43530 rows × 5 columns</p>
</div>
</div>
</div>
</section>
<section id="calculating-variation-statistics" class="level3">
<h3 class="anchored" data-anchor-id="calculating-variation-statistics">Calculating Variation Statistics</h3>
<p>With the full data table cleaned up into a long format, we can now calculate the variation statistics for each region. We will calculate variance (Var).</p>
<ol type="A">
<li>First calculate statistics on the raw FPS data.</li>
</ol>
<div id="cell-26" class="cell" data-execution_count="17">
<div class="sourceCode cell-code" id="cb18"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb18-1"><a href="#cb18-1" aria-hidden="true" tabindex="-1"></a><span class="co"># extract fps columns</span></span>
<span id="cb18-2"><a href="#cb18-2" aria-hidden="true" tabindex="-1"></a>fps_df <span class="op">=</span> dt_afps.<span class="bu">filter</span>(regex<span class="op">=</span><span class="st">'_fps$|_id$'</span>).copy()</span>
<span id="cb18-3"><a href="#cb18-3" aria-hidden="true" tabindex="-1"></a>fps_df</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display" data-execution_count="17">
<div>


<table class="dataframe table table-sm table-striped small" data-quarto-postprocess="true" data-border="1">
<thead>
<tr class="header">
<th data-quarto-table-cell-role="th"></th>
<th data-quarto-table-cell-role="th">98JKPD8_lumA_fps</th>
<th data-quarto-table-cell-role="th">ANAB5F7_basal_fps</th>
<th data-quarto-table-cell-role="th">S6R691V_her2_fps</th>
<th data-quarto-table-cell-role="th">PU24GB8_lumB_fps</th>
<th data-quarto-table-cell-role="th">2GAMBDQ_norm_fps</th>
<th data-quarto-table-cell-role="th">region_id</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td data-quarto-table-cell-role="th">0</td>
<td>0.00000</td>
<td>0.00000</td>
<td>0.00000</td>
<td>0.00000</td>
<td>0.00000</td>
<td>chr1:191257-191268</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">1</td>
<td>0.02966</td>
<td>0.04399</td>
<td>0.02131</td>
<td>0.03470</td>
<td>0.03443</td>
<td>chr1:818021-818032</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">2</td>
<td>0.02572</td>
<td>0.02376</td>
<td>0.02271</td>
<td>0.02814</td>
<td>0.03391</td>
<td>chr1:818780-818791</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">3</td>
<td>0.28046</td>
<td>0.33500</td>
<td>0.28495</td>
<td>0.27618</td>
<td>0.22986</td>
<td>chr1:960570-960581</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">4</td>
<td>1.10047</td>
<td>0.50514</td>
<td>0.71035</td>
<td>1.21305</td>
<td>1.30155</td>
<td>chr1:977015-977026</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">...</td>
<td>...</td>
<td>...</td>
<td>...</td>
<td>...</td>
<td>...</td>
<td>...</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">9555</td>
<td>0.19670</td>
<td>0.16790</td>
<td>0.17212</td>
<td>0.19046</td>
<td>0.16161</td>
<td>chrX:153934861-153934872</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">9556</td>
<td>0.04900</td>
<td>0.07431</td>
<td>0.07307</td>
<td>0.09013</td>
<td>0.04233</td>
<td>chrX:154444629-154444640</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">9557</td>
<td>0.01177</td>
<td>0.00497</td>
<td>0.00875</td>
<td>0.01298</td>
<td>0.01987</td>
<td>chrX:155010397-155010408</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">9558</td>
<td>0.01249</td>
<td>0.01243</td>
<td>0.01329</td>
<td>0.01492</td>
<td>0.01740</td>
<td>chrX:155641610-155641621</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">9559</td>
<td>0.00000</td>
<td>0.00000</td>
<td>0.00000</td>
<td>0.00000</td>
<td>0.00000</td>
<td>chrY:11108019-11108030</td>
</tr>
</tbody>
</table>

<p>9560 rows × 6 columns</p>
</div>
</div>
</div>
<p>Now filter out the rows whose <code>region_id</code> is not in the <code>region_id</code> column of <code>merged_filt</code>. First, reduce the repeated <code>region_id</code> column to unique values in <code>merged_filt</code>.</p>
<div id="cell-28" class="cell">
<div class="sourceCode cell-code" id="cb19"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb19-1"><a href="#cb19-1" aria-hidden="true" tabindex="-1"></a><span class="co"># this creates an array</span></span>
<span id="cb19-2"><a href="#cb19-2" aria-hidden="true" tabindex="-1"></a>merged_filt_uniq_region_id <span class="op">=</span> merged_filt_dfl[<span class="st">'region_id'</span>].unique()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div id="cell-29" class="cell" data-execution_count="20">
<div class="sourceCode cell-code" id="cb20"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb20-1"><a href="#cb20-1" aria-hidden="true" tabindex="-1"></a><span class="co"># then using this array, keep only the rows in the fps_df that have `region_id` values in the array</span></span>
<span id="cb20-2"><a href="#cb20-2" aria-hidden="true" tabindex="-1"></a>fps_df_filt <span class="op">=</span> fps_df[fps_df[<span class="st">'region_id'</span>].isin(merged_filt_uniq_region_id)]</span>
<span id="cb20-3"><a href="#cb20-3" aria-hidden="true" tabindex="-1"></a>fps_df_filt</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display" data-execution_count="20">
<div>


<table class="dataframe table table-sm table-striped small" data-quarto-postprocess="true" data-border="1">
<thead>
<tr class="header">
<th data-quarto-table-cell-role="th"></th>
<th data-quarto-table-cell-role="th">98JKPD8_lumA_fps</th>
<th data-quarto-table-cell-role="th">ANAB5F7_basal_fps</th>
<th data-quarto-table-cell-role="th">S6R691V_her2_fps</th>
<th data-quarto-table-cell-role="th">PU24GB8_lumB_fps</th>
<th data-quarto-table-cell-role="th">2GAMBDQ_norm_fps</th>
<th data-quarto-table-cell-role="th">region_id</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td data-quarto-table-cell-role="th">1</td>
<td>0.02966</td>
<td>0.04399</td>
<td>0.02131</td>
<td>0.03470</td>
<td>0.03443</td>
<td>chr1:818021-818032</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">2</td>
<td>0.02572</td>
<td>0.02376</td>
<td>0.02271</td>
<td>0.02814</td>
<td>0.03391</td>
<td>chr1:818780-818791</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">3</td>
<td>0.28046</td>
<td>0.33500</td>
<td>0.28495</td>
<td>0.27618</td>
<td>0.22986</td>
<td>chr1:960570-960581</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">4</td>
<td>1.10047</td>
<td>0.50514</td>
<td>0.71035</td>
<td>1.21305</td>
<td>1.30155</td>
<td>chr1:977015-977026</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">5</td>
<td>0.88179</td>
<td>0.40974</td>
<td>0.53173</td>
<td>0.98628</td>
<td>1.07101</td>
<td>chr1:977047-977058</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">...</td>
<td>...</td>
<td>...</td>
<td>...</td>
<td>...</td>
<td>...</td>
<td>...</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">9552</td>
<td>0.02192</td>
<td>0.03699</td>
<td>0.01875</td>
<td>0.01472</td>
<td>0.02839</td>
<td>chrX:153642534-153642545</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">9555</td>
<td>0.19670</td>
<td>0.16790</td>
<td>0.17212</td>
<td>0.19046</td>
<td>0.16161</td>
<td>chrX:153934861-153934872</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">9556</td>
<td>0.04900</td>
<td>0.07431</td>
<td>0.07307</td>
<td>0.09013</td>
<td>0.04233</td>
<td>chrX:154444629-154444640</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">9557</td>
<td>0.01177</td>
<td>0.00497</td>
<td>0.00875</td>
<td>0.01298</td>
<td>0.01987</td>
<td>chrX:155010397-155010408</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">9558</td>
<td>0.01249</td>
<td>0.01243</td>
<td>0.01329</td>
<td>0.01492</td>
<td>0.01740</td>
<td>chrX:155641610-155641621</td>
</tr>
</tbody>
</table>

<p>8706 rows × 6 columns</p>
</div>
</div>
</div>
<div id="cell-30" class="cell" data-execution_count="21">
<div class="sourceCode cell-code" id="cb21"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb21-1"><a href="#cb21-1" aria-hidden="true" tabindex="-1"></a><span class="co"># calculate variance of fps values across samples per region_id and add to a new column called 'fps_var'</span></span>
<span id="cb21-2"><a href="#cb21-2" aria-hidden="true" tabindex="-1"></a>fps_df_filt_ind <span class="op">=</span> fps_df_filt.set_index(<span class="st">'region_id'</span>)</span>
<span id="cb21-3"><a href="#cb21-3" aria-hidden="true" tabindex="-1"></a>fps_df_filt_ind[<span class="st">'FPS_var'</span>] <span class="op">=</span> fps_df_filt_ind.var(axis<span class="op">=</span><span class="dv">1</span>)</span>
<span id="cb21-4"><a href="#cb21-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb21-5"><a href="#cb21-5" aria-hidden="true" tabindex="-1"></a>fps_df_filt_ind</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display" data-execution_count="21">
<div>


<table class="dataframe table table-sm table-striped small" data-quarto-postprocess="true" data-border="1">
<thead>
<tr class="header">
<th data-quarto-table-cell-role="th"></th>
<th data-quarto-table-cell-role="th">98JKPD8_lumA_fps</th>
<th data-quarto-table-cell-role="th">ANAB5F7_basal_fps</th>
<th data-quarto-table-cell-role="th">S6R691V_her2_fps</th>
<th data-quarto-table-cell-role="th">PU24GB8_lumB_fps</th>
<th data-quarto-table-cell-role="th">2GAMBDQ_norm_fps</th>
<th data-quarto-table-cell-role="th">FPS_var</th>
</tr>
<tr class="odd">
<th data-quarto-table-cell-role="th">region_id</th>
<th data-quarto-table-cell-role="th"></th>
<th data-quarto-table-cell-role="th"></th>
<th data-quarto-table-cell-role="th"></th>
<th data-quarto-table-cell-role="th"></th>
<th data-quarto-table-cell-role="th"></th>
<th data-quarto-table-cell-role="th"></th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td data-quarto-table-cell-role="th">chr1:818021-818032</td>
<td>0.02966</td>
<td>0.04399</td>
<td>0.02131</td>
<td>0.03470</td>
<td>0.03443</td>
<td>0.000068</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">chr1:818780-818791</td>
<td>0.02572</td>
<td>0.02376</td>
<td>0.02271</td>
<td>0.02814</td>
<td>0.03391</td>
<td>0.000020</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">chr1:960570-960581</td>
<td>0.28046</td>
<td>0.33500</td>
<td>0.28495</td>
<td>0.27618</td>
<td>0.22986</td>
<td>0.001393</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">chr1:977015-977026</td>
<td>1.10047</td>
<td>0.50514</td>
<td>0.71035</td>
<td>1.21305</td>
<td>1.30155</td>
<td>0.117365</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">chr1:977047-977058</td>
<td>0.88179</td>
<td>0.40974</td>
<td>0.53173</td>
<td>0.98628</td>
<td>1.07101</td>
<td>0.084064</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">...</td>
<td>...</td>
<td>...</td>
<td>...</td>
<td>...</td>
<td>...</td>
<td>...</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">chrX:153642534-153642545</td>
<td>0.02192</td>
<td>0.03699</td>
<td>0.01875</td>
<td>0.01472</td>
<td>0.02839</td>
<td>0.000076</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">chrX:153934861-153934872</td>
<td>0.19670</td>
<td>0.16790</td>
<td>0.17212</td>
<td>0.19046</td>
<td>0.16161</td>
<td>0.000227</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">chrX:154444629-154444640</td>
<td>0.04900</td>
<td>0.07431</td>
<td>0.07307</td>
<td>0.09013</td>
<td>0.04233</td>
<td>0.000388</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">chrX:155010397-155010408</td>
<td>0.01177</td>
<td>0.00497</td>
<td>0.00875</td>
<td>0.01298</td>
<td>0.01987</td>
<td>0.000031</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">chrX:155641610-155641621</td>
<td>0.01249</td>
<td>0.01243</td>
<td>0.01329</td>
<td>0.01492</td>
<td>0.01740</td>
<td>0.000004</td>
</tr>
</tbody>
</table>

<p>8706 rows × 6 columns</p>
</div>
</div>
</div>
<ol start="2" type="A">
<li>Then do the same on the fps scaled data.</li>
</ol>
<p>Filter out the rows whose <code>region_id</code> is not in the <code>region_id</code> column of <code>merged_filt_scaled</code>.</p>
<div id="cell-33" class="cell" data-execution_count="25">
<div class="sourceCode cell-code" id="cb22"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb22-1"><a href="#cb22-1" aria-hidden="true" tabindex="-1"></a>fps_df_scaled</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display" data-execution_count="25">
<div>


<table class="dataframe table table-sm table-striped small" data-quarto-postprocess="true" data-border="1">
<thead>
<tr class="header">
<th data-quarto-table-cell-role="th"></th>
<th data-quarto-table-cell-role="th">98JKPD8_lumA_fps_scaled</th>
<th data-quarto-table-cell-role="th">ANAB5F7_basal_fps_scaled</th>
<th data-quarto-table-cell-role="th">S6R691V_her2_fps_scaled</th>
<th data-quarto-table-cell-role="th">PU24GB8_lumB_fps_scaled</th>
<th data-quarto-table-cell-role="th">2GAMBDQ_norm_fps_scaled</th>
</tr>
<tr class="odd">
<th data-quarto-table-cell-role="th">region_id</th>
<th data-quarto-table-cell-role="th"></th>
<th data-quarto-table-cell-role="th"></th>
<th data-quarto-table-cell-role="th"></th>
<th data-quarto-table-cell-role="th"></th>
<th data-quarto-table-cell-role="th"></th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td data-quarto-table-cell-role="th">chr1:191257-191268</td>
<td>0.000000</td>
<td>0.000000</td>
<td>0.000000</td>
<td>0.000000</td>
<td>0.000000</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">chr1:818021-818032</td>
<td>0.007823</td>
<td>0.011176</td>
<td>0.002785</td>
<td>0.009443</td>
<td>0.009147</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">chr1:818780-818791</td>
<td>0.006783</td>
<td>0.006036</td>
<td>0.002968</td>
<td>0.007658</td>
<td>0.009009</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">chr1:960570-960581</td>
<td>0.073969</td>
<td>0.085110</td>
<td>0.037244</td>
<td>0.075158</td>
<td>0.061066</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">chr1:977015-977026</td>
<td>0.290240</td>
<td>0.128336</td>
<td>0.092845</td>
<td>0.330112</td>
<td>0.345779</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">...</td>
<td>...</td>
<td>...</td>
<td>...</td>
<td>...</td>
<td>...</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">chrX:153934861-153934872</td>
<td>0.051878</td>
<td>0.042657</td>
<td>0.022497</td>
<td>0.051831</td>
<td>0.042934</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">chrX:154444629-154444640</td>
<td>0.012923</td>
<td>0.018879</td>
<td>0.009550</td>
<td>0.024527</td>
<td>0.011246</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">chrX:155010397-155010408</td>
<td>0.003104</td>
<td>0.001263</td>
<td>0.001144</td>
<td>0.003532</td>
<td>0.005279</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">chrX:155641610-155641621</td>
<td>0.003294</td>
<td>0.003158</td>
<td>0.001737</td>
<td>0.004060</td>
<td>0.004623</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">chrY:11108019-11108030</td>
<td>0.000000</td>
<td>0.000000</td>
<td>0.000000</td>
<td>0.000000</td>
<td>0.000000</td>
</tr>
</tbody>
</table>

<p>9560 rows × 5 columns</p>
</div>
</div>
</div>
<div id="cell-34" class="cell" data-execution_count="22">
<div class="sourceCode cell-code" id="cb23"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb23-1"><a href="#cb23-1" aria-hidden="true" tabindex="-1"></a><span class="co"># region_id in the scaled df is already used as index, so filter by index</span></span>
<span id="cb23-2"><a href="#cb23-2" aria-hidden="true" tabindex="-1"></a>fps_df_scaled_filt <span class="op">=</span> fps_df_scaled[fps_df_scaled.index.isin(merged_filt_uniq_region_id)]</span>
<span id="cb23-3"><a href="#cb23-3" aria-hidden="true" tabindex="-1"></a>fps_df_scaled_filt</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display" data-execution_count="22">
<div>


<table class="dataframe table table-sm table-striped small" data-quarto-postprocess="true" data-border="1">
<thead>
<tr class="header">
<th data-quarto-table-cell-role="th"></th>
<th data-quarto-table-cell-role="th">98JKPD8_lumA_fps_scaled</th>
<th data-quarto-table-cell-role="th">ANAB5F7_basal_fps_scaled</th>
<th data-quarto-table-cell-role="th">S6R691V_her2_fps_scaled</th>
<th data-quarto-table-cell-role="th">PU24GB8_lumB_fps_scaled</th>
<th data-quarto-table-cell-role="th">2GAMBDQ_norm_fps_scaled</th>
</tr>
<tr class="odd">
<th data-quarto-table-cell-role="th">region_id</th>
<th data-quarto-table-cell-role="th"></th>
<th data-quarto-table-cell-role="th"></th>
<th data-quarto-table-cell-role="th"></th>
<th data-quarto-table-cell-role="th"></th>
<th data-quarto-table-cell-role="th"></th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td data-quarto-table-cell-role="th">chr1:818021-818032</td>
<td>0.007823</td>
<td>0.011176</td>
<td>0.002785</td>
<td>0.009443</td>
<td>0.009147</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">chr1:818780-818791</td>
<td>0.006783</td>
<td>0.006036</td>
<td>0.002968</td>
<td>0.007658</td>
<td>0.009009</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">chr1:960570-960581</td>
<td>0.073969</td>
<td>0.085110</td>
<td>0.037244</td>
<td>0.075158</td>
<td>0.061066</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">chr1:977015-977026</td>
<td>0.290240</td>
<td>0.128336</td>
<td>0.092845</td>
<td>0.330112</td>
<td>0.345779</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">chr1:977047-977058</td>
<td>0.232565</td>
<td>0.104098</td>
<td>0.069499</td>
<td>0.268400</td>
<td>0.284532</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">...</td>
<td>...</td>
<td>...</td>
<td>...</td>
<td>...</td>
<td>...</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">chrX:153642534-153642545</td>
<td>0.005781</td>
<td>0.009398</td>
<td>0.002451</td>
<td>0.004006</td>
<td>0.007542</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">chrX:153934861-153934872</td>
<td>0.051878</td>
<td>0.042657</td>
<td>0.022497</td>
<td>0.051831</td>
<td>0.042934</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">chrX:154444629-154444640</td>
<td>0.012923</td>
<td>0.018879</td>
<td>0.009550</td>
<td>0.024527</td>
<td>0.011246</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">chrX:155010397-155010408</td>
<td>0.003104</td>
<td>0.001263</td>
<td>0.001144</td>
<td>0.003532</td>
<td>0.005279</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">chrX:155641610-155641621</td>
<td>0.003294</td>
<td>0.003158</td>
<td>0.001737</td>
<td>0.004060</td>
<td>0.004623</td>
</tr>
</tbody>
</table>

<p>8706 rows × 5 columns</p>
</div>
</div>
</div>
<div id="cell-35" class="cell" data-execution_count="23">
<div class="sourceCode cell-code" id="cb24"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb24-1"><a href="#cb24-1" aria-hidden="true" tabindex="-1"></a><span class="co"># copy the slice</span></span>
<span id="cb24-2"><a href="#cb24-2" aria-hidden="true" tabindex="-1"></a>fps_df_scaled_filt_ind <span class="op">=</span> fps_df_scaled_filt.copy()</span>
<span id="cb24-3"><a href="#cb24-3" aria-hidden="true" tabindex="-1"></a>fps_df_scaled_filt_ind</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display" data-execution_count="23">
<div>


<table class="dataframe table table-sm table-striped small" data-quarto-postprocess="true" data-border="1">
<thead>
<tr class="header">
<th data-quarto-table-cell-role="th"></th>
<th data-quarto-table-cell-role="th">98JKPD8_lumA_fps_scaled</th>
<th data-quarto-table-cell-role="th">ANAB5F7_basal_fps_scaled</th>
<th data-quarto-table-cell-role="th">S6R691V_her2_fps_scaled</th>
<th data-quarto-table-cell-role="th">PU24GB8_lumB_fps_scaled</th>
<th data-quarto-table-cell-role="th">2GAMBDQ_norm_fps_scaled</th>
</tr>
<tr class="odd">
<th data-quarto-table-cell-role="th">region_id</th>
<th data-quarto-table-cell-role="th"></th>
<th data-quarto-table-cell-role="th"></th>
<th data-quarto-table-cell-role="th"></th>
<th data-quarto-table-cell-role="th"></th>
<th data-quarto-table-cell-role="th"></th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td data-quarto-table-cell-role="th">chr1:818021-818032</td>
<td>0.007823</td>
<td>0.011176</td>
<td>0.002785</td>
<td>0.009443</td>
<td>0.009147</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">chr1:818780-818791</td>
<td>0.006783</td>
<td>0.006036</td>
<td>0.002968</td>
<td>0.007658</td>
<td>0.009009</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">chr1:960570-960581</td>
<td>0.073969</td>
<td>0.085110</td>
<td>0.037244</td>
<td>0.075158</td>
<td>0.061066</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">chr1:977015-977026</td>
<td>0.290240</td>
<td>0.128336</td>
<td>0.092845</td>
<td>0.330112</td>
<td>0.345779</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">chr1:977047-977058</td>
<td>0.232565</td>
<td>0.104098</td>
<td>0.069499</td>
<td>0.268400</td>
<td>0.284532</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">...</td>
<td>...</td>
<td>...</td>
<td>...</td>
<td>...</td>
<td>...</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">chrX:153642534-153642545</td>
<td>0.005781</td>
<td>0.009398</td>
<td>0.002451</td>
<td>0.004006</td>
<td>0.007542</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">chrX:153934861-153934872</td>
<td>0.051878</td>
<td>0.042657</td>
<td>0.022497</td>
<td>0.051831</td>
<td>0.042934</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">chrX:154444629-154444640</td>
<td>0.012923</td>
<td>0.018879</td>
<td>0.009550</td>
<td>0.024527</td>
<td>0.011246</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">chrX:155010397-155010408</td>
<td>0.003104</td>
<td>0.001263</td>
<td>0.001144</td>
<td>0.003532</td>
<td>0.005279</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">chrX:155641610-155641621</td>
<td>0.003294</td>
<td>0.003158</td>
<td>0.001737</td>
<td>0.004060</td>
<td>0.004623</td>
</tr>
</tbody>
</table>

<p>8706 rows × 5 columns</p>
</div>
</div>
</div>
<div id="cell-36" class="cell" data-execution_count="24">
<div class="sourceCode cell-code" id="cb25"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb25-1"><a href="#cb25-1" aria-hidden="true" tabindex="-1"></a><span class="co"># calculate variance of fps_scaled values across samples per region_id and add to a new column called 'fps_scaled_var'</span></span>
<span id="cb25-2"><a href="#cb25-2" aria-hidden="true" tabindex="-1"></a>fps_df_scaled_filt_ind[<span class="st">'FPS_scaled_var'</span>] <span class="op">=</span> fps_df_scaled_filt_ind.var(axis<span class="op">=</span><span class="dv">1</span>)</span>
<span id="cb25-3"><a href="#cb25-3" aria-hidden="true" tabindex="-1"></a>fps_df_scaled_filt_ind</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display" data-execution_count="24">
<div>


<table class="dataframe table table-sm table-striped small" data-quarto-postprocess="true" data-border="1">
<thead>
<tr class="header">
<th data-quarto-table-cell-role="th"></th>
<th data-quarto-table-cell-role="th">98JKPD8_lumA_fps_scaled</th>
<th data-quarto-table-cell-role="th">ANAB5F7_basal_fps_scaled</th>
<th data-quarto-table-cell-role="th">S6R691V_her2_fps_scaled</th>
<th data-quarto-table-cell-role="th">PU24GB8_lumB_fps_scaled</th>
<th data-quarto-table-cell-role="th">2GAMBDQ_norm_fps_scaled</th>
<th data-quarto-table-cell-role="th">FPS_scaled_var</th>
</tr>
<tr class="odd">
<th data-quarto-table-cell-role="th">region_id</th>
<th data-quarto-table-cell-role="th"></th>
<th data-quarto-table-cell-role="th"></th>
<th data-quarto-table-cell-role="th"></th>
<th data-quarto-table-cell-role="th"></th>
<th data-quarto-table-cell-role="th"></th>
<th data-quarto-table-cell-role="th"></th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td data-quarto-table-cell-role="th">chr1:818021-818032</td>
<td>0.007823</td>
<td>0.011176</td>
<td>0.002785</td>
<td>0.009443</td>
<td>0.009147</td>
<td>0.000010</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">chr1:818780-818791</td>
<td>0.006783</td>
<td>0.006036</td>
<td>0.002968</td>
<td>0.007658</td>
<td>0.009009</td>
<td>0.000005</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">chr1:960570-960581</td>
<td>0.073969</td>
<td>0.085110</td>
<td>0.037244</td>
<td>0.075158</td>
<td>0.061066</td>
<td>0.000341</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">chr1:977015-977026</td>
<td>0.290240</td>
<td>0.128336</td>
<td>0.092845</td>
<td>0.330112</td>
<td>0.345779</td>
<td>0.013981</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">chr1:977047-977058</td>
<td>0.232565</td>
<td>0.104098</td>
<td>0.069499</td>
<td>0.268400</td>
<td>0.284532</td>
<td>0.009694</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">...</td>
<td>...</td>
<td>...</td>
<td>...</td>
<td>...</td>
<td>...</td>
<td>...</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">chrX:153642534-153642545</td>
<td>0.005781</td>
<td>0.009398</td>
<td>0.002451</td>
<td>0.004006</td>
<td>0.007542</td>
<td>0.000008</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">chrX:153934861-153934872</td>
<td>0.051878</td>
<td>0.042657</td>
<td>0.022497</td>
<td>0.051831</td>
<td>0.042934</td>
<td>0.000144</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">chrX:154444629-154444640</td>
<td>0.012923</td>
<td>0.018879</td>
<td>0.009550</td>
<td>0.024527</td>
<td>0.011246</td>
<td>0.000038</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">chrX:155010397-155010408</td>
<td>0.003104</td>
<td>0.001263</td>
<td>0.001144</td>
<td>0.003532</td>
<td>0.005279</td>
<td>0.000003</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">chrX:155641610-155641621</td>
<td>0.003294</td>
<td>0.003158</td>
<td>0.001737</td>
<td>0.004060</td>
<td>0.004623</td>
<td>0.000001</td>
</tr>
</tbody>
</table>

<p>8706 rows × 6 columns</p>
</div>
</div>
</div>
<ol start="3" type="A">
<li>Now do the same for the AF data.</li>
</ol>
<div id="cell-38" class="cell" data-execution_count="25">
<div class="sourceCode cell-code" id="cb26"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb26-1"><a href="#cb26-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Do the same for AF values</span></span>
<span id="cb26-2"><a href="#cb26-2" aria-hidden="true" tabindex="-1"></a><span class="co"># extract af columns</span></span>
<span id="cb26-3"><a href="#cb26-3" aria-hidden="true" tabindex="-1"></a>af_df <span class="op">=</span> dt_afps.<span class="bu">filter</span>(regex<span class="op">=</span><span class="st">'_AF$|_id$'</span>)</span>
<span id="cb26-4"><a href="#cb26-4" aria-hidden="true" tabindex="-1"></a>af_df</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display" data-execution_count="25">
<div>


<table class="dataframe table table-sm table-striped small" data-quarto-postprocess="true" data-border="1">
<thead>
<tr class="header">
<th data-quarto-table-cell-role="th"></th>
<th data-quarto-table-cell-role="th">2GAMBDQ_norm_AF</th>
<th data-quarto-table-cell-role="th">98JKPD8_lumA_AF</th>
<th data-quarto-table-cell-role="th">ANAB5F7_basal_AF</th>
<th data-quarto-table-cell-role="th">PU24GB8_lumB_AF</th>
<th data-quarto-table-cell-role="th">S6R691V_her2_AF</th>
<th data-quarto-table-cell-role="th">region_id</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td data-quarto-table-cell-role="th">0</td>
<td>0.000000</td>
<td>0.000000</td>
<td>0.000000</td>
<td>0.026316</td>
<td>0.000000</td>
<td>chr1:191257-191268</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">1</td>
<td>0.875000</td>
<td>0.900000</td>
<td>0.812500</td>
<td>0.833333</td>
<td>0.900000</td>
<td>chr1:818021-818032</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">2</td>
<td>0.000000</td>
<td>0.125000</td>
<td>0.033333</td>
<td>0.055556</td>
<td>0.062500</td>
<td>chr1:818780-818791</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">3</td>
<td>0.000000</td>
<td>0.041667</td>
<td>0.000000</td>
<td>0.000000</td>
<td>0.000000</td>
<td>chr1:960570-960581</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">4</td>
<td>1.000000</td>
<td>1.000000</td>
<td>0.941176</td>
<td>0.947368</td>
<td>0.911765</td>
<td>chr1:977015-977026</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">...</td>
<td>...</td>
<td>...</td>
<td>...</td>
<td>...</td>
<td>...</td>
<td>...</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">9555</td>
<td>0.062500</td>
<td>0.000000</td>
<td>0.000000</td>
<td>0.000000</td>
<td>0.000000</td>
<td>chrX:153934861-153934872</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">9556</td>
<td>0.000000</td>
<td>0.125000</td>
<td>0.058824</td>
<td>0.105263</td>
<td>0.000000</td>
<td>chrX:154444629-154444640</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">9557</td>
<td>0.083333</td>
<td>0.312500</td>
<td>0.000000</td>
<td>0.392857</td>
<td>0.100000</td>
<td>chrX:155010397-155010408</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">9558</td>
<td>0.000000</td>
<td>0.000000</td>
<td>0.000000</td>
<td>0.029412</td>
<td>0.000000</td>
<td>chrX:155641610-155641621</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">9559</td>
<td>0.437500</td>
<td>0.100000</td>
<td>0.382353</td>
<td>0.264706</td>
<td>0.437500</td>
<td>chrY:11108019-11108030</td>
</tr>
</tbody>
</table>

<p>9560 rows × 6 columns</p>
</div>
</div>
</div>
<div id="cell-39" class="cell" data-execution_count="26">
<div class="sourceCode cell-code" id="cb27"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb27-1"><a href="#cb27-1" aria-hidden="true" tabindex="-1"></a><span class="co"># filter by region_id</span></span>
<span id="cb27-2"><a href="#cb27-2" aria-hidden="true" tabindex="-1"></a>af_df_filt <span class="op">=</span> af_df[af_df[<span class="st">'region_id'</span>].isin(merged_filt_uniq_region_id)]</span>
<span id="cb27-3"><a href="#cb27-3" aria-hidden="true" tabindex="-1"></a>af_df_filt</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display" data-execution_count="26">
<div>


<table class="dataframe table table-sm table-striped small" data-quarto-postprocess="true" data-border="1">
<thead>
<tr class="header">
<th data-quarto-table-cell-role="th"></th>
<th data-quarto-table-cell-role="th">2GAMBDQ_norm_AF</th>
<th data-quarto-table-cell-role="th">98JKPD8_lumA_AF</th>
<th data-quarto-table-cell-role="th">ANAB5F7_basal_AF</th>
<th data-quarto-table-cell-role="th">PU24GB8_lumB_AF</th>
<th data-quarto-table-cell-role="th">S6R691V_her2_AF</th>
<th data-quarto-table-cell-role="th">region_id</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td data-quarto-table-cell-role="th">1</td>
<td>0.875000</td>
<td>0.900000</td>
<td>0.812500</td>
<td>0.833333</td>
<td>0.900000</td>
<td>chr1:818021-818032</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">2</td>
<td>0.000000</td>
<td>0.125000</td>
<td>0.033333</td>
<td>0.055556</td>
<td>0.062500</td>
<td>chr1:818780-818791</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">3</td>
<td>0.000000</td>
<td>0.041667</td>
<td>0.000000</td>
<td>0.000000</td>
<td>0.000000</td>
<td>chr1:960570-960581</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">4</td>
<td>1.000000</td>
<td>1.000000</td>
<td>0.941176</td>
<td>0.947368</td>
<td>0.911765</td>
<td>chr1:977015-977026</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">5</td>
<td>0.750000</td>
<td>1.000000</td>
<td>0.500000</td>
<td>0.789474</td>
<td>0.588235</td>
<td>chr1:977047-977058</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">...</td>
<td>...</td>
<td>...</td>
<td>...</td>
<td>...</td>
<td>...</td>
<td>...</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">9552</td>
<td>0.062500</td>
<td>0.000000</td>
<td>0.000000</td>
<td>0.000000</td>
<td>0.000000</td>
<td>chrX:153642534-153642545</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">9555</td>
<td>0.062500</td>
<td>0.000000</td>
<td>0.000000</td>
<td>0.000000</td>
<td>0.000000</td>
<td>chrX:153934861-153934872</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">9556</td>
<td>0.000000</td>
<td>0.125000</td>
<td>0.058824</td>
<td>0.105263</td>
<td>0.000000</td>
<td>chrX:154444629-154444640</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">9557</td>
<td>0.083333</td>
<td>0.312500</td>
<td>0.000000</td>
<td>0.392857</td>
<td>0.100000</td>
<td>chrX:155010397-155010408</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">9558</td>
<td>0.000000</td>
<td>0.000000</td>
<td>0.000000</td>
<td>0.029412</td>
<td>0.000000</td>
<td>chrX:155641610-155641621</td>
</tr>
</tbody>
</table>

<p>8706 rows × 6 columns</p>
</div>
</div>
</div>
<div id="cell-40" class="cell" data-execution_count="27">
<div class="sourceCode cell-code" id="cb28"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb28-1"><a href="#cb28-1" aria-hidden="true" tabindex="-1"></a>af_df_filt_ind <span class="op">=</span> af_df_filt.set_index(<span class="st">'region_id'</span>)</span>
<span id="cb28-2"><a href="#cb28-2" aria-hidden="true" tabindex="-1"></a><span class="co"># then calculate variance of af values across samples per region_id and add to a new column called 'af_var'</span></span>
<span id="cb28-3"><a href="#cb28-3" aria-hidden="true" tabindex="-1"></a>af_df_filt_ind[<span class="st">'AF_var'</span>] <span class="op">=</span> af_df_filt_ind.var(axis<span class="op">=</span><span class="dv">1</span>)</span>
<span id="cb28-4"><a href="#cb28-4" aria-hidden="true" tabindex="-1"></a>af_df_filt_ind</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display" data-execution_count="27">
<div>


<table class="dataframe table table-sm table-striped small" data-quarto-postprocess="true" data-border="1">
<thead>
<tr class="header">
<th data-quarto-table-cell-role="th"></th>
<th data-quarto-table-cell-role="th">2GAMBDQ_norm_AF</th>
<th data-quarto-table-cell-role="th">98JKPD8_lumA_AF</th>
<th data-quarto-table-cell-role="th">ANAB5F7_basal_AF</th>
<th data-quarto-table-cell-role="th">PU24GB8_lumB_AF</th>
<th data-quarto-table-cell-role="th">S6R691V_her2_AF</th>
<th data-quarto-table-cell-role="th">AF_var</th>
</tr>
<tr class="odd">
<th data-quarto-table-cell-role="th">region_id</th>
<th data-quarto-table-cell-role="th"></th>
<th data-quarto-table-cell-role="th"></th>
<th data-quarto-table-cell-role="th"></th>
<th data-quarto-table-cell-role="th"></th>
<th data-quarto-table-cell-role="th"></th>
<th data-quarto-table-cell-role="th"></th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td data-quarto-table-cell-role="th">chr1:818021-818032</td>
<td>0.875000</td>
<td>0.900000</td>
<td>0.812500</td>
<td>0.833333</td>
<td>0.900000</td>
<td>0.001576</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">chr1:818780-818791</td>
<td>0.000000</td>
<td>0.125000</td>
<td>0.033333</td>
<td>0.055556</td>
<td>0.062500</td>
<td>0.002113</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">chr1:960570-960581</td>
<td>0.000000</td>
<td>0.041667</td>
<td>0.000000</td>
<td>0.000000</td>
<td>0.000000</td>
<td>0.000347</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">chr1:977015-977026</td>
<td>1.000000</td>
<td>1.000000</td>
<td>0.941176</td>
<td>0.947368</td>
<td>0.911765</td>
<td>0.001510</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">chr1:977047-977058</td>
<td>0.750000</td>
<td>1.000000</td>
<td>0.500000</td>
<td>0.789474</td>
<td>0.588235</td>
<td>0.037434</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">...</td>
<td>...</td>
<td>...</td>
<td>...</td>
<td>...</td>
<td>...</td>
<td>...</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">chrX:153642534-153642545</td>
<td>0.062500</td>
<td>0.000000</td>
<td>0.000000</td>
<td>0.000000</td>
<td>0.000000</td>
<td>0.000781</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">chrX:153934861-153934872</td>
<td>0.062500</td>
<td>0.000000</td>
<td>0.000000</td>
<td>0.000000</td>
<td>0.000000</td>
<td>0.000781</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">chrX:154444629-154444640</td>
<td>0.000000</td>
<td>0.125000</td>
<td>0.058824</td>
<td>0.105263</td>
<td>0.000000</td>
<td>0.003363</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">chrX:155010397-155010408</td>
<td>0.083333</td>
<td>0.312500</td>
<td>0.000000</td>
<td>0.392857</td>
<td>0.100000</td>
<td>0.027746</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">chrX:155641610-155641621</td>
<td>0.000000</td>
<td>0.000000</td>
<td>0.000000</td>
<td>0.029412</td>
<td>0.000000</td>
<td>0.000173</td>
</tr>
</tbody>
</table>

<p>8706 rows × 6 columns</p>
</div>
</div>
</div>
<ol start="4" type="A">
<li>Now merge the statistics tables together on <code>region_id</code> index.</li>
</ol>
<div id="cell-42" class="cell" data-execution_count="28">
<div class="sourceCode cell-code" id="cb29"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb29-1"><a href="#cb29-1" aria-hidden="true" tabindex="-1"></a>fps_filt_var <span class="op">=</span> fps_df_filt_ind.<span class="bu">filter</span>(regex<span class="op">=</span><span class="st">'_var$|_id$'</span>).copy()</span>
<span id="cb29-2"><a href="#cb29-2" aria-hidden="true" tabindex="-1"></a>fps_filt_var</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display" data-execution_count="28">
<div>


<table class="dataframe table table-sm table-striped small" data-quarto-postprocess="true" data-border="1">
<thead>
<tr class="header">
<th data-quarto-table-cell-role="th"></th>
<th data-quarto-table-cell-role="th">FPS_var</th>
</tr>
<tr class="odd">
<th data-quarto-table-cell-role="th">region_id</th>
<th data-quarto-table-cell-role="th"></th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td data-quarto-table-cell-role="th">chr1:818021-818032</td>
<td>0.000068</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">chr1:818780-818791</td>
<td>0.000020</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">chr1:960570-960581</td>
<td>0.001393</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">chr1:977015-977026</td>
<td>0.117365</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">chr1:977047-977058</td>
<td>0.084064</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">...</td>
<td>...</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">chrX:153642534-153642545</td>
<td>0.000076</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">chrX:153934861-153934872</td>
<td>0.000227</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">chrX:154444629-154444640</td>
<td>0.000388</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">chrX:155010397-155010408</td>
<td>0.000031</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">chrX:155641610-155641621</td>
<td>0.000004</td>
</tr>
</tbody>
</table>

<p>8706 rows × 1 columns</p>
</div>
</div>
</div>
<div id="cell-43" class="cell" data-execution_count="29">
<div class="sourceCode cell-code" id="cb30"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb30-1"><a href="#cb30-1" aria-hidden="true" tabindex="-1"></a>fps_sc_filt_var <span class="op">=</span> fps_df_scaled_filt_ind.<span class="bu">filter</span>(regex<span class="op">=</span><span class="st">'_var$|_id$'</span>).copy()</span>
<span id="cb30-2"><a href="#cb30-2" aria-hidden="true" tabindex="-1"></a>af_filt_var <span class="op">=</span> af_df_filt_ind.<span class="bu">filter</span>(regex<span class="op">=</span><span class="st">'_var$|_id$'</span>).copy()</span>
<span id="cb30-3"><a href="#cb30-3" aria-hidden="true" tabindex="-1"></a><span class="co"># merge on region_id index from both tables</span></span>
<span id="cb30-4"><a href="#cb30-4" aria-hidden="true" tabindex="-1"></a>merged_var_tmp <span class="op">=</span> fps_filt_var.merge(fps_sc_filt_var, left_index<span class="op">=</span><span class="va">True</span>, right_index<span class="op">=</span><span class="va">True</span>)</span>
<span id="cb30-5"><a href="#cb30-5" aria-hidden="true" tabindex="-1"></a><span class="co"># now merge with af_df</span></span>
<span id="cb30-6"><a href="#cb30-6" aria-hidden="true" tabindex="-1"></a>merged_var <span class="op">=</span> merged_var_tmp.merge(af_filt_var, left_index<span class="op">=</span><span class="va">True</span>, right_index<span class="op">=</span><span class="va">True</span>)</span>
<span id="cb30-7"><a href="#cb30-7" aria-hidden="true" tabindex="-1"></a>merged_var</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display" data-execution_count="29">
<div>


<table class="dataframe table table-sm table-striped small" data-quarto-postprocess="true" data-border="1">
<thead>
<tr class="header">
<th data-quarto-table-cell-role="th"></th>
<th data-quarto-table-cell-role="th">FPS_var</th>
<th data-quarto-table-cell-role="th">FPS_scaled_var</th>
<th data-quarto-table-cell-role="th">AF_var</th>
</tr>
<tr class="odd">
<th data-quarto-table-cell-role="th">region_id</th>
<th data-quarto-table-cell-role="th"></th>
<th data-quarto-table-cell-role="th"></th>
<th data-quarto-table-cell-role="th"></th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td data-quarto-table-cell-role="th">chr1:818021-818032</td>
<td>0.000068</td>
<td>0.000010</td>
<td>0.001576</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">chr1:818780-818791</td>
<td>0.000020</td>
<td>0.000005</td>
<td>0.002113</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">chr1:960570-960581</td>
<td>0.001393</td>
<td>0.000341</td>
<td>0.000347</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">chr1:977015-977026</td>
<td>0.117365</td>
<td>0.013981</td>
<td>0.001510</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">chr1:977047-977058</td>
<td>0.084064</td>
<td>0.009694</td>
<td>0.037434</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">...</td>
<td>...</td>
<td>...</td>
<td>...</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">chrX:153642534-153642545</td>
<td>0.000076</td>
<td>0.000008</td>
<td>0.000781</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">chrX:153934861-153934872</td>
<td>0.000227</td>
<td>0.000144</td>
<td>0.000781</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">chrX:154444629-154444640</td>
<td>0.000388</td>
<td>0.000038</td>
<td>0.003363</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">chrX:155010397-155010408</td>
<td>0.000031</td>
<td>0.000003</td>
<td>0.027746</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">chrX:155641610-155641621</td>
<td>0.000004</td>
<td>0.000001</td>
<td>0.000173</td>
</tr>
</tbody>
</table>

<p>8706 rows × 3 columns</p>
</div>
</div>
</div>
<p>Now we have a table of stats data to merge with the <code>merge_filt</code> raw data.</p>
<div id="cell-45" class="cell" data-execution_count="30">
<div class="sourceCode cell-code" id="cb31"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb31-1"><a href="#cb31-1" aria-hidden="true" tabindex="-1"></a><span class="co"># set the index to 'region_id'</span></span>
<span id="cb31-2"><a href="#cb31-2" aria-hidden="true" tabindex="-1"></a>merged_filt_i <span class="op">=</span> merged_filt_dfl.set_index(<span class="st">'region_id'</span>)</span>
<span id="cb31-3"><a href="#cb31-3" aria-hidden="true" tabindex="-1"></a>merged_filt_i</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display" data-execution_count="30">
<div>


<table class="dataframe table table-sm table-striped small" data-quarto-postprocess="true" data-border="1">
<thead>
<tr class="header">
<th data-quarto-table-cell-role="th"></th>
<th data-quarto-table-cell-role="th">sample_id</th>
<th data-quarto-table-cell-role="th">AF</th>
<th data-quarto-table-cell-role="th">FPS</th>
<th data-quarto-table-cell-role="th">FPS_scaled</th>
</tr>
<tr class="odd">
<th data-quarto-table-cell-role="th">region_id</th>
<th data-quarto-table-cell-role="th"></th>
<th data-quarto-table-cell-role="th"></th>
<th data-quarto-table-cell-role="th"></th>
<th data-quarto-table-cell-role="th"></th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td data-quarto-table-cell-role="th">chr1:818021-818032</td>
<td>2GAMBDQ_norm</td>
<td>0.875000</td>
<td>0.03443</td>
<td>0.009147</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">chr1:818021-818032</td>
<td>98JKPD8_lumA</td>
<td>0.900000</td>
<td>0.02966</td>
<td>0.007823</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">chr1:818021-818032</td>
<td>ANAB5F7_basal</td>
<td>0.812500</td>
<td>0.04399</td>
<td>0.011176</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">chr1:818021-818032</td>
<td>PU24GB8_lumB</td>
<td>0.833333</td>
<td>0.03470</td>
<td>0.009443</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">chr1:818021-818032</td>
<td>S6R691V_her2</td>
<td>0.900000</td>
<td>0.02131</td>
<td>0.002785</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">...</td>
<td>...</td>
<td>...</td>
<td>...</td>
<td>...</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">chrX:155641610-155641621</td>
<td>2GAMBDQ_norm</td>
<td>0.000000</td>
<td>0.01740</td>
<td>0.004623</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">chrX:155641610-155641621</td>
<td>98JKPD8_lumA</td>
<td>0.000000</td>
<td>0.01249</td>
<td>0.003294</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">chrX:155641610-155641621</td>
<td>ANAB5F7_basal</td>
<td>0.000000</td>
<td>0.01243</td>
<td>0.003158</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">chrX:155641610-155641621</td>
<td>PU24GB8_lumB</td>
<td>0.029412</td>
<td>0.01492</td>
<td>0.004060</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">chrX:155641610-155641621</td>
<td>S6R691V_her2</td>
<td>0.000000</td>
<td>0.01329</td>
<td>0.001737</td>
</tr>
</tbody>
</table>

<p>43530 rows × 4 columns</p>
</div>
</div>
</div>
<div id="cell-46" class="cell" data-execution_count="31">
<div class="sourceCode cell-code" id="cb32"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb32-1"><a href="#cb32-1" aria-hidden="true" tabindex="-1"></a><span class="co"># merge afps_full_dfli with stats df on region_id</span></span>
<span id="cb32-2"><a href="#cb32-2" aria-hidden="true" tabindex="-1"></a>merged_stat <span class="op">=</span> merged_filt_i.merge(merged_var, left_index<span class="op">=</span><span class="va">True</span>, right_index<span class="op">=</span><span class="va">True</span>, how<span class="op">=</span><span class="st">'left'</span>)</span>
<span id="cb32-3"><a href="#cb32-3" aria-hidden="true" tabindex="-1"></a>merged_stat</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display" data-execution_count="31">
<div>


<table class="dataframe table table-sm table-striped small" data-quarto-postprocess="true" data-border="1">
<thead>
<tr class="header">
<th data-quarto-table-cell-role="th"></th>
<th data-quarto-table-cell-role="th">sample_id</th>
<th data-quarto-table-cell-role="th">AF</th>
<th data-quarto-table-cell-role="th">FPS</th>
<th data-quarto-table-cell-role="th">FPS_scaled</th>
<th data-quarto-table-cell-role="th">FPS_var</th>
<th data-quarto-table-cell-role="th">FPS_scaled_var</th>
<th data-quarto-table-cell-role="th">AF_var</th>
</tr>
<tr class="odd">
<th data-quarto-table-cell-role="th">region_id</th>
<th data-quarto-table-cell-role="th"></th>
<th data-quarto-table-cell-role="th"></th>
<th data-quarto-table-cell-role="th"></th>
<th data-quarto-table-cell-role="th"></th>
<th data-quarto-table-cell-role="th"></th>
<th data-quarto-table-cell-role="th"></th>
<th data-quarto-table-cell-role="th"></th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td data-quarto-table-cell-role="th">chr1:818021-818032</td>
<td>2GAMBDQ_norm</td>
<td>0.875000</td>
<td>0.03443</td>
<td>0.009147</td>
<td>0.000068</td>
<td>0.000010</td>
<td>0.001576</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">chr1:818021-818032</td>
<td>98JKPD8_lumA</td>
<td>0.900000</td>
<td>0.02966</td>
<td>0.007823</td>
<td>0.000068</td>
<td>0.000010</td>
<td>0.001576</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">chr1:818021-818032</td>
<td>ANAB5F7_basal</td>
<td>0.812500</td>
<td>0.04399</td>
<td>0.011176</td>
<td>0.000068</td>
<td>0.000010</td>
<td>0.001576</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">chr1:818021-818032</td>
<td>PU24GB8_lumB</td>
<td>0.833333</td>
<td>0.03470</td>
<td>0.009443</td>
<td>0.000068</td>
<td>0.000010</td>
<td>0.001576</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">chr1:818021-818032</td>
<td>S6R691V_her2</td>
<td>0.900000</td>
<td>0.02131</td>
<td>0.002785</td>
<td>0.000068</td>
<td>0.000010</td>
<td>0.001576</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">...</td>
<td>...</td>
<td>...</td>
<td>...</td>
<td>...</td>
<td>...</td>
<td>...</td>
<td>...</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">chrX:155641610-155641621</td>
<td>2GAMBDQ_norm</td>
<td>0.000000</td>
<td>0.01740</td>
<td>0.004623</td>
<td>0.000004</td>
<td>0.000001</td>
<td>0.000173</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">chrX:155641610-155641621</td>
<td>98JKPD8_lumA</td>
<td>0.000000</td>
<td>0.01249</td>
<td>0.003294</td>
<td>0.000004</td>
<td>0.000001</td>
<td>0.000173</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">chrX:155641610-155641621</td>
<td>ANAB5F7_basal</td>
<td>0.000000</td>
<td>0.01243</td>
<td>0.003158</td>
<td>0.000004</td>
<td>0.000001</td>
<td>0.000173</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">chrX:155641610-155641621</td>
<td>PU24GB8_lumB</td>
<td>0.029412</td>
<td>0.01492</td>
<td>0.004060</td>
<td>0.000004</td>
<td>0.000001</td>
<td>0.000173</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">chrX:155641610-155641621</td>
<td>S6R691V_her2</td>
<td>0.000000</td>
<td>0.01329</td>
<td>0.001737</td>
<td>0.000004</td>
<td>0.000001</td>
<td>0.000173</td>
</tr>
</tbody>
</table>

<p>43530 rows × 7 columns</p>
</div>
</div>
</div>
<div id="cell-47" class="cell" data-execution_count="42">
<div class="sourceCode cell-code" id="cb33"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb33-1"><a href="#cb33-1" aria-hidden="true" tabindex="-1"></a><span class="co"># drop the FPS and FPS_var columns</span></span>
<span id="cb33-2"><a href="#cb33-2" aria-hidden="true" tabindex="-1"></a>merged_stat <span class="op">=</span> merged_stat.drop(columns<span class="op">=</span>merged_stat.<span class="bu">filter</span>(regex<span class="op">=</span><span class="st">'FPS_var|FPS$'</span>).columns)</span>
<span id="cb33-3"><a href="#cb33-3" aria-hidden="true" tabindex="-1"></a>merged_stat</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display" data-execution_count="42">
<div>


<table class="dataframe table table-sm table-striped small" data-quarto-postprocess="true" data-border="1">
<thead>
<tr class="header">
<th data-quarto-table-cell-role="th"></th>
<th data-quarto-table-cell-role="th">sample_id</th>
<th data-quarto-table-cell-role="th">AF</th>
<th data-quarto-table-cell-role="th">FPS_scaled</th>
<th data-quarto-table-cell-role="th">FPS_scaled_var</th>
<th data-quarto-table-cell-role="th">AF_var</th>
</tr>
<tr class="odd">
<th data-quarto-table-cell-role="th">region_id</th>
<th data-quarto-table-cell-role="th"></th>
<th data-quarto-table-cell-role="th"></th>
<th data-quarto-table-cell-role="th"></th>
<th data-quarto-table-cell-role="th"></th>
<th data-quarto-table-cell-role="th"></th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td data-quarto-table-cell-role="th">chr1:818021-818032</td>
<td>2GAMBDQ_norm</td>
<td>0.875000</td>
<td>0.009147</td>
<td>0.000010</td>
<td>0.001576</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">chr1:818021-818032</td>
<td>98JKPD8_lumA</td>
<td>0.900000</td>
<td>0.007823</td>
<td>0.000010</td>
<td>0.001576</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">chr1:818021-818032</td>
<td>ANAB5F7_basal</td>
<td>0.812500</td>
<td>0.011176</td>
<td>0.000010</td>
<td>0.001576</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">chr1:818021-818032</td>
<td>PU24GB8_lumB</td>
<td>0.833333</td>
<td>0.009443</td>
<td>0.000010</td>
<td>0.001576</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">chr1:818021-818032</td>
<td>S6R691V_her2</td>
<td>0.900000</td>
<td>0.002785</td>
<td>0.000010</td>
<td>0.001576</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">...</td>
<td>...</td>
<td>...</td>
<td>...</td>
<td>...</td>
<td>...</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">chrX:155641610-155641621</td>
<td>2GAMBDQ_norm</td>
<td>0.000000</td>
<td>0.004623</td>
<td>0.000001</td>
<td>0.000173</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">chrX:155641610-155641621</td>
<td>98JKPD8_lumA</td>
<td>0.000000</td>
<td>0.003294</td>
<td>0.000001</td>
<td>0.000173</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">chrX:155641610-155641621</td>
<td>ANAB5F7_basal</td>
<td>0.000000</td>
<td>0.003158</td>
<td>0.000001</td>
<td>0.000173</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">chrX:155641610-155641621</td>
<td>PU24GB8_lumB</td>
<td>0.029412</td>
<td>0.004060</td>
<td>0.000001</td>
<td>0.000173</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">chrX:155641610-155641621</td>
<td>S6R691V_her2</td>
<td>0.000000</td>
<td>0.001737</td>
<td>0.000001</td>
<td>0.000173</td>
</tr>
</tbody>
</table>

<p>43530 rows × 5 columns</p>
</div>
</div>
</div>
<div id="cell-48" class="cell" data-execution_count="32">
<div class="sourceCode cell-code" id="cb34"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb34-1"><a href="#cb34-1" aria-hidden="true" tabindex="-1"></a><span class="co"># rearrange columns</span></span>
<span id="cb34-2"><a href="#cb34-2" aria-hidden="true" tabindex="-1"></a>merged_stat <span class="op">=</span> merged_stat[[<span class="st">'sample_id'</span>, <span class="st">'AF'</span>, <span class="st">'FPS_scaled'</span>, <span class="st">'AF_var'</span>, <span class="st">'FPS_scaled_var'</span>]]</span>
<span id="cb34-3"><a href="#cb34-3" aria-hidden="true" tabindex="-1"></a>merged_stat</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display" data-execution_count="32">
<div>


<table class="dataframe table table-sm table-striped small" data-quarto-postprocess="true" data-border="1">
<thead>
<tr class="header">
<th data-quarto-table-cell-role="th"></th>
<th data-quarto-table-cell-role="th">sample_id</th>
<th data-quarto-table-cell-role="th">AF</th>
<th data-quarto-table-cell-role="th">FPS_scaled</th>
<th data-quarto-table-cell-role="th">AF_var</th>
<th data-quarto-table-cell-role="th">FPS_scaled_var</th>
</tr>
<tr class="odd">
<th data-quarto-table-cell-role="th">region_id</th>
<th data-quarto-table-cell-role="th"></th>
<th data-quarto-table-cell-role="th"></th>
<th data-quarto-table-cell-role="th"></th>
<th data-quarto-table-cell-role="th"></th>
<th data-quarto-table-cell-role="th"></th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td data-quarto-table-cell-role="th">chr1:818021-818032</td>
<td>2GAMBDQ_norm</td>
<td>0.875000</td>
<td>0.009147</td>
<td>0.001576</td>
<td>0.000010</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">chr1:818021-818032</td>
<td>98JKPD8_lumA</td>
<td>0.900000</td>
<td>0.007823</td>
<td>0.001576</td>
<td>0.000010</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">chr1:818021-818032</td>
<td>ANAB5F7_basal</td>
<td>0.812500</td>
<td>0.011176</td>
<td>0.001576</td>
<td>0.000010</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">chr1:818021-818032</td>
<td>PU24GB8_lumB</td>
<td>0.833333</td>
<td>0.009443</td>
<td>0.001576</td>
<td>0.000010</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">chr1:818021-818032</td>
<td>S6R691V_her2</td>
<td>0.900000</td>
<td>0.002785</td>
<td>0.001576</td>
<td>0.000010</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">...</td>
<td>...</td>
<td>...</td>
<td>...</td>
<td>...</td>
<td>...</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">chrX:155641610-155641621</td>
<td>2GAMBDQ_norm</td>
<td>0.000000</td>
<td>0.004623</td>
<td>0.000173</td>
<td>0.000001</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">chrX:155641610-155641621</td>
<td>98JKPD8_lumA</td>
<td>0.000000</td>
<td>0.003294</td>
<td>0.000173</td>
<td>0.000001</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">chrX:155641610-155641621</td>
<td>ANAB5F7_basal</td>
<td>0.000000</td>
<td>0.003158</td>
<td>0.000173</td>
<td>0.000001</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">chrX:155641610-155641621</td>
<td>PU24GB8_lumB</td>
<td>0.029412</td>
<td>0.004060</td>
<td>0.000173</td>
<td>0.000001</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">chrX:155641610-155641621</td>
<td>S6R691V_her2</td>
<td>0.000000</td>
<td>0.001737</td>
<td>0.000173</td>
<td>0.000001</td>
</tr>
</tbody>
</table>

<p>43530 rows × 5 columns</p>
</div>
</div>
</div>
<section id="aside-visualizing-af_var-and-fps_scaled_var" class="level4">
<h4 class="anchored" data-anchor-id="aside-visualizing-af_var-and-fps_scaled_var">ASIDE: Visualizing <code>AF_var</code> and <code>FPS_scaled_var</code></h4>
<p>We can now see if there is a correlation between the variance of the AF and FPS values, per <code>region_id</code>. As the statistics are calculated per <code>region_id</code>, we can plot the <code>FPS_scaled_var</code> and <code>AF_var</code> values against each other.</p>
<div id="cell-50" class="cell" data-execution_count="33">
<div class="sourceCode cell-code" id="cb35"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb35-1"><a href="#cb35-1" aria-hidden="true" tabindex="-1"></a><span class="co"># subset a copy of the dataframe with only the AF_var and FPS_scaled_var columns and region_id</span></span>
<span id="cb35-2"><a href="#cb35-2" aria-hidden="true" tabindex="-1"></a>merged_stat_corrplot <span class="op">=</span> merged_stat[[<span class="st">'AF_var'</span>, <span class="st">'FPS_scaled_var'</span>]].copy().drop_duplicates()</span>
<span id="cb35-3"><a href="#cb35-3" aria-hidden="true" tabindex="-1"></a>merged_stat_corrplot</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display" data-execution_count="33">
<div>


<table class="dataframe table table-sm table-striped small" data-quarto-postprocess="true" data-border="1">
<thead>
<tr class="header">
<th data-quarto-table-cell-role="th"></th>
<th data-quarto-table-cell-role="th">AF_var</th>
<th data-quarto-table-cell-role="th">FPS_scaled_var</th>
</tr>
<tr class="odd">
<th data-quarto-table-cell-role="th">region_id</th>
<th data-quarto-table-cell-role="th"></th>
<th data-quarto-table-cell-role="th"></th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td data-quarto-table-cell-role="th">chr1:818021-818032</td>
<td>0.001576</td>
<td>0.000010</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">chr1:818780-818791</td>
<td>0.002113</td>
<td>0.000005</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">chr1:960570-960581</td>
<td>0.000347</td>
<td>0.000341</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">chr1:977015-977026</td>
<td>0.001510</td>
<td>0.013981</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">chr1:977047-977058</td>
<td>0.037434</td>
<td>0.009694</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">...</td>
<td>...</td>
<td>...</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">chrX:153642534-153642545</td>
<td>0.000781</td>
<td>0.000008</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">chrX:153934861-153934872</td>
<td>0.000781</td>
<td>0.000144</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">chrX:154444629-154444640</td>
<td>0.003363</td>
<td>0.000038</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">chrX:155010397-155010408</td>
<td>0.027746</td>
<td>0.000003</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">chrX:155641610-155641621</td>
<td>0.000173</td>
<td>0.000001</td>
</tr>
</tbody>
</table>

<p>8706 rows × 2 columns</p>
</div>
</div>
</div>
<div id="cell-51" class="cell" data-execution_count="34">
<div class="sourceCode cell-code" id="cb36"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb36-1"><a href="#cb36-1" aria-hidden="true" tabindex="-1"></a><span class="co"># plot lmplot</span></span>
<span id="cb36-2"><a href="#cb36-2" aria-hidden="true" tabindex="-1"></a>sns.lmplot(data<span class="op">=</span>merged_stat_corrplot, x<span class="op">=</span><span class="st">'AF_var'</span>, y<span class="op">=</span><span class="st">'FPS_scaled_var'</span>, fit_reg<span class="op">=</span><span class="va">True</span>, height<span class="op">=</span><span class="dv">10</span>)</span>
<span id="cb36-3"><a href="#cb36-3" aria-hidden="true" tabindex="-1"></a>plt.title(<span class="ss">f'Variance of AF vs Variance of FPS_scaled for </span><span class="sc">{</span>motif_id<span class="sc">}</span><span class="ss">'</span>)</span>
<span id="cb36-4"><a href="#cb36-4" aria-hidden="true" tabindex="-1"></a>plt.show()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><img src="variant-freq-analysis-v5_files/figure-html/cell-35-output-1.png" class="img-fluid figure-img"></p>
</figure>
</div>
</div>
</div>
</section>
</section>
<section id="finding-covariant-sites" class="level3">
<h3 class="anchored" data-anchor-id="finding-covariant-sites">Finding Covariant Sites</h3>
<p>Now that <code>merged_stat</code> contains the variance statistics for both AF and FPS values, we can find the sites that have high variance in both AF and FPS values. We can do this by filtering the <code>merged_stat</code> table to include only the rows where both <code>AF_var</code> and <code>FPS_scaled_var</code> are above a certain threshold.</p>
<p>To remove arbitrariness, <strong>IQR outlier method</strong> can be used to find the threshold for high variance.</p>
<div id="cell-53" class="cell" data-execution_count="35">
<div class="sourceCode cell-code" id="cb37"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb37-1"><a href="#cb37-1" aria-hidden="true" tabindex="-1"></a>merged_stat</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display" data-execution_count="35">
<div>


<table class="dataframe table table-sm table-striped small" data-quarto-postprocess="true" data-border="1">
<thead>
<tr class="header">
<th data-quarto-table-cell-role="th"></th>
<th data-quarto-table-cell-role="th">sample_id</th>
<th data-quarto-table-cell-role="th">AF</th>
<th data-quarto-table-cell-role="th">FPS_scaled</th>
<th data-quarto-table-cell-role="th">AF_var</th>
<th data-quarto-table-cell-role="th">FPS_scaled_var</th>
</tr>
<tr class="odd">
<th data-quarto-table-cell-role="th">region_id</th>
<th data-quarto-table-cell-role="th"></th>
<th data-quarto-table-cell-role="th"></th>
<th data-quarto-table-cell-role="th"></th>
<th data-quarto-table-cell-role="th"></th>
<th data-quarto-table-cell-role="th"></th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td data-quarto-table-cell-role="th">chr1:818021-818032</td>
<td>2GAMBDQ_norm</td>
<td>0.875000</td>
<td>0.009147</td>
<td>0.001576</td>
<td>0.000010</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">chr1:818021-818032</td>
<td>98JKPD8_lumA</td>
<td>0.900000</td>
<td>0.007823</td>
<td>0.001576</td>
<td>0.000010</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">chr1:818021-818032</td>
<td>ANAB5F7_basal</td>
<td>0.812500</td>
<td>0.011176</td>
<td>0.001576</td>
<td>0.000010</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">chr1:818021-818032</td>
<td>PU24GB8_lumB</td>
<td>0.833333</td>
<td>0.009443</td>
<td>0.001576</td>
<td>0.000010</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">chr1:818021-818032</td>
<td>S6R691V_her2</td>
<td>0.900000</td>
<td>0.002785</td>
<td>0.001576</td>
<td>0.000010</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">...</td>
<td>...</td>
<td>...</td>
<td>...</td>
<td>...</td>
<td>...</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">chrX:155641610-155641621</td>
<td>2GAMBDQ_norm</td>
<td>0.000000</td>
<td>0.004623</td>
<td>0.000173</td>
<td>0.000001</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">chrX:155641610-155641621</td>
<td>98JKPD8_lumA</td>
<td>0.000000</td>
<td>0.003294</td>
<td>0.000173</td>
<td>0.000001</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">chrX:155641610-155641621</td>
<td>ANAB5F7_basal</td>
<td>0.000000</td>
<td>0.003158</td>
<td>0.000173</td>
<td>0.000001</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">chrX:155641610-155641621</td>
<td>PU24GB8_lumB</td>
<td>0.029412</td>
<td>0.004060</td>
<td>0.000173</td>
<td>0.000001</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">chrX:155641610-155641621</td>
<td>S6R691V_her2</td>
<td>0.000000</td>
<td>0.001737</td>
<td>0.000173</td>
<td>0.000001</td>
</tr>
</tbody>
</table>

<p>43530 rows × 5 columns</p>
</div>
</div>
</div>
<div id="cell-54" class="cell" data-execution_count="37">
<div class="sourceCode cell-code" id="cb38"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb38-1"><a href="#cb38-1" aria-hidden="true" tabindex="-1"></a><span class="co"># subset the dataframe to only include the AF_var and FPS_scaled_var columns and keep only unique rows</span></span>
<span id="cb38-2"><a href="#cb38-2" aria-hidden="true" tabindex="-1"></a>merged_stat_vars <span class="op">=</span> merged_stat[[<span class="st">'AF_var'</span>, <span class="st">'FPS_scaled_var'</span>]].copy().drop_duplicates()</span>
<span id="cb38-3"><a href="#cb38-3" aria-hidden="true" tabindex="-1"></a>merged_stat_vars</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display" data-execution_count="37">
<div>


<table class="dataframe table table-sm table-striped small" data-quarto-postprocess="true" data-border="1">
<thead>
<tr class="header">
<th data-quarto-table-cell-role="th"></th>
<th data-quarto-table-cell-role="th">AF_var</th>
<th data-quarto-table-cell-role="th">FPS_scaled_var</th>
</tr>
<tr class="odd">
<th data-quarto-table-cell-role="th">region_id</th>
<th data-quarto-table-cell-role="th"></th>
<th data-quarto-table-cell-role="th"></th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td data-quarto-table-cell-role="th">chr1:818021-818032</td>
<td>0.001576</td>
<td>0.000010</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">chr1:818780-818791</td>
<td>0.002113</td>
<td>0.000005</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">chr1:960570-960581</td>
<td>0.000347</td>
<td>0.000341</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">chr1:977015-977026</td>
<td>0.001510</td>
<td>0.013981</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">chr1:977047-977058</td>
<td>0.037434</td>
<td>0.009694</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">...</td>
<td>...</td>
<td>...</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">chrX:153642534-153642545</td>
<td>0.000781</td>
<td>0.000008</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">chrX:153934861-153934872</td>
<td>0.000781</td>
<td>0.000144</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">chrX:154444629-154444640</td>
<td>0.003363</td>
<td>0.000038</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">chrX:155010397-155010408</td>
<td>0.027746</td>
<td>0.000003</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">chrX:155641610-155641621</td>
<td>0.000173</td>
<td>0.000001</td>
</tr>
</tbody>
</table>

<p>8706 rows × 2 columns</p>
</div>
</div>
</div>
<p>We can plot boxplots to visualize the distribution of <code>AF_var</code> and <code>FPS_scaled_var</code> values.</p>
<div id="cell-56" class="cell" data-execution_count="38">
<div class="sourceCode cell-code" id="cb39"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb39-1"><a href="#cb39-1" aria-hidden="true" tabindex="-1"></a><span class="co"># plot bloxplots</span></span>
<span id="cb39-2"><a href="#cb39-2" aria-hidden="true" tabindex="-1"></a>fig, axs <span class="op">=</span> plt.subplots(<span class="dv">1</span>, <span class="dv">2</span>, figsize<span class="op">=</span>(<span class="dv">10</span>, <span class="dv">10</span>), dpi<span class="op">=</span><span class="dv">300</span>)</span>
<span id="cb39-3"><a href="#cb39-3" aria-hidden="true" tabindex="-1"></a>sns.boxplot(data<span class="op">=</span>merged_stat_vars, y<span class="op">=</span><span class="st">'AF_var'</span>, ax<span class="op">=</span>axs[<span class="dv">0</span>])</span>
<span id="cb39-4"><a href="#cb39-4" aria-hidden="true" tabindex="-1"></a>sns.boxplot(data<span class="op">=</span>merged_stat_vars, y<span class="op">=</span><span class="st">'FPS_scaled_var'</span>, ax<span class="op">=</span>axs[<span class="dv">1</span>])</span>
<span id="cb39-5"><a href="#cb39-5" aria-hidden="true" tabindex="-1"></a>plt.show()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><img src="variant-freq-analysis-v5_files/figure-html/cell-38-output-1.png" class="img-fluid figure-img"></p>
</figure>
</div>
</div>
</div>
<p>The distribution is highly skewed and we are looking at variances, not raw data points, so we can probably ignore the lower bounds and just use the upper bounds as the threshold to filter for potentially interesting sites.</p>
<p>Calculate the distribution of <code>AF_var</code> and <code>FPS_scaled_var</code> values and then calculate the IQR for each. Then, calculate the upper and lower bounds for the <code>AF_var</code> and <code>FPS_scaled_var</code> values.</p>
<div id="cell-59" class="cell" data-execution_count="39">
<div class="sourceCode cell-code" id="cb40"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb40-1"><a href="#cb40-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Calculate the IQR for AF_var and FPS_scaled_var separately</span></span>
<span id="cb40-2"><a href="#cb40-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb40-3"><a href="#cb40-3" aria-hidden="true" tabindex="-1"></a>q1_vaf <span class="op">=</span> merged_stat_vars[<span class="st">'AF_var'</span>].quantile(<span class="fl">0.25</span>)</span>
<span id="cb40-4"><a href="#cb40-4" aria-hidden="true" tabindex="-1"></a>q3_vaf <span class="op">=</span> merged_stat_vars[<span class="st">'AF_var'</span>].quantile(<span class="fl">0.75</span>)</span>
<span id="cb40-5"><a href="#cb40-5" aria-hidden="true" tabindex="-1"></a>iqr_vaf <span class="op">=</span> q3_vaf <span class="op">-</span> q1_vaf</span>
<span id="cb40-6"><a href="#cb40-6" aria-hidden="true" tabindex="-1"></a>lower_bound_outliers_vaf <span class="op">=</span> q1_vaf <span class="op">-</span> (<span class="fl">1.5</span> <span class="op">*</span> iqr_vaf)</span>
<span id="cb40-7"><a href="#cb40-7" aria-hidden="true" tabindex="-1"></a>upper_bound_outliers_vaf <span class="op">=</span> q3_vaf <span class="op">+</span> (<span class="fl">1.5</span> <span class="op">*</span> iqr_vaf)</span>
<span id="cb40-8"><a href="#cb40-8" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb40-9"><a href="#cb40-9" aria-hidden="true" tabindex="-1"></a><span class="co"># calculate IQR for FPS_scaled_var</span></span>
<span id="cb40-10"><a href="#cb40-10" aria-hidden="true" tabindex="-1"></a>q1_vfps <span class="op">=</span> merged_stat_vars[<span class="st">'FPS_scaled_var'</span>].quantile(<span class="fl">0.25</span>)</span>
<span id="cb40-11"><a href="#cb40-11" aria-hidden="true" tabindex="-1"></a>q3_vfps <span class="op">=</span> merged_stat_vars[<span class="st">'FPS_scaled_var'</span>].quantile(<span class="fl">0.75</span>)</span>
<span id="cb40-12"><a href="#cb40-12" aria-hidden="true" tabindex="-1"></a>iqr_vfps <span class="op">=</span> q3_vfps <span class="op">-</span> q1_vfps</span>
<span id="cb40-13"><a href="#cb40-13" aria-hidden="true" tabindex="-1"></a>lower_bound_outliers_vfps <span class="op">=</span> q1_vfps <span class="op">-</span> (<span class="fl">1.5</span> <span class="op">*</span> iqr_vfps)</span>
<span id="cb40-14"><a href="#cb40-14" aria-hidden="true" tabindex="-1"></a>upper_bound_outliers_vfps <span class="op">=</span> q3_vfps <span class="op">+</span> (<span class="fl">1.5</span> <span class="op">*</span> iqr_vfps)</span>
<span id="cb40-15"><a href="#cb40-15" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb40-16"><a href="#cb40-16" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(<span class="ss">f'Outlier bounds for </span><span class="sc">{</span>motif_id<span class="sc">}</span><span class="ss"> AF variance: </span><span class="sc">{</span>lower_bound_outliers_vaf<span class="sc">,</span> upper_bound_outliers_vaf<span class="sc">}</span><span class="ss">'</span>)</span>
<span id="cb40-17"><a href="#cb40-17" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb40-18"><a href="#cb40-18" aria-hidden="true" tabindex="-1"></a><span class="bu">print</span>(<span class="ss">f'Outlier bounds for </span><span class="sc">{</span>motif_id<span class="sc">}</span><span class="ss"> FPS_scaled variance: </span><span class="sc">{</span>lower_bound_outliers_vfps<span class="sc">,</span> upper_bound_outliers_vfps<span class="sc">}</span><span class="ss">'</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>Outlier bounds for TFDP1_M08108_2.00 AF variance: (-0.006431253983107125, 0.011311348045919874)
Outlier bounds for TFDP1_M08108_2.00 FPS_scaled variance: (-0.00022509123785722238, 0.0003832948356750625)</code></pre>
</div>
</div>
<div id="cell-60" class="cell" data-execution_count="40">
<div class="sourceCode cell-code" id="cb42"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb42-1"><a href="#cb42-1" aria-hidden="true" tabindex="-1"></a><span class="co"># using the outlier fence of upper bound, get the region_id that are outliers for both AF_var and FPS_scaled_var</span></span>
<span id="cb42-2"><a href="#cb42-2" aria-hidden="true" tabindex="-1"></a>outlier_af_fps_vars <span class="op">=</span> merged_stat_vars[(merged_stat_vars[<span class="st">'AF_var'</span>] <span class="op">&gt;</span> upper_bound_outliers_vaf) <span class="op">&amp;</span> (merged_stat_vars[<span class="st">'FPS_scaled_var'</span>] <span class="op">&gt;</span> upper_bound_outliers_vfps)]</span>
<span id="cb42-3"><a href="#cb42-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb42-4"><a href="#cb42-4" aria-hidden="true" tabindex="-1"></a>outlier_af_fps_vars</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display" data-execution_count="40">
<div>


<table class="dataframe table table-sm table-striped small" data-quarto-postprocess="true" data-border="1">
<thead>
<tr class="header">
<th data-quarto-table-cell-role="th"></th>
<th data-quarto-table-cell-role="th">AF_var</th>
<th data-quarto-table-cell-role="th">FPS_scaled_var</th>
</tr>
<tr class="odd">
<th data-quarto-table-cell-role="th">region_id</th>
<th data-quarto-table-cell-role="th"></th>
<th data-quarto-table-cell-role="th"></th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td data-quarto-table-cell-role="th">chr1:977047-977058</td>
<td>0.037434</td>
<td>0.009694</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">chr1:977174-977185</td>
<td>0.200000</td>
<td>0.000478</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">chr1:977206-977217</td>
<td>0.156171</td>
<td>0.000789</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">chr1:977394-977405</td>
<td>0.300000</td>
<td>0.000556</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">chr1:3857731-3857742</td>
<td>0.014514</td>
<td>0.000814</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">...</td>
<td>...</td>
<td>...</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">chr22:39520618-39520629</td>
<td>0.014366</td>
<td>0.000983</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">chr22:42369353-42369364</td>
<td>0.015027</td>
<td>0.005037</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">chr22:50474997-50475008</td>
<td>0.037910</td>
<td>0.001630</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">chrX:24054420-24054431</td>
<td>0.011948</td>
<td>0.000549</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">chrX:119606547-119606558</td>
<td>0.011720</td>
<td>0.000570</td>
</tr>
</tbody>
</table>

<p>118 rows × 2 columns</p>
</div>
</div>
</div>
<div id="cell-61" class="cell" data-execution_count="41">
<div class="sourceCode cell-code" id="cb43"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb43-1"><a href="#cb43-1" aria-hidden="true" tabindex="-1"></a><span class="co"># now extract the unique region IDs as a list from the outlier_af_fps_vars</span></span>
<span id="cb43-2"><a href="#cb43-2" aria-hidden="true" tabindex="-1"></a>outliers_list <span class="op">=</span> outlier_af_fps_vars.index.tolist()</span>
<span id="cb43-3"><a href="#cb43-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb43-4"><a href="#cb43-4" aria-hidden="true" tabindex="-1"></a><span class="co"># subset the merged_stat dataframe to get the highly covariant sites (outlier sites)</span></span>
<span id="cb43-5"><a href="#cb43-5" aria-hidden="true" tabindex="-1"></a>covar_sites <span class="op">=</span> merged_stat[merged_stat.index.isin(outliers_list)]</span>
<span id="cb43-6"><a href="#cb43-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb43-7"><a href="#cb43-7" aria-hidden="true" tabindex="-1"></a>covar_sites</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display" data-execution_count="41">
<div>


<table class="dataframe table table-sm table-striped small" data-quarto-postprocess="true" data-border="1">
<thead>
<tr class="header">
<th data-quarto-table-cell-role="th"></th>
<th data-quarto-table-cell-role="th">sample_id</th>
<th data-quarto-table-cell-role="th">AF</th>
<th data-quarto-table-cell-role="th">FPS_scaled</th>
<th data-quarto-table-cell-role="th">AF_var</th>
<th data-quarto-table-cell-role="th">FPS_scaled_var</th>
</tr>
<tr class="odd">
<th data-quarto-table-cell-role="th">region_id</th>
<th data-quarto-table-cell-role="th"></th>
<th data-quarto-table-cell-role="th"></th>
<th data-quarto-table-cell-role="th"></th>
<th data-quarto-table-cell-role="th"></th>
<th data-quarto-table-cell-role="th"></th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td data-quarto-table-cell-role="th">chr1:977047-977058</td>
<td>2GAMBDQ_norm</td>
<td>0.750000</td>
<td>0.284532</td>
<td>0.037434</td>
<td>0.009694</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">chr1:977047-977058</td>
<td>98JKPD8_lumA</td>
<td>1.000000</td>
<td>0.232565</td>
<td>0.037434</td>
<td>0.009694</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">chr1:977047-977058</td>
<td>ANAB5F7_basal</td>
<td>0.500000</td>
<td>0.104098</td>
<td>0.037434</td>
<td>0.009694</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">chr1:977047-977058</td>
<td>PU24GB8_lumB</td>
<td>0.789474</td>
<td>0.268400</td>
<td>0.037434</td>
<td>0.009694</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">chr1:977047-977058</td>
<td>S6R691V_her2</td>
<td>0.588235</td>
<td>0.069499</td>
<td>0.037434</td>
<td>0.009694</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">...</td>
<td>...</td>
<td>...</td>
<td>...</td>
<td>...</td>
<td>...</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">chrX:119606547-119606558</td>
<td>2GAMBDQ_norm</td>
<td>0.750000</td>
<td>0.094577</td>
<td>0.011720</td>
<td>0.000570</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">chrX:119606547-119606558</td>
<td>98JKPD8_lumA</td>
<td>0.666667</td>
<td>0.111742</td>
<td>0.011720</td>
<td>0.000570</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">chrX:119606547-119606558</td>
<td>ANAB5F7_basal</td>
<td>0.941176</td>
<td>0.108128</td>
<td>0.011720</td>
<td>0.000570</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">chrX:119606547-119606558</td>
<td>PU24GB8_lumB</td>
<td>0.710526</td>
<td>0.117186</td>
<td>0.011720</td>
<td>0.000570</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">chrX:119606547-119606558</td>
<td>S6R691V_her2</td>
<td>0.705882</td>
<td>0.057896</td>
<td>0.011720</td>
<td>0.000570</td>
</tr>
</tbody>
</table>

<p>590 rows × 5 columns</p>
</div>
</div>
</div>
<div id="cell-62" class="cell" data-execution_count="43">
<div class="sourceCode cell-code" id="cb44"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb44-1"><a href="#cb44-1" aria-hidden="true" tabindex="-1"></a><span class="co"># sort the outlier sites by descending order of AF_var and FPS_scaled_var</span></span>
<span id="cb44-2"><a href="#cb44-2" aria-hidden="true" tabindex="-1"></a>covar_sites_sorted <span class="op">=</span> covar_sites.sort_values(by<span class="op">=</span>[<span class="st">'AF_var'</span>, <span class="st">'FPS_scaled_var'</span>], ascending<span class="op">=</span>[<span class="va">False</span>, <span class="va">False</span>])</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>We can plot to see the distributions.</p>
<div id="cell-64" class="cell" data-execution_count="44">
<div class="sourceCode cell-code" id="cb45"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb45-1"><a href="#cb45-1" aria-hidden="true" tabindex="-1"></a><span class="co"># create color dictionary</span></span>
<span id="cb45-2"><a href="#cb45-2" aria-hidden="true" tabindex="-1"></a>dutchfield_colordict <span class="op">=</span> {<span class="st">'S6R691V_her2'</span>: <span class="st">"#e60049"</span>, <span class="st">'ANAB5F7_basal'</span>: <span class="st">"#0bb4ff"</span>, <span class="st">'98JKPD8_lumA'</span>: <span class="st">"#87bc45"</span>, <span class="st">'PU24GB8_lumB'</span>: <span class="st">"#ef9b20"</span>, <span class="st">'2GAMBDQ_norm'</span>: <span class="st">"#b33dc6"</span>}</span>
<span id="cb45-3"><a href="#cb45-3" aria-hidden="true" tabindex="-1"></a>plt.figure(figsize<span class="op">=</span>(<span class="dv">10</span>, <span class="dv">10</span>), dpi<span class="op">=</span><span class="dv">300</span>)</span>
<span id="cb45-4"><a href="#cb45-4" aria-hidden="true" tabindex="-1"></a><span class="co"># specify subplot</span></span>
<span id="cb45-5"><a href="#cb45-5" aria-hidden="true" tabindex="-1"></a>plt.subplot(<span class="dv">4</span>, <span class="dv">1</span>, <span class="dv">1</span>)</span>
<span id="cb45-6"><a href="#cb45-6" aria-hidden="true" tabindex="-1"></a>sns.boxplot(x<span class="op">=</span><span class="st">'region_id'</span>, y<span class="op">=</span><span class="st">'AF'</span>, data<span class="op">=</span>covar_sites_sorted, color<span class="op">=</span><span class="st">'whitesmoke'</span>, linecolor<span class="op">=</span><span class="st">'black'</span>, showfliers<span class="op">=</span><span class="va">False</span>)</span>
<span id="cb45-7"><a href="#cb45-7" aria-hidden="true" tabindex="-1"></a>sns.stripplot(x<span class="op">=</span><span class="st">'region_id'</span>, y<span class="op">=</span><span class="st">'AF'</span>, data<span class="op">=</span>covar_sites_sorted, hue<span class="op">=</span><span class="st">'sample_id'</span>, palette<span class="op">=</span>dutchfield_colordict, size<span class="op">=</span><span class="dv">4</span>, jitter<span class="op">=</span><span class="va">True</span>, linewidth<span class="op">=</span><span class="fl">0.5</span>, edgecolor<span class="op">=</span><span class="st">'black'</span>)</span>
<span id="cb45-8"><a href="#cb45-8" aria-hidden="true" tabindex="-1"></a>plt.xticks(rotation<span class="op">=</span><span class="dv">90</span>, fontsize<span class="op">=</span><span class="dv">4</span>)</span>
<span id="cb45-9"><a href="#cb45-9" aria-hidden="true" tabindex="-1"></a>plt.xlabel(<span class="st">''</span>)</span>
<span id="cb45-10"><a href="#cb45-10" aria-hidden="true" tabindex="-1"></a>plt.ylabel(<span class="st">'AF per site'</span>, fontsize<span class="op">=</span><span class="dv">10</span>)</span>
<span id="cb45-11"><a href="#cb45-11" aria-hidden="true" tabindex="-1"></a><span class="co"># place legend outside of the plot</span></span>
<span id="cb45-12"><a href="#cb45-12" aria-hidden="true" tabindex="-1"></a>plt.legend(bbox_to_anchor<span class="op">=</span>(<span class="fl">1.01</span>, <span class="dv">1</span>),borderaxespad<span class="op">=</span><span class="dv">0</span>, markerscale<span class="op">=</span><span class="dv">2</span>, fontsize<span class="op">=</span><span class="dv">10</span>)</span>
<span id="cb45-13"><a href="#cb45-13" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb45-14"><a href="#cb45-14" aria-hidden="true" tabindex="-1"></a>plt.subplot(<span class="dv">4</span>, <span class="dv">1</span>, <span class="dv">2</span>)</span>
<span id="cb45-15"><a href="#cb45-15" aria-hidden="true" tabindex="-1"></a>sns.barplot(x<span class="op">=</span><span class="st">'region_id'</span>, y<span class="op">=</span><span class="st">'AF_var'</span>, data<span class="op">=</span>covar_sites_sorted, color<span class="op">=</span><span class="st">'darkslateblue'</span>, edgecolor<span class="op">=</span><span class="st">'black'</span>)</span>
<span id="cb45-16"><a href="#cb45-16" aria-hidden="true" tabindex="-1"></a>plt.xticks(rotation<span class="op">=</span><span class="dv">90</span>, fontsize<span class="op">=</span><span class="dv">4</span>)</span>
<span id="cb45-17"><a href="#cb45-17" aria-hidden="true" tabindex="-1"></a>plt.xlabel(<span class="st">''</span>)</span>
<span id="cb45-18"><a href="#cb45-18" aria-hidden="true" tabindex="-1"></a>plt.ylabel(<span class="st">'AF variance'</span>, fontsize<span class="op">=</span><span class="dv">10</span>)</span>
<span id="cb45-19"><a href="#cb45-19" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb45-20"><a href="#cb45-20" aria-hidden="true" tabindex="-1"></a>plt.subplot(<span class="dv">4</span>, <span class="dv">1</span>, <span class="dv">3</span>)</span>
<span id="cb45-21"><a href="#cb45-21" aria-hidden="true" tabindex="-1"></a>sns.boxplot(x<span class="op">=</span><span class="st">'region_id'</span>, y<span class="op">=</span><span class="st">'FPS_scaled'</span>, data<span class="op">=</span>covar_sites_sorted, color<span class="op">=</span><span class="st">'whitesmoke'</span>, linecolor<span class="op">=</span><span class="st">'black'</span>, showfliers<span class="op">=</span><span class="va">False</span>)</span>
<span id="cb45-22"><a href="#cb45-22" aria-hidden="true" tabindex="-1"></a>sns.stripplot(x<span class="op">=</span><span class="st">'region_id'</span>, y<span class="op">=</span><span class="st">'FPS_scaled'</span>, data<span class="op">=</span>covar_sites_sorted, hue<span class="op">=</span><span class="st">'sample_id'</span>, palette<span class="op">=</span>dutchfield_colordict, size<span class="op">=</span><span class="dv">4</span>, jitter<span class="op">=</span><span class="va">True</span>, legend<span class="op">=</span><span class="va">False</span>, linewidth<span class="op">=</span><span class="fl">0.5</span>, edgecolor<span class="op">=</span><span class="st">'black'</span>)</span>
<span id="cb45-23"><a href="#cb45-23" aria-hidden="true" tabindex="-1"></a>plt.xticks(rotation<span class="op">=</span><span class="dv">90</span>, fontsize<span class="op">=</span><span class="dv">4</span>)</span>
<span id="cb45-24"><a href="#cb45-24" aria-hidden="true" tabindex="-1"></a>plt.xlabel(<span class="st">''</span>)</span>
<span id="cb45-25"><a href="#cb45-25" aria-hidden="true" tabindex="-1"></a>plt.ylabel(<span class="st">'Scaled FPS per site'</span>, fontsize<span class="op">=</span><span class="dv">10</span>)</span>
<span id="cb45-26"><a href="#cb45-26" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb45-27"><a href="#cb45-27" aria-hidden="true" tabindex="-1"></a>plt.subplot(<span class="dv">4</span>, <span class="dv">1</span>, <span class="dv">4</span>)</span>
<span id="cb45-28"><a href="#cb45-28" aria-hidden="true" tabindex="-1"></a>sns.barplot(x<span class="op">=</span><span class="st">'region_id'</span>, y<span class="op">=</span><span class="st">'FPS_scaled_var'</span>, data<span class="op">=</span>covar_sites_sorted, color<span class="op">=</span><span class="st">'darkslateblue'</span>, edgecolor<span class="op">=</span><span class="st">'black'</span>)</span>
<span id="cb45-29"><a href="#cb45-29" aria-hidden="true" tabindex="-1"></a>plt.xticks(rotation<span class="op">=</span><span class="dv">90</span>, fontsize<span class="op">=</span><span class="dv">4</span>)</span>
<span id="cb45-30"><a href="#cb45-30" aria-hidden="true" tabindex="-1"></a>plt.xlabel(<span class="ss">f'</span><span class="sc">{</span>motif_id<span class="sc">}</span><span class="ss"> TF binding sites with which are outlying covariant in AF and FPS'</span>, fontsize<span class="op">=</span><span class="dv">10</span>)</span>
<span id="cb45-31"><a href="#cb45-31" aria-hidden="true" tabindex="-1"></a>plt.ylabel(<span class="st">'FPS Variance'</span>, fontsize<span class="op">=</span><span class="dv">10</span>)</span>
<span id="cb45-32"><a href="#cb45-32" aria-hidden="true" tabindex="-1"></a>plt.subplots_adjust(hspace<span class="op">=</span><span class="fl">0.8</span>)</span>
<span id="cb45-33"><a href="#cb45-33" aria-hidden="true" tabindex="-1"></a>plt.tight_layout()</span>
<span id="cb45-34"><a href="#cb45-34" aria-hidden="true" tabindex="-1"></a>plt.show()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><img src="variant-freq-analysis-v5_files/figure-html/cell-43-output-1.png" class="img-fluid figure-img"></p>
</figure>
</div>
</div>
</div>
<p>We can plot <code>lmplot</code> hued by <code>sample_id</code> to see if there are any patterns in the raw data.</p>
<div id="cell-66" class="cell" data-execution_count="45">
<div class="sourceCode cell-code" id="cb46"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb46-1"><a href="#cb46-1" aria-hidden="true" tabindex="-1"></a><span class="co"># plot jointplot of site_counts_sorted_top5</span></span>
<span id="cb46-2"><a href="#cb46-2" aria-hidden="true" tabindex="-1"></a>sns.lmplot(x<span class="op">=</span><span class="st">'AF'</span>, y<span class="op">=</span><span class="st">'FPS_scaled'</span>, data<span class="op">=</span>covar_sites_sorted, hue<span class="op">=</span><span class="st">'sample_id'</span>, fit_reg<span class="op">=</span><span class="va">True</span>, height<span class="op">=</span><span class="dv">10</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><img src="variant-freq-analysis-v5_files/figure-html/cell-44-output-1.png" class="img-fluid figure-img"></p>
</figure>
</div>
</div>
</div>
<div id="cell-67" class="cell" data-execution_count="47">
<div class="sourceCode cell-code" id="cb47"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb47-1"><a href="#cb47-1" aria-hidden="true" tabindex="-1"></a><span class="co"># test for correlation between AF and FPS_scaled for each subtype</span></span>
<span id="cb47-2"><a href="#cb47-2" aria-hidden="true" tabindex="-1"></a><span class="im">from</span> scipy.stats <span class="im">import</span> spearmanr</span>
<span id="cb47-3"><a href="#cb47-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb47-4"><a href="#cb47-4" aria-hidden="true" tabindex="-1"></a><span class="co"># # Group by 'sample_id' and calculate the correlation for each group</span></span>
<span id="cb47-5"><a href="#cb47-5" aria-hidden="true" tabindex="-1"></a><span class="co"># correlations = covar_sites_sorted.groupby('sample_id').apply(lambda group: spearmanr(group['AF'], group['FPS_scaled']), include_groups=False)</span></span>
<span id="cb47-6"><a href="#cb47-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb47-7"><a href="#cb47-7" aria-hidden="true" tabindex="-1"></a><span class="co"># correlations</span></span>
<span id="cb47-8"><a href="#cb47-8" aria-hidden="true" tabindex="-1"></a></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>How about we group by region_id?</p>
<div id="cell-69" class="cell" data-execution_count="48">
<div class="sourceCode cell-code" id="cb48"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb48-1"><a href="#cb48-1" aria-hidden="true" tabindex="-1"></a>covar_sites_sorted_novars <span class="op">=</span> covar_sites_sorted.drop(columns<span class="op">=</span>[<span class="st">'AF_var'</span>, <span class="st">'FPS_scaled_var'</span>])</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div id="cell-70" class="cell" data-execution_count="49">
<div class="sourceCode cell-code" id="cb49"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb49-1"><a href="#cb49-1" aria-hidden="true" tabindex="-1"></a>covar_sites_sorted_novars</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display" data-execution_count="49">
<div>


<table class="dataframe table table-sm table-striped small" data-quarto-postprocess="true" data-border="1">
<thead>
<tr class="header">
<th data-quarto-table-cell-role="th"></th>
<th data-quarto-table-cell-role="th">sample_id</th>
<th data-quarto-table-cell-role="th">AF</th>
<th data-quarto-table-cell-role="th">FPS_scaled</th>
</tr>
<tr class="odd">
<th data-quarto-table-cell-role="th">region_id</th>
<th data-quarto-table-cell-role="th"></th>
<th data-quarto-table-cell-role="th"></th>
<th data-quarto-table-cell-role="th"></th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td data-quarto-table-cell-role="th">chr1:977394-977405</td>
<td>2GAMBDQ_norm</td>
<td>1.000000</td>
<td>0.064366</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">chr1:977394-977405</td>
<td>98JKPD8_lumA</td>
<td>0.000000</td>
<td>0.059305</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">chr1:977394-977405</td>
<td>ANAB5F7_basal</td>
<td>0.000000</td>
<td>0.024563</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">chr1:977394-977405</td>
<td>PU24GB8_lumB</td>
<td>1.000000</td>
<td>0.066256</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">chr1:977394-977405</td>
<td>S6R691V_her2</td>
<td>0.000000</td>
<td>0.016988</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">...</td>
<td>...</td>
<td>...</td>
<td>...</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">chr3:161854658-161854669</td>
<td>2GAMBDQ_norm</td>
<td>0.562500</td>
<td>0.018679</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">chr3:161854658-161854669</td>
<td>98JKPD8_lumA</td>
<td>0.625000</td>
<td>0.008329</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">chr3:161854658-161854669</td>
<td>ANAB5F7_basal</td>
<td>0.441176</td>
<td>0.064638</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">chr3:161854658-161854669</td>
<td>PU24GB8_lumB</td>
<td>0.736842</td>
<td>0.012126</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">chr3:161854658-161854669</td>
<td>S6R691V_her2</td>
<td>0.588235</td>
<td>0.004887</td>
</tr>
</tbody>
</table>

<p>590 rows × 3 columns</p>
</div>
</div>
</div>
<div id="cell-71" class="cell" data-execution_count="50">
<div class="sourceCode cell-code" id="cb50"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb50-1"><a href="#cb50-1" aria-hidden="true" tabindex="-1"></a><span class="co"># reset index to make region_id a column</span></span>
<span id="cb50-2"><a href="#cb50-2" aria-hidden="true" tabindex="-1"></a>covar_sites_sorted_novars <span class="op">=</span> covar_sites_sorted_novars.reset_index()</span>
<span id="cb50-3"><a href="#cb50-3" aria-hidden="true" tabindex="-1"></a>covar_sites_sorted_novars</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display" data-execution_count="50">
<div>


<table class="dataframe table table-sm table-striped small" data-quarto-postprocess="true" data-border="1">
<thead>
<tr class="header">
<th data-quarto-table-cell-role="th"></th>
<th data-quarto-table-cell-role="th">region_id</th>
<th data-quarto-table-cell-role="th">sample_id</th>
<th data-quarto-table-cell-role="th">AF</th>
<th data-quarto-table-cell-role="th">FPS_scaled</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td data-quarto-table-cell-role="th">0</td>
<td>chr1:977394-977405</td>
<td>2GAMBDQ_norm</td>
<td>1.000000</td>
<td>0.064366</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">1</td>
<td>chr1:977394-977405</td>
<td>98JKPD8_lumA</td>
<td>0.000000</td>
<td>0.059305</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">2</td>
<td>chr1:977394-977405</td>
<td>ANAB5F7_basal</td>
<td>0.000000</td>
<td>0.024563</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">3</td>
<td>chr1:977394-977405</td>
<td>PU24GB8_lumB</td>
<td>1.000000</td>
<td>0.066256</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">4</td>
<td>chr1:977394-977405</td>
<td>S6R691V_her2</td>
<td>0.000000</td>
<td>0.016988</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">...</td>
<td>...</td>
<td>...</td>
<td>...</td>
<td>...</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">585</td>
<td>chr3:161854658-161854669</td>
<td>2GAMBDQ_norm</td>
<td>0.562500</td>
<td>0.018679</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">586</td>
<td>chr3:161854658-161854669</td>
<td>98JKPD8_lumA</td>
<td>0.625000</td>
<td>0.008329</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">587</td>
<td>chr3:161854658-161854669</td>
<td>ANAB5F7_basal</td>
<td>0.441176</td>
<td>0.064638</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">588</td>
<td>chr3:161854658-161854669</td>
<td>PU24GB8_lumB</td>
<td>0.736842</td>
<td>0.012126</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">589</td>
<td>chr3:161854658-161854669</td>
<td>S6R691V_her2</td>
<td>0.588235</td>
<td>0.004887</td>
</tr>
</tbody>
</table>

<p>590 rows × 4 columns</p>
</div>
</div>
</div>
<div id="cell-72" class="cell" data-execution_count="52">
<div class="sourceCode cell-code" id="cb51"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb51-1"><a href="#cb51-1" aria-hidden="true" tabindex="-1"></a><span class="co"># get the rows where the region_id == chr10:24251881-24251892</span></span>
<span id="cb51-2"><a href="#cb51-2" aria-hidden="true" tabindex="-1"></a>covar_sites_sorted_novars[covar_sites_sorted_novars[<span class="st">'region_id'</span>] <span class="op">==</span> <span class="st">'chr10:73997814-73997825'</span>]</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display" data-execution_count="52">
<div>


<table class="dataframe table table-sm table-striped small" data-quarto-postprocess="true" data-border="1">
<thead>
<tr class="header">
<th data-quarto-table-cell-role="th"></th>
<th data-quarto-table-cell-role="th">region_id</th>
<th data-quarto-table-cell-role="th">sample_id</th>
<th data-quarto-table-cell-role="th">AF</th>
<th data-quarto-table-cell-role="th">FPS_scaled</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td data-quarto-table-cell-role="th">320</td>
<td>chr10:73997814-73997825</td>
<td>2GAMBDQ_norm</td>
<td>0.750000</td>
<td>0.391551</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">321</td>
<td>chr10:73997814-73997825</td>
<td>98JKPD8_lumA</td>
<td>0.875000</td>
<td>0.408824</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">322</td>
<td>chr10:73997814-73997825</td>
<td>ANAB5F7_basal</td>
<td>0.676471</td>
<td>0.336919</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">323</td>
<td>chr10:73997814-73997825</td>
<td>PU24GB8_lumB</td>
<td>0.657895</td>
<td>0.371692</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">324</td>
<td>chr10:73997814-73997825</td>
<td>S6R691V_her2</td>
<td>0.529412</td>
<td>0.210583</td>
</tr>
</tbody>
</table>

</div>
</div>
</div>
<section id="spearman-correlation-test-between-af-and-fps_scaled" class="level4">
<h4 class="anchored" data-anchor-id="spearman-correlation-test-between-af-and-fps_scaled">Spearman Correlation Test between AF and FPS_scaled</h4>
<div id="cell-74" class="cell" data-execution_count="55">
<div class="sourceCode cell-code" id="cb52"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb52-1"><a href="#cb52-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Group by 'region_id' and calculate the correlation between AF and FPS_scaled for each group</span></span>
<span id="cb52-2"><a href="#cb52-2" aria-hidden="true" tabindex="-1"></a>correlations <span class="op">=</span> covar_sites_sorted_novars.groupby(<span class="st">'region_id'</span>).<span class="bu">apply</span>(<span class="kw">lambda</span> group: spearmanr(group[<span class="st">'AF'</span>], group[<span class="st">'FPS_scaled'</span>]), include_groups<span class="op">=</span><span class="va">False</span>)</span>
<span id="cb52-3"><a href="#cb52-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb52-4"><a href="#cb52-4" aria-hidden="true" tabindex="-1"></a>correlations</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display" data-execution_count="55">
<pre><code>region_id
chr10:15168778-15168789        (0.20519567041703085, 0.7405819415910722)
chr10:24251881-24251892        (0.09999999999999999, 0.8728885715695383)
chr10:73997814-73997825        (0.8999999999999998, 0.03738607346849875)
chr11:119121466-119121477                     (0.7, 0.18812040437418728)
chr11:560866-560877          (-0.49999999999999994, 0.39100221895577053)
                                                ...                     
chr9:88256422-88256433                        (-0.3, 0.6238376647810728)
chr9:92098292-92098303         (0.09999999999999999, 0.8728885715695383)
chr9:93095668-93095679                        (0.7, 0.18812040437418728)
chrX:119606547-119606558      (-0.09999999999999999, 0.8728885715695383)
chrX:24054420-24054431                        (0.6, 0.28475697986529375)
Length: 118, dtype: object</code></pre>
</div>
</div>
<div id="cell-75" class="cell" data-execution_count="56">
<div class="sourceCode cell-code" id="cb54"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb54-1"><a href="#cb54-1" aria-hidden="true" tabindex="-1"></a><span class="co"># that returned a pandas Series, so convert to dataframe</span></span>
<span id="cb54-2"><a href="#cb54-2" aria-hidden="true" tabindex="-1"></a>correlations_df <span class="op">=</span> pd.DataFrame(correlations, columns<span class="op">=</span>[<span class="st">'corr_coeff_and_pvalue'</span>])</span>
<span id="cb54-3"><a href="#cb54-3" aria-hidden="true" tabindex="-1"></a><span class="co"># split the 'corr_coeff_and_pvalue' column into two separate columns using apply() and pd.Series</span></span>
<span id="cb54-4"><a href="#cb54-4" aria-hidden="true" tabindex="-1"></a>correlations_df[[<span class="st">'corr_coeff'</span>, <span class="st">'pvalue'</span>]] <span class="op">=</span> correlations_df[<span class="st">'corr_coeff_and_pvalue'</span>].<span class="bu">apply</span>(pd.Series)</span>
<span id="cb54-5"><a href="#cb54-5" aria-hidden="true" tabindex="-1"></a><span class="co"># drop the original column</span></span>
<span id="cb54-6"><a href="#cb54-6" aria-hidden="true" tabindex="-1"></a>correlations_df <span class="op">=</span> correlations_df.drop(columns<span class="op">=</span>[<span class="st">'corr_coeff_and_pvalue'</span>]).reset_index()</span>
<span id="cb54-7"><a href="#cb54-7" aria-hidden="true" tabindex="-1"></a>correlations_df</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display" data-execution_count="56">
<div>


<table class="dataframe table table-sm table-striped small" data-quarto-postprocess="true" data-border="1">
<thead>
<tr class="header">
<th data-quarto-table-cell-role="th"></th>
<th data-quarto-table-cell-role="th">region_id</th>
<th data-quarto-table-cell-role="th">corr_coeff</th>
<th data-quarto-table-cell-role="th">pvalue</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td data-quarto-table-cell-role="th">0</td>
<td>chr10:15168778-15168789</td>
<td>0.205196</td>
<td>0.740582</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">1</td>
<td>chr10:24251881-24251892</td>
<td>0.100000</td>
<td>0.872889</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">2</td>
<td>chr10:73997814-73997825</td>
<td>0.900000</td>
<td>0.037386</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">3</td>
<td>chr11:119121466-119121477</td>
<td>0.700000</td>
<td>0.188120</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">4</td>
<td>chr11:560866-560877</td>
<td>-0.500000</td>
<td>0.391002</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">...</td>
<td>...</td>
<td>...</td>
<td>...</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">113</td>
<td>chr9:88256422-88256433</td>
<td>-0.300000</td>
<td>0.623838</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">114</td>
<td>chr9:92098292-92098303</td>
<td>0.100000</td>
<td>0.872889</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">115</td>
<td>chr9:93095668-93095679</td>
<td>0.700000</td>
<td>0.188120</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">116</td>
<td>chrX:119606547-119606558</td>
<td>-0.100000</td>
<td>0.872889</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">117</td>
<td>chrX:24054420-24054431</td>
<td>0.600000</td>
<td>0.284757</td>
</tr>
</tbody>
</table>

<p>118 rows × 3 columns</p>
</div>
</div>
</div>
<div id="cell-76" class="cell" data-execution_count="57">
<div class="sourceCode cell-code" id="cb55"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb55-1"><a href="#cb55-1" aria-hidden="true" tabindex="-1"></a><span class="co"># now sort region_id naturally</span></span>
<span id="cb55-2"><a href="#cb55-2" aria-hidden="true" tabindex="-1"></a>correlations_df_srt <span class="op">=</span> correlations_df.reindex(index<span class="op">=</span>index_natsorted(correlations_df[<span class="st">'region_id'</span>]))</span>
<span id="cb55-3"><a href="#cb55-3" aria-hidden="true" tabindex="-1"></a>corr_df <span class="op">=</span> correlations_df_srt.reset_index(drop<span class="op">=</span><span class="va">True</span>)</span>
<span id="cb55-4"><a href="#cb55-4" aria-hidden="true" tabindex="-1"></a>corr_df</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display" data-execution_count="57">
<div>


<table class="dataframe table table-sm table-striped small" data-quarto-postprocess="true" data-border="1">
<thead>
<tr class="header">
<th data-quarto-table-cell-role="th"></th>
<th data-quarto-table-cell-role="th">region_id</th>
<th data-quarto-table-cell-role="th">corr_coeff</th>
<th data-quarto-table-cell-role="th">pvalue</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td data-quarto-table-cell-role="th">0</td>
<td>chr1:977047-977058</td>
<td>0.500000</td>
<td>0.391002</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">1</td>
<td>chr1:977174-977185</td>
<td>0.000000</td>
<td>1.000000</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">2</td>
<td>chr1:977206-977217</td>
<td>0.900000</td>
<td>0.037386</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">3</td>
<td>chr1:977394-977405</td>
<td>0.866025</td>
<td>0.057669</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">4</td>
<td>chr1:3857731-3857742</td>
<td>-0.300000</td>
<td>0.623838</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">...</td>
<td>...</td>
<td>...</td>
<td>...</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">113</td>
<td>chr22:39520618-39520629</td>
<td>0.700000</td>
<td>0.188120</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">114</td>
<td>chr22:42369353-42369364</td>
<td>0.500000</td>
<td>0.391002</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">115</td>
<td>chr22:50474997-50475008</td>
<td>-0.200000</td>
<td>0.747060</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">116</td>
<td>chrX:24054420-24054431</td>
<td>0.600000</td>
<td>0.284757</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">117</td>
<td>chrX:119606547-119606558</td>
<td>-0.100000</td>
<td>0.872889</td>
</tr>
</tbody>
</table>

<p>118 rows × 3 columns</p>
</div>
</div>
</div>
<div id="cell-77" class="cell" data-execution_count="58">
<div class="sourceCode cell-code" id="cb56"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb56-1"><a href="#cb56-1" aria-hidden="true" tabindex="-1"></a><span class="co"># filter for pvalue &lt; 0.05</span></span>
<span id="cb56-2"><a href="#cb56-2" aria-hidden="true" tabindex="-1"></a>corr_df_sig <span class="op">=</span> corr_df[corr_df[<span class="st">'pvalue'</span>] <span class="op">&lt;</span> <span class="fl">0.05</span>]</span>
<span id="cb56-3"><a href="#cb56-3" aria-hidden="true" tabindex="-1"></a>corr_df_sig</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display" data-execution_count="58">
<div>


<table class="dataframe table table-sm table-striped small" data-quarto-postprocess="true" data-border="1">
<thead>
<tr class="header">
<th data-quarto-table-cell-role="th"></th>
<th data-quarto-table-cell-role="th">region_id</th>
<th data-quarto-table-cell-role="th">corr_coeff</th>
<th data-quarto-table-cell-role="th">pvalue</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td data-quarto-table-cell-role="th">2</td>
<td>chr1:977206-977217</td>
<td>0.9</td>
<td>3.738607e-02</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">65</td>
<td>chr10:73997814-73997825</td>
<td>0.9</td>
<td>3.738607e-02</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">75</td>
<td>chr12:120291913-120291924</td>
<td>-0.9</td>
<td>3.738607e-02</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">77</td>
<td>chr13:98484867-98484878</td>
<td>1.0</td>
<td>1.404265e-24</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">88</td>
<td>chr16:382034-382045</td>
<td>0.9</td>
<td>3.738607e-02</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">95</td>
<td>chr17:18682506-18682517</td>
<td>-0.9</td>
<td>3.738607e-02</td>
</tr>
</tbody>
</table>

</div>
</div>
</div>
</section>
<section id="multiple-testing-correction" class="level4">
<h4 class="anchored" data-anchor-id="multiple-testing-correction">Multiple Testing Correction</h4>
<p>Now we need to correct for multiple testing. We can use the Benjamini-Hochberg method to correct for multiple testing.</p>
<div id="cell-79" class="cell">
<div class="sourceCode cell-code" id="cb57"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb57-1"><a href="#cb57-1" aria-hidden="true" tabindex="-1"></a><span class="im">from</span> statsmodels.stats.multitest <span class="im">import</span> multipletests</span>
<span id="cb57-2"><a href="#cb57-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb57-3"><a href="#cb57-3" aria-hidden="true" tabindex="-1"></a><span class="co"># apply Benjamini/Hochberg FDR correction</span></span>
<span id="cb57-4"><a href="#cb57-4" aria-hidden="true" tabindex="-1"></a>pvals <span class="op">=</span> corr_df[<span class="st">'pvalue'</span>]</span>
<span id="cb57-5"><a href="#cb57-5" aria-hidden="true" tabindex="-1"></a>fdr_corr <span class="op">=</span> multipletests(pvals, alpha<span class="op">=</span><span class="fl">0.05</span>, method<span class="op">=</span><span class="st">'fdr_bh'</span>)</span>
<span id="cb57-6"><a href="#cb57-6" aria-hidden="true" tabindex="-1"></a>fdr_corr</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div id="cell-80" class="cell" data-execution_count="60">
<div class="sourceCode cell-code" id="cb58"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb58-1"><a href="#cb58-1" aria-hidden="true" tabindex="-1"></a><span class="co"># add the corrected pvalues to the dataframe</span></span>
<span id="cb58-2"><a href="#cb58-2" aria-hidden="true" tabindex="-1"></a>corr_df_corrected <span class="op">=</span> corr_df.copy()</span>
<span id="cb58-3"><a href="#cb58-3" aria-hidden="true" tabindex="-1"></a>corr_df_corrected[<span class="st">'fdr_adj_pvalue'</span>] <span class="op">=</span> fdr_corr[<span class="dv">1</span>]</span>
<span id="cb58-4"><a href="#cb58-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb58-5"><a href="#cb58-5" aria-hidden="true" tabindex="-1"></a><span class="co"># filter for corrected pvalue &lt; 0.05</span></span>
<span id="cb58-6"><a href="#cb58-6" aria-hidden="true" tabindex="-1"></a>corr_df_corrected_sig <span class="op">=</span> corr_df_corrected[corr_df_corrected[<span class="st">'fdr_adj_pvalue'</span>] <span class="op">&lt;</span> <span class="fl">0.05</span>]</span>
<span id="cb58-7"><a href="#cb58-7" aria-hidden="true" tabindex="-1"></a>corr_df_corrected_sig</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display" data-execution_count="60">
<div>


<table class="dataframe table table-sm table-striped small" data-quarto-postprocess="true" data-border="1">
<thead>
<tr class="header">
<th data-quarto-table-cell-role="th"></th>
<th data-quarto-table-cell-role="th">region_id</th>
<th data-quarto-table-cell-role="th">corr_coeff</th>
<th data-quarto-table-cell-role="th">pvalue</th>
<th data-quarto-table-cell-role="th">fdr_adj_pvalue</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td data-quarto-table-cell-role="th">77</td>
<td>chr13:98484867-98484878</td>
<td>1.0</td>
<td>1.404265e-24</td>
<td>1.657033e-22</td>
</tr>
</tbody>
</table>

</div>
</div>
</div>
</section>
<section id="aside-visualization-of-correlation-coefficients-at-filtered-covariant-sites" class="level4">
<h4 class="anchored" data-anchor-id="aside-visualization-of-correlation-coefficients-at-filtered-covariant-sites">ASIDE: Visualization of Correlation Coefficients at Filtered Covariant Sites</h4>
<div id="cell-82" class="cell" data-execution_count="147">
<div class="sourceCode cell-code" id="cb59"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb59-1"><a href="#cb59-1" aria-hidden="true" tabindex="-1"></a>corr_df_i <span class="op">=</span> corr_df.set_index(<span class="st">'region_id'</span>)</span>
<span id="cb59-2"><a href="#cb59-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb59-3"><a href="#cb59-3" aria-hidden="true" tabindex="-1"></a><span class="co"># Convert the DataFrame to a 2D format suitable for a heatmap</span></span>
<span id="cb59-4"><a href="#cb59-4" aria-hidden="true" tabindex="-1"></a>df_2d <span class="op">=</span> corr_df_i[<span class="st">'corr_coeff'</span>].to_numpy().reshape(<span class="op">-</span><span class="dv">1</span>, <span class="dv">1</span>)</span>
<span id="cb59-5"><a href="#cb59-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb59-6"><a href="#cb59-6" aria-hidden="true" tabindex="-1"></a><span class="co"># Create a heatmap</span></span>
<span id="cb59-7"><a href="#cb59-7" aria-hidden="true" tabindex="-1"></a>plt.figure(figsize<span class="op">=</span>(<span class="fl">0.5</span>,<span class="dv">50</span>))  <span class="co"># Adjust the figure size as needed</span></span>
<span id="cb59-8"><a href="#cb59-8" aria-hidden="true" tabindex="-1"></a>sns.heatmap(df_2d, annot<span class="op">=</span><span class="va">True</span>, cmap<span class="op">=</span><span class="st">'coolwarm'</span>, xticklabels<span class="op">=</span><span class="va">False</span>, yticklabels<span class="op">=</span>corr_df_i.index)</span>
<span id="cb59-9"><a href="#cb59-9" aria-hidden="true" tabindex="-1"></a>plt.show()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><img src="variant-freq-analysis-v5_files/figure-html/cell-56-output-1.png" class="img-fluid figure-img"></p>
</figure>
</div>
</div>
</div>
<div id="cell-83" class="cell" data-execution_count="9">
<div class="sourceCode cell-code" id="cb60"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb60-1"><a href="#cb60-1" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> os</span>
<span id="cb60-2"><a href="#cb60-2" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> csv</span>
<span id="cb60-3"><a href="#cb60-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb60-4"><a href="#cb60-4" aria-hidden="true" tabindex="-1"></a><span class="kw">def</span> extract_motif_names(path, extension, output_file):</span>
<span id="cb60-5"><a href="#cb60-5" aria-hidden="true" tabindex="-1"></a>    data <span class="op">=</span> []</span>
<span id="cb60-6"><a href="#cb60-6" aria-hidden="true" tabindex="-1"></a>    <span class="cf">for</span> root, _, files <span class="kw">in</span> os.walk(path):</span>
<span id="cb60-7"><a href="#cb60-7" aria-hidden="true" tabindex="-1"></a>        <span class="cf">for</span> <span class="bu">file</span> <span class="kw">in</span> files:</span>
<span id="cb60-8"><a href="#cb60-8" aria-hidden="true" tabindex="-1"></a>            <span class="cf">if</span> <span class="bu">file</span>.endswith(extension):</span>
<span id="cb60-9"><a href="#cb60-9" aria-hidden="true" tabindex="-1"></a>                <span class="cf">with</span> <span class="bu">open</span>(os.path.join(root, <span class="bu">file</span>)) <span class="im">as</span> f:</span>
<span id="cb60-10"><a href="#cb60-10" aria-hidden="true" tabindex="-1"></a>                    first_line <span class="op">=</span> f.readline()</span>
<span id="cb60-11"><a href="#cb60-11" aria-hidden="true" tabindex="-1"></a>                    split_line <span class="op">=</span> first_line.split()</span>
<span id="cb60-12"><a href="#cb60-12" aria-hidden="true" tabindex="-1"></a>                    <span class="cf">if</span> <span class="bu">len</span>(split_line) <span class="op">&gt;</span> <span class="dv">1</span>:</span>
<span id="cb60-13"><a href="#cb60-13" aria-hidden="true" tabindex="-1"></a>                        tf_tuple <span class="op">=</span> (split_line[<span class="dv">0</span>][<span class="dv">1</span>:], split_line[<span class="dv">1</span>])</span>
<span id="cb60-14"><a href="#cb60-14" aria-hidden="true" tabindex="-1"></a>                        data.append(tf_tuple)</span>
<span id="cb60-15"><a href="#cb60-15" aria-hidden="true" tabindex="-1"></a>    </span>
<span id="cb60-16"><a href="#cb60-16" aria-hidden="true" tabindex="-1"></a>    <span class="co"># sort the list of tuples by the first element, then second element</span></span>
<span id="cb60-17"><a href="#cb60-17" aria-hidden="true" tabindex="-1"></a>    data.sort(key<span class="op">=</span><span class="kw">lambda</span> x: (x[<span class="dv">1</span>], x[<span class="dv">0</span>]))</span>
<span id="cb60-18"><a href="#cb60-18" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb60-19"><a href="#cb60-19" aria-hidden="true" tabindex="-1"></a>    <span class="co"># write the sorted list of tuples to a file</span></span>
<span id="cb60-20"><a href="#cb60-20" aria-hidden="true" tabindex="-1"></a>    <span class="cf">with</span> <span class="bu">open</span>(output_file, <span class="st">'w'</span>, newline<span class="op">=</span><span class="st">''</span>) <span class="im">as</span> out_file:</span>
<span id="cb60-21"><a href="#cb60-21" aria-hidden="true" tabindex="-1"></a>        writer <span class="op">=</span> csv.writer(out_file, delimiter<span class="op">=</span><span class="st">'</span><span class="ch">\t</span><span class="st">'</span>)</span>
<span id="cb60-22"><a href="#cb60-22" aria-hidden="true" tabindex="-1"></a>        writer.writerows(data)</span>
<span id="cb60-23"><a href="#cb60-23" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb60-24"><a href="#cb60-24" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb60-25"><a href="#cb60-25" aria-hidden="true" tabindex="-1"></a>path <span class="op">=</span> <span class="st">'/Users/sufyazi/Documents/local-storage/bioinf/repos/reference-TF-motifs/combined-filt-individual-motifs/individuals'</span></span>
<span id="cb60-26"><a href="#cb60-26" aria-hidden="true" tabindex="-1"></a>extension <span class="op">=</span> <span class="st">'.jaspar'</span></span>
<span id="cb60-27"><a href="#cb60-27" aria-hidden="true" tabindex="-1"></a>output_file <span class="op">=</span> <span class="st">'/Users/sufyazi/Documents/local-storage/bioinf/repos/reference-TF-motifs/combined-filt-individual-motifs/tf-motif-IDs-and-tf-names.tsv'</span></span>
<span id="cb60-28"><a href="#cb60-28" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb60-29"><a href="#cb60-29" aria-hidden="true" tabindex="-1"></a>extract_motif_names(path, extension, output_file)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
</section>
</section>
<section id="annotating-covariant-sites" class="level2">
<h2 class="anchored" data-anchor-id="annotating-covariant-sites">Annotating Covariant Sites</h2>
<p>First extract TF IDs and TF classes from a manually curated TSV.</p>
<div id="cell-86" class="cell" data-execution_count="9">
<div class="sourceCode cell-code" id="cb61"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb61-1"><a href="#cb61-1" aria-hidden="true" tabindex="-1"></a><span class="co"># first load the data table</span></span>
<span id="cb61-2"><a href="#cb61-2" aria-hidden="true" tabindex="-1"></a>annot_data <span class="op">=</span> pd.read_csv(<span class="st">'/Users/sufyazi/Documents/local-storage/bioinf/repos/reference-TF-motifs/combined-filt-individual-motifs/TF-motif-full-metadata-manual.tsv'</span>, sep<span class="op">=</span><span class="st">'</span><span class="ch">\t</span><span class="st">'</span>, header<span class="op">=</span><span class="dv">0</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div id="cell-87" class="cell" data-execution_count="10">
<div class="sourceCode cell-code" id="cb62"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb62-1"><a href="#cb62-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Replace NaN values with 'N/A'</span></span>
<span id="cb62-2"><a href="#cb62-2" aria-hidden="true" tabindex="-1"></a>annot_data.fillna(<span class="st">'N/A'</span>, inplace<span class="op">=</span><span class="va">True</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div id="cell-88" class="cell" data-execution_count="11">
<div class="sourceCode cell-code" id="cb63"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb63-1"><a href="#cb63-1" aria-hidden="true" tabindex="-1"></a>annot_data[annot_data[<span class="st">'tf_name'</span>] <span class="op">==</span> <span class="st">'STAT1STAT2'</span>]</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display" data-execution_count="11">
<div>


<table class="dataframe table table-sm table-striped small" data-quarto-postprocess="true" data-border="1">
<thead>
<tr class="header">
<th data-quarto-table-cell-role="th"></th>
<th data-quarto-table-cell-role="th">motif_id</th>
<th data-quarto-table-cell-role="th">tf_name</th>
<th data-quarto-table-cell-role="th">tf_class</th>
<th data-quarto-table-cell-role="th">tf_family</th>
<th data-quarto-table-cell-role="th">tf_subfamily</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td data-quarto-table-cell-role="th">771</td>
<td>MA0517.1</td>
<td>STAT1STAT2</td>
<td>STAT domain factors</td>
<td>STAT factors</td>
<td>N/A</td>
</tr>
</tbody>
</table>

</div>
</div>
</div>
<p>The <code>motif_id</code> in this data table is not the same as the <code>motif_id</code> in the <code>merged_filt</code> table. This is because for some databases the TF name is not a part of the motif ID and thus need to be manually added.</p>
<div id="cell-90" class="cell" data-execution_count="12">
<div class="sourceCode cell-code" id="cb64"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb64-1"><a href="#cb64-1" aria-hidden="true" tabindex="-1"></a><span class="co"># go through the motif_id column and use the corresponding value in the tf_name column to do a match</span></span>
<span id="cb64-2"><a href="#cb64-2" aria-hidden="true" tabindex="-1"></a><span class="co"># if there is a match, do nothing, else, add the tf_name to the motif_id by prefixing it and adding an underscore</span></span>
<span id="cb64-3"><a href="#cb64-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb64-4"><a href="#cb64-4" aria-hidden="true" tabindex="-1"></a><span class="co"># def add_prefix(row):</span></span>
<span id="cb64-5"><a href="#cb64-5" aria-hidden="true" tabindex="-1"></a><span class="co">#     if row['tf_name'] not in row['motif_id']:</span></span>
<span id="cb64-6"><a href="#cb64-6" aria-hidden="true" tabindex="-1"></a><span class="co">#         return row['tf_name'] + '_' + row['motif_id']</span></span>
<span id="cb64-7"><a href="#cb64-7" aria-hidden="true" tabindex="-1"></a><span class="co">#     else:</span></span>
<span id="cb64-8"><a href="#cb64-8" aria-hidden="true" tabindex="-1"></a><span class="co">#         return row['motif_id']</span></span>
<span id="cb64-9"><a href="#cb64-9" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb64-10"><a href="#cb64-10" aria-hidden="true" tabindex="-1"></a><span class="co"># scratch that; the motif_id in the other data table have redundant tf_name prefixes, so just use the motif_id as is</span></span>
<span id="cb64-11"><a href="#cb64-11" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb64-12"><a href="#cb64-12" aria-hidden="true" tabindex="-1"></a><span class="kw">def</span> add_prefix(row):</span>
<span id="cb64-13"><a href="#cb64-13" aria-hidden="true" tabindex="-1"></a>    <span class="cf">return</span> row[<span class="st">'tf_name'</span>] <span class="op">+</span> <span class="st">'_'</span> <span class="op">+</span> row[<span class="st">'motif_id'</span>]</span>
<span id="cb64-14"><a href="#cb64-14" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb64-15"><a href="#cb64-15" aria-hidden="true" tabindex="-1"></a>annot_data[<span class="st">'motif_id'</span>] <span class="op">=</span> annot_data.<span class="bu">apply</span>(add_prefix, axis<span class="op">=</span><span class="dv">1</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div id="cell-91" class="cell" data-execution_count="13">
<div class="sourceCode cell-code" id="cb65"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb65-1"><a href="#cb65-1" aria-hidden="true" tabindex="-1"></a>annot_data</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display" data-execution_count="13">
<div>


<table class="dataframe table table-sm table-striped small" data-quarto-postprocess="true" data-border="1">
<thead>
<tr class="header">
<th data-quarto-table-cell-role="th"></th>
<th data-quarto-table-cell-role="th">motif_id</th>
<th data-quarto-table-cell-role="th">tf_name</th>
<th data-quarto-table-cell-role="th">tf_class</th>
<th data-quarto-table-cell-role="th">tf_family</th>
<th data-quarto-table-cell-role="th">tf_subfamily</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td data-quarto-table-cell-role="th">0</td>
<td>AHR_AHR_HUMAN.H11MO.0.B</td>
<td>AHR</td>
<td>basic helix-loop-helix (bHLH)</td>
<td>PAS domain factors</td>
<td>Ahr-like factors</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">1</td>
<td>AIRE_AIRE_HUMAN.H11MO.0.C</td>
<td>AIRE</td>
<td>SAND domain factors</td>
<td>AIRE</td>
<td>AIRE</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">2</td>
<td>ALX1_ALX1_HUMAN.H11MO.0.B</td>
<td>ALX1</td>
<td>homeo domain factors</td>
<td>paired-related HD factors</td>
<td>ALX</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">3</td>
<td>ANDR_ANDR_HUMAN.H11MO.0.A</td>
<td>ANDR</td>
<td>nuclear receptors with C4 zinc fingers</td>
<td>steroid hormone receptors (NR3)</td>
<td>GR-like receptors (NR3C)</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">4</td>
<td>AP2A_AP2A_HUMAN.H11MO.0.A</td>
<td>AP2A</td>
<td>basic helix-span-helix (bHSH) factors</td>
<td>AP2</td>
<td>N/A</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">...</td>
<td>...</td>
<td>...</td>
<td>...</td>
<td>...</td>
<td>...</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">1355</td>
<td>ZSCAN29_MA1602.1</td>
<td>ZSCAN29</td>
<td>C2H2 zinc finger factors</td>
<td>more than 3 adjacent zinc finger factors</td>
<td>ZNF500-like factors</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">1356</td>
<td>ZSCAN30_M08365_2.00</td>
<td>ZSCAN30</td>
<td>C2H2 zinc finger factors</td>
<td>more than 3 adjacent zinc finger factors</td>
<td>unclassified</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">1357</td>
<td>ZSCAN31_M08270_2.00</td>
<td>ZSCAN31</td>
<td>C2H2 zinc finger factors</td>
<td>more than 3 adjacent zinc finger factors</td>
<td>ZNF24-like factors</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">1358</td>
<td>ZSCAN31_M08393_2.00</td>
<td>ZSCAN31</td>
<td>C2H2 zinc finger factors</td>
<td>more than 3 adjacent zinc finger factors</td>
<td>ZNF24-like factors</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">1359</td>
<td>ZSCAN5C_M08390_2.00</td>
<td>ZSCAN5C</td>
<td>C2H2 zinc finger factors</td>
<td>more than 3 adjacent zinc finger factors</td>
<td>ZSCAN5-like zinc finger factors</td>
</tr>
</tbody>
</table>

<p>1360 rows × 5 columns</p>
</div>
</div>
</div>
<div id="cell-92" class="cell" data-execution_count="14">
<div class="sourceCode cell-code" id="cb66"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb66-1"><a href="#cb66-1" aria-hidden="true" tabindex="-1"></a>annot_data[annot_data[<span class="st">'tf_name'</span>] <span class="op">==</span> <span class="st">'STAT1STAT2'</span>]</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display" data-execution_count="14">
<div>


<table class="dataframe table table-sm table-striped small" data-quarto-postprocess="true" data-border="1">
<thead>
<tr class="header">
<th data-quarto-table-cell-role="th"></th>
<th data-quarto-table-cell-role="th">motif_id</th>
<th data-quarto-table-cell-role="th">tf_name</th>
<th data-quarto-table-cell-role="th">tf_class</th>
<th data-quarto-table-cell-role="th">tf_family</th>
<th data-quarto-table-cell-role="th">tf_subfamily</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td data-quarto-table-cell-role="th">771</td>
<td>STAT1STAT2_MA0517.1</td>
<td>STAT1STAT2</td>
<td>STAT domain factors</td>
<td>STAT factors</td>
<td>N/A</td>
</tr>
</tbody>
</table>

</div>
</div>
</div>
<div id="cell-93" class="cell" data-execution_count="15">
<div class="sourceCode cell-code" id="cb67"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb67-1"><a href="#cb67-1" aria-hidden="true" tabindex="-1"></a>annot_data.head()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display" data-execution_count="15">
<div>


<table class="dataframe table table-sm table-striped small" data-quarto-postprocess="true" data-border="1">
<thead>
<tr class="header">
<th data-quarto-table-cell-role="th"></th>
<th data-quarto-table-cell-role="th">motif_id</th>
<th data-quarto-table-cell-role="th">tf_name</th>
<th data-quarto-table-cell-role="th">tf_class</th>
<th data-quarto-table-cell-role="th">tf_family</th>
<th data-quarto-table-cell-role="th">tf_subfamily</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td data-quarto-table-cell-role="th">0</td>
<td>AHR_AHR_HUMAN.H11MO.0.B</td>
<td>AHR</td>
<td>basic helix-loop-helix (bHLH)</td>
<td>PAS domain factors</td>
<td>Ahr-like factors</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">1</td>
<td>AIRE_AIRE_HUMAN.H11MO.0.C</td>
<td>AIRE</td>
<td>SAND domain factors</td>
<td>AIRE</td>
<td>AIRE</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">2</td>
<td>ALX1_ALX1_HUMAN.H11MO.0.B</td>
<td>ALX1</td>
<td>homeo domain factors</td>
<td>paired-related HD factors</td>
<td>ALX</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">3</td>
<td>ANDR_ANDR_HUMAN.H11MO.0.A</td>
<td>ANDR</td>
<td>nuclear receptors with C4 zinc fingers</td>
<td>steroid hormone receptors (NR3)</td>
<td>GR-like receptors (NR3C)</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">4</td>
<td>AP2A_AP2A_HUMAN.H11MO.0.A</td>
<td>AP2A</td>
<td>basic helix-span-helix (bHSH) factors</td>
<td>AP2</td>
<td>N/A</td>
</tr>
</tbody>
</table>

</div>
</div>
</div>
<div id="cell-94" class="cell" data-execution_count="16">
<div class="sourceCode cell-code" id="cb68"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb68-1"><a href="#cb68-1" aria-hidden="true" tabindex="-1"></a><span class="co"># save the dataframe to a new file</span></span>
<span id="cb68-2"><a href="#cb68-2" aria-hidden="true" tabindex="-1"></a>annot_data.to_csv(<span class="st">'/Users/sufyazi/Documents/local-storage/bioinf/repos/reference-TF-motifs/combined-filt-individual-motifs/TF-motif-full-metadata-manual.processed.tsv'</span>, sep<span class="op">=</span><span class="st">'</span><span class="ch">\t</span><span class="st">'</span>, index<span class="op">=</span><span class="va">False</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<section id="visualizing-annotation-data" class="level4">
<h4 class="anchored" data-anchor-id="visualizing-annotation-data">Visualizing Annotation Data</h4>
<p>We can visualize the distribution of TF classes and TF IDs in the annotation data with a barplot.</p>
<div id="cell-96" class="cell" data-execution_count="29">
<div class="sourceCode cell-code" id="cb69"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb69-1"><a href="#cb69-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Get the order of categories based on their counts</span></span>
<span id="cb69-2"><a href="#cb69-2" aria-hidden="true" tabindex="-1"></a>order <span class="op">=</span> annot_data[<span class="st">'tf_class'</span>].value_counts().index</span>
<span id="cb69-3"><a href="#cb69-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb69-4"><a href="#cb69-4" aria-hidden="true" tabindex="-1"></a><span class="co"># Create a larger figure</span></span>
<span id="cb69-5"><a href="#cb69-5" aria-hidden="true" tabindex="-1"></a>plt.figure(figsize<span class="op">=</span>(<span class="dv">10</span>, <span class="bu">len</span>(order) <span class="op">*</span> <span class="fl">0.4</span>))</span>
<span id="cb69-6"><a href="#cb69-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb69-7"><a href="#cb69-7" aria-hidden="true" tabindex="-1"></a><span class="co"># Create a countplot with the order</span></span>
<span id="cb69-8"><a href="#cb69-8" aria-hidden="true" tabindex="-1"></a>ax <span class="op">=</span> sns.countplot(y<span class="op">=</span><span class="st">'tf_class'</span>, data<span class="op">=</span>annot_data, order<span class="op">=</span>order,color<span class="op">=</span><span class="st">'royalblue'</span>)</span>
<span id="cb69-9"><a href="#cb69-9" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb69-10"><a href="#cb69-10" aria-hidden="true" tabindex="-1"></a><span class="co"># Rotate x-axis labels for readability if they are long</span></span>
<span id="cb69-11"><a href="#cb69-11" aria-hidden="true" tabindex="-1"></a>plt.xticks(rotation<span class="op">=</span><span class="dv">90</span>)</span>
<span id="cb69-12"><a href="#cb69-12" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb69-13"><a href="#cb69-13" aria-hidden="true" tabindex="-1"></a><span class="co">#  y axis label</span></span>
<span id="cb69-14"><a href="#cb69-14" aria-hidden="true" tabindex="-1"></a>plt.ylabel(<span class="st">'TF class'</span>)</span>
<span id="cb69-15"><a href="#cb69-15" aria-hidden="true" tabindex="-1"></a><span class="co"># x axis label</span></span>
<span id="cb69-16"><a href="#cb69-16" aria-hidden="true" tabindex="-1"></a>plt.xlabel(<span class="st">'Number of motifs'</span>)</span>
<span id="cb69-17"><a href="#cb69-17" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb69-18"><a href="#cb69-18" aria-hidden="true" tabindex="-1"></a><span class="co"># Increase the space between the y-axis labels</span></span>
<span id="cb69-19"><a href="#cb69-19" aria-hidden="true" tabindex="-1"></a><span class="co"># plt.setp(ax.get_yticklabels(), fontsize=12, verticalalignment='center')</span></span>
<span id="cb69-20"><a href="#cb69-20" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb69-21"><a href="#cb69-21" aria-hidden="true" tabindex="-1"></a><span class="co"># Show the plot</span></span>
<span id="cb69-22"><a href="#cb69-22" aria-hidden="true" tabindex="-1"></a>plt.show()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><img src="variant-freq-analysis-v5_files/figure-html/cell-66-output-1.png" class="img-fluid figure-img"></p>
</figure>
</div>
</div>
</div>
<div id="cell-97" class="cell" data-execution_count="31">
<div class="sourceCode cell-code" id="cb70"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb70-1"><a href="#cb70-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Get the counts of each category in the 'tf_class' column</span></span>
<span id="cb70-2"><a href="#cb70-2" aria-hidden="true" tabindex="-1"></a>class_counts <span class="op">=</span> annot_data[<span class="st">'tf_class'</span>].value_counts().to_frame()</span>
<span id="cb70-3"><a href="#cb70-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb70-4"><a href="#cb70-4" aria-hidden="true" tabindex="-1"></a><span class="co"># Rename the column for clarity</span></span>
<span id="cb70-5"><a href="#cb70-5" aria-hidden="true" tabindex="-1"></a>class_counts.columns <span class="op">=</span> [<span class="st">'Motif Counts'</span>]</span>
<span id="cb70-6"><a href="#cb70-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb70-7"><a href="#cb70-7" aria-hidden="true" tabindex="-1"></a><span class="co"># rename the index to 'TF class'</span></span>
<span id="cb70-8"><a href="#cb70-8" aria-hidden="true" tabindex="-1"></a>class_counts.index.name <span class="op">=</span> <span class="st">'TF class'</span></span>
<span id="cb70-9"><a href="#cb70-9" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb70-10"><a href="#cb70-10" aria-hidden="true" tabindex="-1"></a><span class="co"># Display the DataFrame with conditional formatting</span></span>
<span id="cb70-11"><a href="#cb70-11" aria-hidden="true" tabindex="-1"></a>class_counts.style.background_gradient(cmap<span class="op">=</span><span class="st">'Blues'</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display" data-execution_count="31">
<style type="text/css">
#T_d297c_row0_col0 {
  background-color: #08306b;
  color: #f1f1f1;
}
#T_d297c_row1_col0 {
  background-color: #ddeaf7;
  color: #000000;
}
#T_d297c_row2_col0 {
  background-color: #dfebf7;
  color: #000000;
}
#T_d297c_row3_col0 {
  background-color: #e4eff9;
  color: #000000;
}
#T_d297c_row4_col0 {
  background-color: #e5eff9;
  color: #000000;
}
#T_d297c_row5_col0 {
  background-color: #e7f0fa;
  color: #000000;
}
#T_d297c_row6_col0 {
  background-color: #edf4fc;
  color: #000000;
}
#T_d297c_row7_col0 {
  background-color: #eef5fc;
  color: #000000;
}
#T_d297c_row8_col0 {
  background-color: #eff6fc;
  color: #000000;
}
#T_d297c_row9_col0 {
  background-color: #f0f6fd;
  color: #000000;
}
#T_d297c_row10_col0, #T_d297c_row11_col0, #T_d297c_row12_col0 {
  background-color: #f2f7fd;
  color: #000000;
}
#T_d297c_row13_col0 {
  background-color: #f3f8fe;
  color: #000000;
}
#T_d297c_row14_col0 {
  background-color: #f4f9fe;
  color: #000000;
}
#T_d297c_row15_col0 {
  background-color: #f5f9fe;
  color: #000000;
}
#T_d297c_row16_col0, #T_d297c_row17_col0 {
  background-color: #f5fafe;
  color: #000000;
}
#T_d297c_row18_col0, #T_d297c_row19_col0, #T_d297c_row20_col0, #T_d297c_row21_col0, #T_d297c_row22_col0, #T_d297c_row23_col0 {
  background-color: #f6faff;
  color: #000000;
}
#T_d297c_row24_col0, #T_d297c_row25_col0, #T_d297c_row26_col0, #T_d297c_row27_col0, #T_d297c_row28_col0, #T_d297c_row29_col0, #T_d297c_row30_col0, #T_d297c_row31_col0 {
  background-color: #f7fbff;
  color: #000000;
}
</style>

<table id="T_d297c" data-quarto-postprocess="true" class="table table-sm table-striped small">
<thead>
<tr class="header">
<th class="blank level0" data-quarto-table-cell-role="th">&nbsp;</th>
<th id="T_d297c_level0_col0" class="col_heading level0 col0" data-quarto-table-cell-role="th">Motif Counts</th>
</tr>
<tr class="odd">
<th class="index_name level0" data-quarto-table-cell-role="th">TF class</th>
<th class="blank col0" data-quarto-table-cell-role="th">&nbsp;</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td id="T_d297c_level0_row0" class="row_heading level0 row0" data-quarto-table-cell-role="th">C2H2 zinc finger factors</td>
<td id="T_d297c_row0_col0" class="data row0 col0">699</td>
</tr>
<tr class="even">
<td id="T_d297c_level0_row1" class="row_heading level0 row1" data-quarto-table-cell-role="th">basic leucine zipper (bZIP) factors</td>
<td id="T_d297c_row1_col0" class="data row1 col0">92</td>
</tr>
<tr class="odd">
<td id="T_d297c_level0_row2" class="row_heading level0 row2" data-quarto-table-cell-role="th">homeo domain factors</td>
<td id="T_d297c_row2_col0" class="data row2 col0">87</td>
</tr>
<tr class="even">
<td id="T_d297c_level0_row3" class="row_heading level0 row3" data-quarto-table-cell-role="th">nuclear receptors with C4 zinc fingers</td>
<td id="T_d297c_row3_col0" class="data row3 col0">68</td>
</tr>
<tr class="odd">
<td id="T_d297c_level0_row4" class="row_heading level0 row4" data-quarto-table-cell-role="th">forkhead/winged helix factors</td>
<td id="T_d297c_row4_col0" class="data row4 col0">64</td>
</tr>
<tr class="even">
<td id="T_d297c_level0_row5" class="row_heading level0 row5" data-quarto-table-cell-role="th">tryptophan cluster factors</td>
<td id="T_d297c_row5_col0" class="data row5 col0">60</td>
</tr>
<tr class="odd">
<td id="T_d297c_level0_row6" class="row_heading level0 row6" data-quarto-table-cell-role="th">basic helix-loop-helix factors (bHLH)</td>
<td id="T_d297c_row6_col0" class="data row6 col0">38</td>
</tr>
<tr class="even">
<td id="T_d297c_level0_row7" class="row_heading level0 row7" data-quarto-table-cell-role="th">basic helix-loop-helix (bHLH)</td>
<td id="T_d297c_row7_col0" class="data row7 col0">31</td>
</tr>
<tr class="odd">
<td id="T_d297c_level0_row8" class="row_heading level0 row8" data-quarto-table-cell-role="th">basic helix-loop-helix (bHLH) factors</td>
<td id="T_d297c_row8_col0" class="data row8 col0">29</td>
</tr>
<tr class="even">
<td id="T_d297c_level0_row9" class="row_heading level0 row9" data-quarto-table-cell-role="th">Rel homology region (RHR) factors</td>
<td id="T_d297c_row9_col0" class="data row9 col0">27</td>
</tr>
<tr class="odd">
<td id="T_d297c_level0_row10" class="row_heading level0 row10" data-quarto-table-cell-role="th">other C4 zinc finger-type factors</td>
<td id="T_d297c_row10_col0" class="data row10 col0">22</td>
</tr>
<tr class="even">
<td id="T_d297c_level0_row11" class="row_heading level0 row11" data-quarto-table-cell-role="th">STAT domain factors</td>
<td id="T_d297c_row11_col0" class="data row11 col0">21</td>
</tr>
<tr class="odd">
<td id="T_d297c_level0_row12" class="row_heading level0 row12" data-quarto-table-cell-role="th">high-mobility group (HMG) domain factors</td>
<td id="T_d297c_row12_col0" class="data row12 col0">21</td>
</tr>
<tr class="even">
<td id="T_d297c_level0_row13" class="row_heading level0 row13" data-quarto-table-cell-role="th">MADS box factors</td>
<td id="T_d297c_row13_col0" class="data row13 col0">16</td>
</tr>
<tr class="odd">
<td id="T_d297c_level0_row14" class="row_heading level0 row14" data-quarto-table-cell-role="th">basic leucine zipper factors (bZIP)</td>
<td id="T_d297c_row14_col0" class="data row14 col0">14</td>
</tr>
<tr class="even">
<td id="T_d297c_level0_row15" class="row_heading level0 row15" data-quarto-table-cell-role="th">paired box factors</td>
<td id="T_d297c_row15_col0" class="data row15 col0">10</td>
</tr>
<tr class="odd">
<td id="T_d297c_level0_row16" class="row_heading level0 row16" data-quarto-table-cell-role="th">SMAD/NF-1 DNA-binding domain factors</td>
<td id="T_d297c_row16_col0" class="data row16 col0">8</td>
</tr>
<tr class="even">
<td id="T_d297c_level0_row17" class="row_heading level0 row17" data-quarto-table-cell-role="th">heteromeric CCAAT-binding factors</td>
<td id="T_d297c_row17_col0" class="data row17 col0">8</td>
</tr>
<tr class="odd">
<td id="T_d297c_level0_row18" class="row_heading level0 row18" data-quarto-table-cell-role="th">p53 domain factors</td>
<td id="T_d297c_row18_col0" class="data row18 col0">6</td>
</tr>
<tr class="even">
<td id="T_d297c_level0_row19" class="row_heading level0 row19" data-quarto-table-cell-role="th">TEA domain factors</td>
<td id="T_d297c_row19_col0" class="data row19 col0">6</td>
</tr>
<tr class="odd">
<td id="T_d297c_level0_row20" class="row_heading level0 row20" data-quarto-table-cell-role="th">N/A</td>
<td id="T_d297c_row20_col0" class="data row20 col0">5</td>
</tr>
<tr class="even">
<td id="T_d297c_level0_row21" class="row_heading level0 row21" data-quarto-table-cell-role="th">basic helix-span-helix factors (bHSH)</td>
<td id="T_d297c_row21_col0" class="data row21 col0">4</td>
</tr>
<tr class="odd">
<td id="T_d297c_level0_row22" class="row_heading level0 row22" data-quarto-table-cell-role="th">Runt domain factors</td>
<td id="T_d297c_row22_col0" class="data row22 col0">4</td>
</tr>
<tr class="even">
<td id="T_d297c_level0_row23" class="row_heading level0 row23" data-quarto-table-cell-role="th">C2CH THAP-type zinc finger factors</td>
<td id="T_d297c_row23_col0" class="data row23 col0">4</td>
</tr>
<tr class="odd">
<td id="T_d297c_level0_row24" class="row_heading level0 row24" data-quarto-table-cell-role="th">heat shock factors</td>
<td id="T_d297c_row24_col0" class="data row24 col0">3</td>
</tr>
<tr class="even">
<td id="T_d297c_level0_row25" class="row_heading level0 row25" data-quarto-table-cell-role="th">basic helix-span-helix (bHSH) factors</td>
<td id="T_d297c_row25_col0" class="data row25 col0">3</td>
</tr>
<tr class="odd">
<td id="T_d297c_level0_row26" class="row_heading level0 row26" data-quarto-table-cell-role="th">Grainyhead domain factors</td>
<td id="T_d297c_row26_col0" class="data row26 col0">3</td>
</tr>
<tr class="even">
<td id="T_d297c_level0_row27" class="row_heading level0 row27" data-quarto-table-cell-role="th">TATA-binding proteins</td>
<td id="T_d297c_row27_col0" class="data row27 col0">2</td>
</tr>
<tr class="odd">
<td id="T_d297c_level0_row28" class="row_heading level0 row28" data-quarto-table-cell-role="th">T-Box factors</td>
<td id="T_d297c_row28_col0" class="data row28 col0">2</td>
</tr>
<tr class="even">
<td id="T_d297c_level0_row29" class="row_heading level0 row29" data-quarto-table-cell-role="th">T-box factors</td>
<td id="T_d297c_row29_col0" class="data row29 col0">1</td>
</tr>
<tr class="odd">
<td id="T_d297c_level0_row30" class="row_heading level0 row30" data-quarto-table-cell-role="th">ARID domain factors</td>
<td id="T_d297c_row30_col0" class="data row30 col0">1</td>
</tr>
<tr class="even">
<td id="T_d297c_level0_row31" class="row_heading level0 row31" data-quarto-table-cell-role="th">SAND domain factors</td>
<td id="T_d297c_row31_col0" class="data row31 col0">1</td>
</tr>
</tbody>
</table>
</div>
</div>
</section>
<section id="merging-annotation-data-with-covariant-sites" class="level3">
<h3 class="anchored" data-anchor-id="merging-annotation-data-with-covariant-sites">Merging Annotation Data with Covariant Sites</h3>
<div id="cell-99" class="cell" data-execution_count="23">
<div class="sourceCode cell-code" id="cb71"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb71-1"><a href="#cb71-1" aria-hidden="true" tabindex="-1"></a><span class="co"># load the processed data tables</span></span>
<span id="cb71-2"><a href="#cb71-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb71-3"><a href="#cb71-3" aria-hidden="true" tabindex="-1"></a>annot_data <span class="op">=</span> pd.read_csv(<span class="st">'/Users/sufyazi/Documents/local-storage/bioinf/repos/reference-TF-motifs/combined-filt-individual-motifs/TF-motif-full-metadata-manual.processed.tsv'</span>, sep<span class="op">=</span><span class="st">'</span><span class="ch">\t</span><span class="st">'</span>, header<span class="op">=</span><span class="dv">0</span>)</span>
<span id="cb71-4"><a href="#cb71-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb71-5"><a href="#cb71-5" aria-hidden="true" tabindex="-1"></a>annot_data</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display" data-execution_count="23">
<div>


<table class="dataframe table table-sm table-striped small" data-quarto-postprocess="true" data-border="1">
<thead>
<tr class="header">
<th data-quarto-table-cell-role="th"></th>
<th data-quarto-table-cell-role="th">motif_id</th>
<th data-quarto-table-cell-role="th">tf_name</th>
<th data-quarto-table-cell-role="th">tf_class</th>
<th data-quarto-table-cell-role="th">tf_family</th>
<th data-quarto-table-cell-role="th">tf_subfamily</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td data-quarto-table-cell-role="th">0</td>
<td>AHR_AHR_HUMAN.H11MO.0.B</td>
<td>AHR</td>
<td>basic helix-loop-helix (bHLH)</td>
<td>PAS domain factors</td>
<td>Ahr-like factors</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">1</td>
<td>AIRE_AIRE_HUMAN.H11MO.0.C</td>
<td>AIRE</td>
<td>SAND domain factors</td>
<td>AIRE</td>
<td>AIRE</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">2</td>
<td>ALX1_ALX1_HUMAN.H11MO.0.B</td>
<td>ALX1</td>
<td>homeo domain factors</td>
<td>paired-related HD factors</td>
<td>ALX</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">3</td>
<td>ANDR_ANDR_HUMAN.H11MO.0.A</td>
<td>ANDR</td>
<td>nuclear receptors with C4 zinc fingers</td>
<td>steroid hormone receptors (NR3)</td>
<td>GR-like receptors (NR3C)</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">4</td>
<td>AP2A_AP2A_HUMAN.H11MO.0.A</td>
<td>AP2A</td>
<td>basic helix-span-helix (bHSH) factors</td>
<td>AP2</td>
<td>NaN</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">...</td>
<td>...</td>
<td>...</td>
<td>...</td>
<td>...</td>
<td>...</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">1355</td>
<td>ZSCAN29_MA1602.1</td>
<td>ZSCAN29</td>
<td>C2H2 zinc finger factors</td>
<td>more than 3 adjacent zinc finger factors</td>
<td>ZNF500-like factors</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">1356</td>
<td>ZSCAN30_M08365_2.00</td>
<td>ZSCAN30</td>
<td>C2H2 zinc finger factors</td>
<td>more than 3 adjacent zinc finger factors</td>
<td>unclassified</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">1357</td>
<td>ZSCAN31_M08270_2.00</td>
<td>ZSCAN31</td>
<td>C2H2 zinc finger factors</td>
<td>more than 3 adjacent zinc finger factors</td>
<td>ZNF24-like factors</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">1358</td>
<td>ZSCAN31_M08393_2.00</td>
<td>ZSCAN31</td>
<td>C2H2 zinc finger factors</td>
<td>more than 3 adjacent zinc finger factors</td>
<td>ZNF24-like factors</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">1359</td>
<td>ZSCAN5C_M08390_2.00</td>
<td>ZSCAN5C</td>
<td>C2H2 zinc finger factors</td>
<td>more than 3 adjacent zinc finger factors</td>
<td>ZSCAN5-like zinc finger factors</td>
</tr>
</tbody>
</table>

<p>1360 rows × 5 columns</p>
</div>
</div>
</div>
<div id="cell-100" class="cell" data-execution_count="24">
<div class="sourceCode cell-code" id="cb72"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb72-1"><a href="#cb72-1" aria-hidden="true" tabindex="-1"></a><span class="co"># load the covariant sites data</span></span>
<span id="cb72-2"><a href="#cb72-2" aria-hidden="true" tabindex="-1"></a>sig_covariant_sites <span class="op">=</span> pd.read_csv(<span class="st">'/Users/sufyazi/Documents/local-storage/bioinf/repos/fps-var-integration-local/output-data/covariant-site-counts/AF_FPS-covariant_sites-significant.combined.tsv'</span>, sep<span class="op">=</span><span class="st">'</span><span class="ch">\t</span><span class="st">'</span>, header<span class="op">=</span><span class="dv">0</span>)</span>
<span id="cb72-3"><a href="#cb72-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb72-4"><a href="#cb72-4" aria-hidden="true" tabindex="-1"></a>sig_covariant_sites</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display" data-execution_count="24">
<div>


<table class="dataframe table table-sm table-striped small" data-quarto-postprocess="true" data-border="1">
<thead>
<tr class="header">
<th data-quarto-table-cell-role="th"></th>
<th data-quarto-table-cell-role="th">orig_index</th>
<th data-quarto-table-cell-role="th">motif_id</th>
<th data-quarto-table-cell-role="th">region_id</th>
<th data-quarto-table-cell-role="th">corr_coeff</th>
<th data-quarto-table-cell-role="th">pvalues</th>
<th data-quarto-table-cell-role="th">adj_pvalues</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td data-quarto-table-cell-role="th">0</td>
<td>71</td>
<td>AIRE_AIRE_HUMAN.H11MO.0.C</td>
<td>chr16:3282261-3282279</td>
<td>-1.0</td>
<td>1.404265e-24</td>
<td>4.446841e-23</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">1</td>
<td>77</td>
<td>AIRE_AIRE_HUMAN.H11MO.0.C</td>
<td>chr17:49374914-49374932</td>
<td>1.0</td>
<td>1.404265e-24</td>
<td>4.446841e-23</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">2</td>
<td>24</td>
<td>AIRE_AIRE_HUMAN.H11MO.0.C</td>
<td>chr6:76866202-76866220</td>
<td>1.0</td>
<td>1.404265e-24</td>
<td>4.446841e-23</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">3</td>
<td>14</td>
<td>ALX1_ALX1_HUMAN.H11MO.0.B</td>
<td>chr2:225782219-225782231</td>
<td>1.0</td>
<td>1.404265e-24</td>
<td>6.740474e-23</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">4</td>
<td>244</td>
<td>ANDR_ANDR_HUMAN.H11MO.0.A</td>
<td>chr14:99571643-99571661</td>
<td>-1.0</td>
<td>1.404265e-24</td>
<td>1.507245e-22</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">...</td>
<td>...</td>
<td>...</td>
<td>...</td>
<td>...</td>
<td>...</td>
<td>...</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">4896</td>
<td>122</td>
<td>ZSCAN31_M08393_2.00</td>
<td>chr18:13246839-13246854</td>
<td>-1.0</td>
<td>1.404265e-24</td>
<td>9.689431e-23</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">4897</td>
<td>36</td>
<td>ZSCAN31_M08393_2.00</td>
<td>chr5:53402883-53402898</td>
<td>-1.0</td>
<td>1.404265e-24</td>
<td>9.689431e-23</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">4898</td>
<td>37</td>
<td>ZSCAN5C_M08390_2.00</td>
<td>chr7:133512687-133512702</td>
<td>1.0</td>
<td>1.404265e-24</td>
<td>3.744708e-23</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">4899</td>
<td>36</td>
<td>ZSCAN5C_M08390_2.00</td>
<td>chr7:98352788-98352803</td>
<td>-1.0</td>
<td>1.404265e-24</td>
<td>3.744708e-23</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">4900</td>
<td>43</td>
<td>ZSCAN5C_M08390_2.00</td>
<td>chr8:142727247-142727262</td>
<td>1.0</td>
<td>1.404265e-24</td>
<td>3.744708e-23</td>
</tr>
</tbody>
</table>

<p>4901 rows × 6 columns</p>
</div>
</div>
</div>
<div id="cell-101" class="cell" data-execution_count="25">
<div class="sourceCode cell-code" id="cb73"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb73-1"><a href="#cb73-1" aria-hidden="true" tabindex="-1"></a><span class="co"># group the merged dataframe by 'motif_id' and count the number of unique 'region_id' values</span></span>
<span id="cb73-2"><a href="#cb73-2" aria-hidden="true" tabindex="-1"></a>motif_counts <span class="op">=</span> sig_covariant_sites.groupby(<span class="st">'motif_id'</span>)[<span class="st">'region_id'</span>].nunique().reset_index()</span>
<span id="cb73-3"><a href="#cb73-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb73-4"><a href="#cb73-4" aria-hidden="true" tabindex="-1"></a><span class="co"># rename the 'region_id' column to 'site_counts'</span></span>
<span id="cb73-5"><a href="#cb73-5" aria-hidden="true" tabindex="-1"></a>motif_counts <span class="op">=</span> motif_counts.rename(columns<span class="op">=</span>{<span class="st">'region_id'</span>: <span class="st">'site_counts'</span>})</span>
<span id="cb73-6"><a href="#cb73-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb73-7"><a href="#cb73-7" aria-hidden="true" tabindex="-1"></a><span class="co"># sort rows by 'region_id' in descending order</span></span>
<span id="cb73-8"><a href="#cb73-8" aria-hidden="true" tabindex="-1"></a>motif_counts <span class="op">=</span> motif_counts.sort_values(by<span class="op">=</span><span class="st">'site_counts'</span>, ascending<span class="op">=</span><span class="va">False</span>)</span>
<span id="cb73-9"><a href="#cb73-9" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb73-10"><a href="#cb73-10" aria-hidden="true" tabindex="-1"></a>motif_counts</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display" data-execution_count="25">
<div>


<table class="dataframe table table-sm table-striped small" data-quarto-postprocess="true" data-border="1">
<thead>
<tr class="header">
<th data-quarto-table-cell-role="th"></th>
<th data-quarto-table-cell-role="th">motif_id</th>
<th data-quarto-table-cell-role="th">site_counts</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td data-quarto-table-cell-role="th">801</td>
<td>ZN770_ZN770_HUMAN.H11MO.0.C</td>
<td>27</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">773</td>
<td>ZN281_ZN281_HUMAN.H11MO.0.A</td>
<td>27</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">869</td>
<td>ZNF263_M08082_2.00</td>
<td>26</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">342</td>
<td>KLF6_KLF6_HUMAN.H11MO.0.A</td>
<td>26</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">695</td>
<td>WT1_WT1_HUMAN.H11MO.0.C</td>
<td>24</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">...</td>
<td>...</td>
<td>...</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">351</td>
<td>LHX2_MA0700.2</td>
<td>1</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">352</td>
<td>LHX3_LHX3_HUMAN.H11MO.0.C</td>
<td>1</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">355</td>
<td>MAFF_M07839_2.00</td>
<td>1</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">357</td>
<td>MAFF_MA0495.3</td>
<td>1</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">278</td>
<td>HXB7_HXB7_HUMAN.H11MO.0.C</td>
<td>1</td>
</tr>
</tbody>
</table>

<p>1153 rows × 2 columns</p>
</div>
</div>
</div>
<div id="cell-102" class="cell" data-execution_count="30">
<div class="sourceCode cell-code" id="cb74"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb74-1"><a href="#cb74-1" aria-hidden="true" tabindex="-1"></a><span class="co"># merge the two dataframes on 'motif_id'</span></span>
<span id="cb74-2"><a href="#cb74-2" aria-hidden="true" tabindex="-1"></a>merged_annot_covar <span class="op">=</span> motif_counts.merge(annot_data, on<span class="op">=</span><span class="st">'motif_id'</span>)</span>
<span id="cb74-3"><a href="#cb74-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb74-4"><a href="#cb74-4" aria-hidden="true" tabindex="-1"></a>merged_annot_covar.head(<span class="dv">10</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display" data-execution_count="30">
<div>


<table class="dataframe table table-sm table-striped small" data-quarto-postprocess="true" data-border="1">
<thead>
<tr class="header">
<th data-quarto-table-cell-role="th"></th>
<th data-quarto-table-cell-role="th">motif_id</th>
<th data-quarto-table-cell-role="th">site_counts</th>
<th data-quarto-table-cell-role="th">tf_name</th>
<th data-quarto-table-cell-role="th">tf_class</th>
<th data-quarto-table-cell-role="th">tf_family</th>
<th data-quarto-table-cell-role="th">tf_subfamily</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td data-quarto-table-cell-role="th">0</td>
<td>ZN770_ZN770_HUMAN.H11MO.0.C</td>
<td>27</td>
<td>ZN770</td>
<td>C2H2 zinc finger factors</td>
<td>factors with multiple dispersed zinc fingers</td>
<td>unclassified</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">1</td>
<td>ZN281_ZN281_HUMAN.H11MO.0.A</td>
<td>27</td>
<td>ZN281</td>
<td>C2H2 zinc finger factors</td>
<td>more than 3 adjacent zinc finger factors</td>
<td>ZNF148-like factors</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">2</td>
<td>ZNF263_M08082_2.00</td>
<td>26</td>
<td>ZNF263</td>
<td>C2H2 zinc finger factors</td>
<td>more than 3 adjacent zinc finger factors</td>
<td>unclassified</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">3</td>
<td>KLF6_KLF6_HUMAN.H11MO.0.A</td>
<td>26</td>
<td>KLF6</td>
<td>C2H2 zinc finger factors</td>
<td>three-zinc finger Krüppel-related factors</td>
<td>Krüppel-like factors</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">4</td>
<td>WT1_WT1_HUMAN.H11MO.0.C</td>
<td>24</td>
<td>WT1</td>
<td>C2H2 zinc finger factors</td>
<td>more than 3 adjacent zinc finger factors</td>
<td>unclassified</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">5</td>
<td>ZN467_ZN467_HUMAN.H11MO.0.C</td>
<td>23</td>
<td>ZN467</td>
<td>C2H2 zinc finger factors</td>
<td>factors with multiple dispersed zinc fingers</td>
<td>unclassified</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">6</td>
<td>SP4_SP4_HUMAN.H11MO.0.A</td>
<td>22</td>
<td>SP4</td>
<td>C2H2 zinc finger factors</td>
<td>three-zinc finger Krüppel-related factors</td>
<td>Sp1-like factors</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">7</td>
<td>SP3_SP3_HUMAN.H11MO.0.B</td>
<td>22</td>
<td>SP3</td>
<td>C2H2 zinc finger factors</td>
<td>three-zinc finger Krüppel-related factors</td>
<td>Sp1-like factors</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">8</td>
<td>SP2_SP2_HUMAN.H11MO.0.A</td>
<td>22</td>
<td>SP2</td>
<td>C2H2 zinc finger factors</td>
<td>three-zinc finger Krüppel-related factors</td>
<td>Sp1-like factors</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">9</td>
<td>PATZ1_PATZ1_HUMAN.H11MO.0.C</td>
<td>22</td>
<td>PATZ1</td>
<td>C2H2 zinc finger factors</td>
<td>factors with multiple dispersed zinc fingers</td>
<td>MAZ-like factors</td>
</tr>
</tbody>
</table>

</div>
</div>
</div>
<div id="cell-103" class="cell" data-execution_count="43">
<div class="sourceCode cell-code" id="cb75"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb75-1"><a href="#cb75-1" aria-hidden="true" tabindex="-1"></a>merged_annot_covar.iloc[<span class="dv">10</span>:<span class="dv">50</span>]</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display" data-execution_count="43">
<div>


<table class="dataframe table table-sm table-striped small" data-quarto-postprocess="true" data-border="1">
<thead>
<tr class="header">
<th data-quarto-table-cell-role="th"></th>
<th data-quarto-table-cell-role="th">motif_id</th>
<th data-quarto-table-cell-role="th">site_counts</th>
<th data-quarto-table-cell-role="th">tf_name</th>
<th data-quarto-table-cell-role="th">tf_class</th>
<th data-quarto-table-cell-role="th">tf_family</th>
<th data-quarto-table-cell-role="th">tf_subfamily</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td data-quarto-table-cell-role="th">10</td>
<td>ZNF383_M07689_2.00</td>
<td>21</td>
<td>ZNF383</td>
<td>C2H2 zinc finger factors</td>
<td>more than 3 adjacent zinc finger factors</td>
<td>ZNF620-like factors</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">11</td>
<td>MAZ_MAZ_HUMAN.H11MO.0.A</td>
<td>21</td>
<td>MAZ</td>
<td>C2H2 zinc finger factors</td>
<td>factors with multiple dispersed zinc fingers</td>
<td>MAZ-like factors</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">12</td>
<td>ZN263_ZN263_HUMAN.H11MO.0.A</td>
<td>19</td>
<td>ZN263</td>
<td>C2H2 zinc finger factors</td>
<td>more than 3 adjacent zinc finger factors</td>
<td>unclassified</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">13</td>
<td>ZNF263_M07843_2.00</td>
<td>19</td>
<td>ZNF263</td>
<td>C2H2 zinc finger factors</td>
<td>more than 3 adjacent zinc finger factors</td>
<td>unclassified</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">14</td>
<td>PRDM9_MA1723.1</td>
<td>19</td>
<td>PRDM9</td>
<td>C2H2 zinc finger factors</td>
<td>more than 3 adjacent zinc finger factors</td>
<td>unclassified</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">15</td>
<td>ZNF529_M07678_2.00</td>
<td>19</td>
<td>ZNF529</td>
<td>C2H2 zinc finger factors</td>
<td>more than 3 adjacent zinc finger factors</td>
<td>unclassified</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">16</td>
<td>ZBT17_ZBT17_HUMAN.H11MO.0.A</td>
<td>19</td>
<td>ZBT17</td>
<td>C2H2 zinc finger factors</td>
<td>factors with multiple dispersed zinc fingers</td>
<td>unclassified</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">17</td>
<td>ZNF341_M08310_2.00</td>
<td>18</td>
<td>ZNF341</td>
<td>C2H2 zinc finger factors</td>
<td>factors with multiple dispersed zinc fingers</td>
<td>unclassified</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">18</td>
<td>ZSC22_ZSC22_HUMAN.H11MO.0.C</td>
<td>18</td>
<td>ZSC22</td>
<td>C2H2 zinc finger factors</td>
<td>more than 3 adjacent zinc finger factors</td>
<td>unclassified</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">19</td>
<td>ZNF432_M07773_2.00</td>
<td>18</td>
<td>ZNF432</td>
<td>C2H2 zinc finger factors</td>
<td>more than 3 adjacent zinc finger factors</td>
<td>ZNF432-like factors</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">20</td>
<td>ZNF880_M07764_2.00</td>
<td>18</td>
<td>ZNF880</td>
<td>C2H2 zinc finger factors</td>
<td>more than 3 adjacent zinc finger factors</td>
<td>unclassified</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">21</td>
<td>ZNF93_M07671_2.00</td>
<td>18</td>
<td>ZNF93</td>
<td>C2H2 zinc finger factors</td>
<td>more than 3 adjacent zinc finger factors</td>
<td>unclassified</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">22</td>
<td>ZNF148_MA1653.1</td>
<td>17</td>
<td>ZNF148</td>
<td>C2H2 zinc finger factors</td>
<td>more than 3 adjacent zinc finger factors</td>
<td>ZNF148-like factors</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">23</td>
<td>SP1_M08363_2.00</td>
<td>17</td>
<td>SP1</td>
<td>C2H2 zinc finger factors</td>
<td>three-zinc finger Krüppel-related factors</td>
<td>Sp1-like factors</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">24</td>
<td>ZNF281_M08321_2.00</td>
<td>17</td>
<td>ZNF281</td>
<td>C2H2 zinc finger factors</td>
<td>more than 3 adjacent zinc finger factors</td>
<td>ZNF148-like factors</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">25</td>
<td>ZNF281_MA1630.2</td>
<td>17</td>
<td>ZNF281</td>
<td>C2H2 zinc finger factors</td>
<td>more than 3 adjacent zinc finger factors</td>
<td>ZNF148-like factors</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">26</td>
<td>ZNF460_MA1596.1</td>
<td>17</td>
<td>ZNF460</td>
<td>C2H2 zinc finger factors</td>
<td>more than 3 adjacent zinc finger factors</td>
<td>ZNF460-like factors</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">27</td>
<td>VEZF1_VEZF1_HUMAN.H11MO.0.C</td>
<td>17</td>
<td>VEZF1</td>
<td>C2H2 zinc finger factors</td>
<td>factors with multiple dispersed zinc fingers</td>
<td>MAZ-like factors</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">28</td>
<td>ZNF180_M07614_2.00</td>
<td>16</td>
<td>ZNF180</td>
<td>C2H2 zinc finger factors</td>
<td>factors with multiple dispersed zinc fingers</td>
<td>ZNF180-like factors</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">29</td>
<td>ZNF467_M08260_2.00</td>
<td>16</td>
<td>ZNF467</td>
<td>C2H2 zinc finger factors</td>
<td>factors with multiple dispersed zinc fingers</td>
<td>unclassified</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">30</td>
<td>ZNF304_M07587_2.00</td>
<td>16</td>
<td>ZNF304</td>
<td>C2H2 zinc finger factors</td>
<td>factors with multiple dispersed zinc fingers</td>
<td>unclassified</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">31</td>
<td>EGR2_EGR2_HUMAN.H11MO.0.A</td>
<td>16</td>
<td>EGR2</td>
<td>C2H2 zinc finger factors</td>
<td>three-zinc finger Krüppel-related factors</td>
<td>EGR factors</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">32</td>
<td>CTCF_MA1929.1</td>
<td>16</td>
<td>CTCF</td>
<td>C2H2 zinc finger factors</td>
<td>more than 3 adjacent zinc finger factors</td>
<td>CTCF-like factors</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">33</td>
<td>PRDM9_M07610_2.00</td>
<td>15</td>
<td>PRDM9</td>
<td>C2H2 zinc finger factors</td>
<td>more than 3 adjacent zinc finger factors</td>
<td>unclassified</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">34</td>
<td>EGR1_EGR1_HUMAN.H11MO.0.A</td>
<td>15</td>
<td>EGR1</td>
<td>C2H2 zinc finger factors</td>
<td>three-zinc finger Krüppel-related factors</td>
<td>EGR factors</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">35</td>
<td>KLF3_KLF3_HUMAN.H11MO.0.B</td>
<td>15</td>
<td>KLF3</td>
<td>C2H2 zinc finger factors</td>
<td>three-zinc finger Krüppel-related factors</td>
<td>Krüppel-like factors</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">36</td>
<td>KLF15_KLF15_HUMAN.H11MO.0.A</td>
<td>15</td>
<td>KLF15</td>
<td>C2H2 zinc finger factors</td>
<td>three-zinc finger Krüppel-related factors</td>
<td>Krüppel-like factors</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">37</td>
<td>ZNF519_M07642_2.00</td>
<td>15</td>
<td>ZNF519</td>
<td>C2H2 zinc finger factors</td>
<td>factors with multiple dispersed zinc fingers</td>
<td>unclassified</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">38</td>
<td>KLF10_M08317_2.00</td>
<td>15</td>
<td>KLF10</td>
<td>C2H2 zinc finger factors</td>
<td>three-zinc finger Krüppel-related factors</td>
<td>Krüppel-like factors</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">39</td>
<td>SP2_M08092_2.00</td>
<td>15</td>
<td>SP2</td>
<td>C2H2 zinc finger factors</td>
<td>three-zinc finger Krüppel-related factors</td>
<td>Sp1-like factors</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">40</td>
<td>MAZ_MA1522.1</td>
<td>14</td>
<td>MAZ</td>
<td>C2H2 zinc finger factors</td>
<td>factors with multiple dispersed zinc fingers</td>
<td>MAZ-like factors</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">41</td>
<td>ZNF121_M08384_2.00</td>
<td>14</td>
<td>ZNF121</td>
<td>C2H2 zinc finger factors</td>
<td>more than 3 adjacent zinc finger factors</td>
<td>unclassified</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">42</td>
<td>ZNF283_M07619_2.00</td>
<td>14</td>
<td>ZNF283</td>
<td>C2H2 zinc finger factors</td>
<td>more than 3 adjacent zinc finger factors</td>
<td>ZFN283-like factors</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">43</td>
<td>ZN341_ZN341_HUMAN.H11MO.0.C</td>
<td>14</td>
<td>ZN341</td>
<td>C2H2 zinc finger factors</td>
<td>factors with multiple dispersed zinc fingers</td>
<td>unclassified</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">44</td>
<td>THAP1_THAP1_HUMAN.H11MO.0.C</td>
<td>14</td>
<td>THAP1</td>
<td>C2CH THAP-type zinc finger factors</td>
<td>THAP-related factors</td>
<td>NaN</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">45</td>
<td>ZFP28_M08378_2.00</td>
<td>14</td>
<td>ZFP28</td>
<td>C2H2 zinc finger factors</td>
<td>more than 3 adjacent zinc finger factors</td>
<td>unclassified</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">46</td>
<td>PATZ1_MA1961.1</td>
<td>14</td>
<td>PATZ1</td>
<td>C2H2 zinc finger factors</td>
<td>factors with multiple dispersed zinc fingers</td>
<td>MAZ-like factors</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">47</td>
<td>ZNF548_M07691_2.00</td>
<td>14</td>
<td>ZNF548</td>
<td>C2H2 zinc finger factors</td>
<td>more than 3 adjacent zinc finger factors</td>
<td>ZNF548-like factors</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">48</td>
<td>ZNF235_M07605_2.00</td>
<td>14</td>
<td>ZNF235</td>
<td>C2H2 zinc finger factors</td>
<td>more than 3 adjacent zinc finger factors</td>
<td>ZNF222-like factors</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">49</td>
<td>ZNF484_M07581_2.00</td>
<td>13</td>
<td>ZNF484</td>
<td>C2H2 zinc finger factors</td>
<td>more than 3 adjacent zinc finger factors</td>
<td>ZFN81-like factors</td>
</tr>
</tbody>
</table>

</div>
</div>
</div>
<div id="cell-104" class="cell" data-execution_count="42">
<div class="sourceCode cell-code" id="cb76"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb76-1"><a href="#cb76-1" aria-hidden="true" tabindex="-1"></a><span class="co"># plot countplot of motif counts</span></span>
<span id="cb76-2"><a href="#cb76-2" aria-hidden="true" tabindex="-1"></a>plt.figure(figsize<span class="op">=</span>(<span class="dv">10</span>, <span class="dv">150</span>), dpi<span class="op">=</span><span class="dv">300</span>)</span>
<span id="cb76-3"><a href="#cb76-3" aria-hidden="true" tabindex="-1"></a>p <span class="op">=</span> sns.barplot(x<span class="op">=</span><span class="st">'site_counts'</span>, y<span class="op">=</span><span class="st">'motif_id'</span>, data<span class="op">=</span>merged_annot_covar, color<span class="op">=</span><span class="st">'green'</span>, edgecolor<span class="op">=</span><span class="st">'black'</span>)</span>
<span id="cb76-4"><a href="#cb76-4" aria-hidden="true" tabindex="-1"></a>p.tick_params(axis<span class="op">=</span><span class="st">'y'</span>, labelsize<span class="op">=</span><span class="dv">6</span>)</span>
<span id="cb76-5"><a href="#cb76-5" aria-hidden="true" tabindex="-1"></a>plt.show()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display">
<div>
<figure class="figure">
<p><img src="variant-freq-analysis-v5_files/figure-html/cell-73-output-1.png" class="img-fluid figure-img"></p>
</figure>
</div>
</div>
</div>
<p>We can get the <code>motif_id</code> column from the <code>motif_counts</code> table, and filter the original <code>sig_covariant_sites</code> table to include only the rows where the <code>motif_id</code> is in the <code>motif_id</code> column of <code>motif_counts</code>.</p>
<div id="cell-106" class="cell" data-execution_count="28">
<div class="sourceCode cell-code" id="cb77"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb77-1"><a href="#cb77-1" aria-hidden="true" tabindex="-1"></a><span class="co"># subset the sig_covariant_sites dataframe to only include the 'motif_id' in the motif_counts dataframe</span></span>
<span id="cb77-2"><a href="#cb77-2" aria-hidden="true" tabindex="-1"></a><span class="co"># the order of the rows in the sig_covariant_sites should follow the order of the motif_counts dataframe</span></span>
<span id="cb77-3"><a href="#cb77-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb77-4"><a href="#cb77-4" aria-hidden="true" tabindex="-1"></a>sig_covar_sites_filt <span class="op">=</span> motif_counts[[<span class="st">'motif_id'</span>]].merge(sig_covariant_sites, on<span class="op">=</span><span class="st">'motif_id'</span>)</span>
<span id="cb77-5"><a href="#cb77-5" aria-hidden="true" tabindex="-1"></a>sig_covar_sites_filt</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display" data-execution_count="28">
<div>


<table class="dataframe table table-sm table-striped small" data-quarto-postprocess="true" data-border="1">
<thead>
<tr class="header">
<th data-quarto-table-cell-role="th"></th>
<th data-quarto-table-cell-role="th">motif_id</th>
<th data-quarto-table-cell-role="th">orig_index</th>
<th data-quarto-table-cell-role="th">region_id</th>
<th data-quarto-table-cell-role="th">corr_coeff</th>
<th data-quarto-table-cell-role="th">pvalues</th>
<th data-quarto-table-cell-role="th">adj_pvalues</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td data-quarto-table-cell-role="th">0</td>
<td>ZN770_ZN770_HUMAN.H11MO.0.C</td>
<td>722</td>
<td>chr10:49916101-49916123</td>
<td>1.0</td>
<td>1.404265e-24</td>
<td>7.687053e-23</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">1</td>
<td>ZN770_ZN770_HUMAN.H11MO.0.C</td>
<td>815</td>
<td>chr11:126421569-126421591</td>
<td>1.0</td>
<td>1.404265e-24</td>
<td>7.687053e-23</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">2</td>
<td>ZN770_ZN770_HUMAN.H11MO.0.C</td>
<td>788</td>
<td>chr11:69123148-69123170</td>
<td>-1.0</td>
<td>1.404265e-24</td>
<td>7.687053e-23</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">3</td>
<td>ZN770_ZN770_HUMAN.H11MO.0.C</td>
<td>767</td>
<td>chr11:9314760-9314782</td>
<td>-1.0</td>
<td>1.404265e-24</td>
<td>7.687053e-23</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">4</td>
<td>ZN770_ZN770_HUMAN.H11MO.0.C</td>
<td>859</td>
<td>chr12:120496042-120496064</td>
<td>1.0</td>
<td>1.404265e-24</td>
<td>7.687053e-23</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">...</td>
<td>...</td>
<td>...</td>
<td>...</td>
<td>...</td>
<td>...</td>
<td>...</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">4896</td>
<td>LHX2_MA0700.2</td>
<td>35</td>
<td>chr20:12153927-12153938</td>
<td>1.0</td>
<td>1.404265e-24</td>
<td>5.617062e-23</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">4897</td>
<td>LHX3_LHX3_HUMAN.H11MO.0.C</td>
<td>3</td>
<td>chr1:109726195-109726208</td>
<td>-1.0</td>
<td>1.404265e-24</td>
<td>6.319194e-23</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">4898</td>
<td>MAFF_M07839_2.00</td>
<td>46</td>
<td>chr8:6541215-6541230</td>
<td>-1.0</td>
<td>1.404265e-24</td>
<td>1.488521e-22</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">4899</td>
<td>MAFF_MA0495.3</td>
<td>108</td>
<td>chr17:57946393-57946409</td>
<td>-1.0</td>
<td>1.404265e-24</td>
<td>1.685119e-22</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">4900</td>
<td>HXB7_HXB7_HUMAN.H11MO.0.C</td>
<td>26</td>
<td>chr12:74537563-74537573</td>
<td>1.0</td>
<td>1.404265e-24</td>
<td>5.336209e-23</td>
</tr>
</tbody>
</table>

<p>4901 rows × 6 columns</p>
</div>
</div>
</div>
<div id="cell-107" class="cell" data-execution_count="29">
<div class="sourceCode cell-code" id="cb78"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb78-1"><a href="#cb78-1" aria-hidden="true" tabindex="-1"></a><span class="co"># show rows where 'motif_id' == 'ZN770_ZN770_HUMAN.H11MO.0.C'</span></span>
<span id="cb78-2"><a href="#cb78-2" aria-hidden="true" tabindex="-1"></a>sig_covar_sites_filt[sig_covar_sites_filt[<span class="st">'motif_id'</span>] <span class="op">==</span> <span class="st">'ZN770_ZN770_HUMAN.H11MO.0.C'</span>]</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display" data-execution_count="29">
<div>


<table class="dataframe table table-sm table-striped small" data-quarto-postprocess="true" data-border="1">
<thead>
<tr class="header">
<th data-quarto-table-cell-role="th"></th>
<th data-quarto-table-cell-role="th">motif_id</th>
<th data-quarto-table-cell-role="th">orig_index</th>
<th data-quarto-table-cell-role="th">region_id</th>
<th data-quarto-table-cell-role="th">corr_coeff</th>
<th data-quarto-table-cell-role="th">pvalues</th>
<th data-quarto-table-cell-role="th">adj_pvalues</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td data-quarto-table-cell-role="th">0</td>
<td>ZN770_ZN770_HUMAN.H11MO.0.C</td>
<td>722</td>
<td>chr10:49916101-49916123</td>
<td>1.0</td>
<td>1.404265e-24</td>
<td>7.687053e-23</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">1</td>
<td>ZN770_ZN770_HUMAN.H11MO.0.C</td>
<td>815</td>
<td>chr11:126421569-126421591</td>
<td>1.0</td>
<td>1.404265e-24</td>
<td>7.687053e-23</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">2</td>
<td>ZN770_ZN770_HUMAN.H11MO.0.C</td>
<td>788</td>
<td>chr11:69123148-69123170</td>
<td>-1.0</td>
<td>1.404265e-24</td>
<td>7.687053e-23</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">3</td>
<td>ZN770_ZN770_HUMAN.H11MO.0.C</td>
<td>767</td>
<td>chr11:9314760-9314782</td>
<td>-1.0</td>
<td>1.404265e-24</td>
<td>7.687053e-23</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">4</td>
<td>ZN770_ZN770_HUMAN.H11MO.0.C</td>
<td>859</td>
<td>chr12:120496042-120496064</td>
<td>1.0</td>
<td>1.404265e-24</td>
<td>7.687053e-23</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">5</td>
<td>ZN770_ZN770_HUMAN.H11MO.0.C</td>
<td>886</td>
<td>chr13:23322762-23322784</td>
<td>-1.0</td>
<td>1.404265e-24</td>
<td>7.687053e-23</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">6</td>
<td>ZN770_ZN770_HUMAN.H11MO.0.C</td>
<td>949</td>
<td>chr15:30991957-30991979</td>
<td>1.0</td>
<td>1.404265e-24</td>
<td>7.687053e-23</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">7</td>
<td>ZN770_ZN770_HUMAN.H11MO.0.C</td>
<td>1020</td>
<td>chr16:23642002-23642024</td>
<td>1.0</td>
<td>1.404265e-24</td>
<td>7.687053e-23</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">8</td>
<td>ZN770_ZN770_HUMAN.H11MO.0.C</td>
<td>1038</td>
<td>chr16:56965005-56965027</td>
<td>-1.0</td>
<td>1.404265e-24</td>
<td>7.687053e-23</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">9</td>
<td>ZN770_ZN770_HUMAN.H11MO.0.C</td>
<td>1111</td>
<td>chr17:39451785-39451807</td>
<td>-1.0</td>
<td>1.404265e-24</td>
<td>7.687053e-23</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">10</td>
<td>ZN770_ZN770_HUMAN.H11MO.0.C</td>
<td>1115</td>
<td>chr17:39676104-39676126</td>
<td>-1.0</td>
<td>1.404265e-24</td>
<td>7.687053e-23</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">11</td>
<td>ZN770_ZN770_HUMAN.H11MO.0.C</td>
<td>1261</td>
<td>chr19:22427192-22427214</td>
<td>1.0</td>
<td>1.404265e-24</td>
<td>7.687053e-23</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">12</td>
<td>ZN770_ZN770_HUMAN.H11MO.0.C</td>
<td>1277</td>
<td>chr19:39157676-39157698</td>
<td>1.0</td>
<td>1.404265e-24</td>
<td>7.687053e-23</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">13</td>
<td>ZN770_ZN770_HUMAN.H11MO.0.C</td>
<td>75</td>
<td>chr1:120341303-120341325</td>
<td>-1.0</td>
<td>1.404265e-24</td>
<td>7.687053e-23</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">14</td>
<td>ZN770_ZN770_HUMAN.H11MO.0.C</td>
<td>123</td>
<td>chr1:203792817-203792839</td>
<td>-1.0</td>
<td>1.404265e-24</td>
<td>7.687053e-23</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">15</td>
<td>ZN770_ZN770_HUMAN.H11MO.0.C</td>
<td>205</td>
<td>chr2:112726845-112726867</td>
<td>1.0</td>
<td>1.404265e-24</td>
<td>7.687053e-23</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">16</td>
<td>ZN770_ZN770_HUMAN.H11MO.0.C</td>
<td>239</td>
<td>chr2:218397454-218397476</td>
<td>1.0</td>
<td>1.404265e-24</td>
<td>7.687053e-23</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">17</td>
<td>ZN770_ZN770_HUMAN.H11MO.0.C</td>
<td>150</td>
<td>chr2:950510-950532</td>
<td>1.0</td>
<td>1.404265e-24</td>
<td>7.687053e-23</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">18</td>
<td>ZN770_ZN770_HUMAN.H11MO.0.C</td>
<td>279</td>
<td>chr3:100492479-100492501</td>
<td>1.0</td>
<td>1.404265e-24</td>
<td>7.687053e-23</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">19</td>
<td>ZN770_ZN770_HUMAN.H11MO.0.C</td>
<td>262</td>
<td>chr3:11771383-11771405</td>
<td>1.0</td>
<td>1.404265e-24</td>
<td>7.687053e-23</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">20</td>
<td>ZN770_ZN770_HUMAN.H11MO.0.C</td>
<td>290</td>
<td>chr3:130112453-130112475</td>
<td>-1.0</td>
<td>1.404265e-24</td>
<td>7.687053e-23</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">21</td>
<td>ZN770_ZN770_HUMAN.H11MO.0.C</td>
<td>406</td>
<td>chr6:3157549-3157571</td>
<td>-1.0</td>
<td>1.404265e-24</td>
<td>7.687053e-23</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">22</td>
<td>ZN770_ZN770_HUMAN.H11MO.0.C</td>
<td>492</td>
<td>chr7:33062891-33062913</td>
<td>1.0</td>
<td>1.404265e-24</td>
<td>7.687053e-23</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">23</td>
<td>ZN770_ZN770_HUMAN.H11MO.0.C</td>
<td>600</td>
<td>chr8:127672188-127672210</td>
<td>1.0</td>
<td>1.404265e-24</td>
<td>7.687053e-23</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">24</td>
<td>ZN770_ZN770_HUMAN.H11MO.0.C</td>
<td>621</td>
<td>chr8:144275596-144275618</td>
<td>-1.0</td>
<td>1.404265e-24</td>
<td>7.687053e-23</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">25</td>
<td>ZN770_ZN770_HUMAN.H11MO.0.C</td>
<td>1476</td>
<td>chrX:153743740-153743762</td>
<td>-1.0</td>
<td>1.404265e-24</td>
<td>7.687053e-23</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">26</td>
<td>ZN770_ZN770_HUMAN.H11MO.0.C</td>
<td>1454</td>
<td>chrX:1661588-1661610</td>
<td>-1.0</td>
<td>1.404265e-24</td>
<td>7.687053e-23</td>
</tr>
</tbody>
</table>

</div>
</div>
</div>
<div id="cell-108" class="cell" data-execution_count="32">
<div class="sourceCode cell-code" id="cb79"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb79-1"><a href="#cb79-1" aria-hidden="true" tabindex="-1"></a>sig_covar_sites_filt[sig_covar_sites_filt[<span class="st">'motif_id'</span>] <span class="op">==</span> <span class="st">'ZN281_ZN281_HUMAN.H11MO.0.A'</span>]</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display" data-execution_count="32">
<div>


<table class="dataframe table table-sm table-striped small" data-quarto-postprocess="true" data-border="1">
<thead>
<tr class="header">
<th data-quarto-table-cell-role="th"></th>
<th data-quarto-table-cell-role="th">motif_id</th>
<th data-quarto-table-cell-role="th">orig_index</th>
<th data-quarto-table-cell-role="th">region_id</th>
<th data-quarto-table-cell-role="th">corr_coeff</th>
<th data-quarto-table-cell-role="th">pvalues</th>
<th data-quarto-table-cell-role="th">adj_pvalues</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td data-quarto-table-cell-role="th">27</td>
<td>ZN281_ZN281_HUMAN.H11MO.0.A</td>
<td>410</td>
<td>chr10:32957975-32957990</td>
<td>1.0</td>
<td>1.404265e-24</td>
<td>4.420836e-23</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">28</td>
<td>ZN281_ZN281_HUMAN.H11MO.0.A</td>
<td>465</td>
<td>chr11:118919038-118919053</td>
<td>-1.0</td>
<td>1.404265e-24</td>
<td>4.420836e-23</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">29</td>
<td>ZN281_ZN281_HUMAN.H11MO.0.A</td>
<td>437</td>
<td>chr11:560800-560815</td>
<td>-1.0</td>
<td>1.404265e-24</td>
<td>4.420836e-23</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">30</td>
<td>ZN281_ZN281_HUMAN.H11MO.0.A</td>
<td>452</td>
<td>chr11:69123153-69123168</td>
<td>-1.0</td>
<td>1.404265e-24</td>
<td>4.420836e-23</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">31</td>
<td>ZN281_ZN281_HUMAN.H11MO.0.A</td>
<td>528</td>
<td>chr14:70907984-70907999</td>
<td>1.0</td>
<td>1.404265e-24</td>
<td>4.420836e-23</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">32</td>
<td>ZN281_ZN281_HUMAN.H11MO.0.A</td>
<td>563</td>
<td>chr15:69358558-69358573</td>
<td>-1.0</td>
<td>1.404265e-24</td>
<td>4.420836e-23</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">33</td>
<td>ZN281_ZN281_HUMAN.H11MO.0.A</td>
<td>598</td>
<td>chr16:85760031-85760046</td>
<td>-1.0</td>
<td>1.404265e-24</td>
<td>4.420836e-23</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">34</td>
<td>ZN281_ZN281_HUMAN.H11MO.0.A</td>
<td>601</td>
<td>chr16:88785254-88785269</td>
<td>-1.0</td>
<td>1.404265e-24</td>
<td>4.420836e-23</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">35</td>
<td>ZN281_ZN281_HUMAN.H11MO.0.A</td>
<td>659</td>
<td>chr17:73283925-73283940</td>
<td>1.0</td>
<td>1.404265e-24</td>
<td>4.420836e-23</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">36</td>
<td>ZN281_ZN281_HUMAN.H11MO.0.A</td>
<td>744</td>
<td>chr19:43439422-43439437</td>
<td>1.0</td>
<td>1.404265e-24</td>
<td>4.420836e-23</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">37</td>
<td>ZN281_ZN281_HUMAN.H11MO.0.A</td>
<td>57</td>
<td>chr1:155167545-155167560</td>
<td>-1.0</td>
<td>1.404265e-24</td>
<td>4.420836e-23</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">38</td>
<td>ZN281_ZN281_HUMAN.H11MO.0.A</td>
<td>69</td>
<td>chr1:183734873-183734888</td>
<td>1.0</td>
<td>1.404265e-24</td>
<td>4.420836e-23</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">39</td>
<td>ZN281_ZN281_HUMAN.H11MO.0.A</td>
<td>76</td>
<td>chr1:203792575-203792590</td>
<td>-1.0</td>
<td>1.404265e-24</td>
<td>4.420836e-23</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">40</td>
<td>ZN281_ZN281_HUMAN.H11MO.0.A</td>
<td>32</td>
<td>chr1:39691469-39691484</td>
<td>-1.0</td>
<td>1.404265e-24</td>
<td>4.420836e-23</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">41</td>
<td>ZN281_ZN281_HUMAN.H11MO.0.A</td>
<td>1</td>
<td>chr1:977051-977066</td>
<td>1.0</td>
<td>1.404265e-24</td>
<td>4.420836e-23</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">42</td>
<td>ZN281_ZN281_HUMAN.H11MO.0.A</td>
<td>794</td>
<td>chr20:63658838-63658853</td>
<td>-1.0</td>
<td>1.404265e-24</td>
<td>4.420836e-23</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">43</td>
<td>ZN281_ZN281_HUMAN.H11MO.0.A</td>
<td>810</td>
<td>chr22:27627133-27627148</td>
<td>-1.0</td>
<td>1.404265e-24</td>
<td>4.420836e-23</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">44</td>
<td>ZN281_ZN281_HUMAN.H11MO.0.A</td>
<td>130</td>
<td>chr2:109146836-109146851</td>
<td>1.0</td>
<td>1.404265e-24</td>
<td>4.420836e-23</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">45</td>
<td>ZN281_ZN281_HUMAN.H11MO.0.A</td>
<td>158</td>
<td>chr3:11771390-11771405</td>
<td>1.0</td>
<td>1.404265e-24</td>
<td>4.420836e-23</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">46</td>
<td>ZN281_ZN281_HUMAN.H11MO.0.A</td>
<td>179</td>
<td>chr3:130112069-130112084</td>
<td>-1.0</td>
<td>1.404265e-24</td>
<td>4.420836e-23</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">47</td>
<td>ZN281_ZN281_HUMAN.H11MO.0.A</td>
<td>201</td>
<td>chr4:67701155-67701170</td>
<td>1.0</td>
<td>1.404265e-24</td>
<td>4.420836e-23</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">48</td>
<td>ZN281_ZN281_HUMAN.H11MO.0.A</td>
<td>251</td>
<td>chr6:31139102-31139117</td>
<td>1.0</td>
<td>1.404265e-24</td>
<td>4.420836e-23</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">49</td>
<td>ZN281_ZN281_HUMAN.H11MO.0.A</td>
<td>324</td>
<td>chr7:127392752-127392767</td>
<td>-1.0</td>
<td>1.404265e-24</td>
<td>4.420836e-23</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">50</td>
<td>ZN281_ZN281_HUMAN.H11MO.0.A</td>
<td>301</td>
<td>chr7:33062167-33062182</td>
<td>1.0</td>
<td>1.404265e-24</td>
<td>4.420836e-23</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">51</td>
<td>ZN281_ZN281_HUMAN.H11MO.0.A</td>
<td>303</td>
<td>chr7:33062898-33062913</td>
<td>1.0</td>
<td>1.404265e-24</td>
<td>4.420836e-23</td>
</tr>
<tr class="even">
<td data-quarto-table-cell-role="th">52</td>
<td>ZN281_ZN281_HUMAN.H11MO.0.A</td>
<td>285</td>
<td>chr7:816260-816275</td>
<td>1.0</td>
<td>1.404265e-24</td>
<td>4.420836e-23</td>
</tr>
<tr class="odd">
<td data-quarto-table-cell-role="th">53</td>
<td>ZN281_ZN281_HUMAN.H11MO.0.A</td>
<td>369</td>
<td>chr8:144275603-144275618</td>
<td>-1.0</td>
<td>1.404265e-24</td>
<td>4.420836e-23</td>
</tr>
</tbody>
</table>

</div>
</div>
</div>
</section>
</section>

</main>
<!-- /main column -->
<script id="quarto-html-after-body" type="application/javascript">
window.document.addEventListener("DOMContentLoaded", function (event) {
  const toggleBodyColorMode = (bsSheetEl) => {
    const mode = bsSheetEl.getAttribute("data-mode");
    const bodyEl = window.document.querySelector("body");
    if (mode === "dark") {
      bodyEl.classList.add("quarto-dark");
      bodyEl.classList.remove("quarto-light");
    } else {
      bodyEl.classList.add("quarto-light");
      bodyEl.classList.remove("quarto-dark");
    }
  }
  const toggleBodyColorPrimary = () => {
    const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
    if (bsSheetEl) {
      toggleBodyColorMode(bsSheetEl);
    }
  }
  toggleBodyColorPrimary();  
  const icon = "";
  const anchorJS = new window.AnchorJS();
  anchorJS.options = {
    placement: 'right',
    icon: icon
  };
  anchorJS.add('.anchored');
  const isCodeAnnotation = (el) => {
    for (const clz of el.classList) {
      if (clz.startsWith('code-annotation-')) {                     
        return true;
      }
    }
    return false;
  }
  const clipboard = new window.ClipboardJS('.code-copy-button', {
    text: function(trigger) {
      const codeEl = trigger.previousElementSibling.cloneNode(true);
      for (const childEl of codeEl.children) {
        if (isCodeAnnotation(childEl)) {
          childEl.remove();
        }
      }
      return codeEl.innerText;
    }
  });
  clipboard.on('success', function(e) {
    // button target
    const button = e.trigger;
    // don't keep focus
    button.blur();
    // flash "checked"
    button.classList.add('code-copy-button-checked');
    var currentTitle = button.getAttribute("title");
    button.setAttribute("title", "Copied!");
    let tooltip;
    if (window.bootstrap) {
      button.setAttribute("data-bs-toggle", "tooltip");
      button.setAttribute("data-bs-placement", "left");
      button.setAttribute("data-bs-title", "Copied!");
      tooltip = new bootstrap.Tooltip(button, 
        { trigger: "manual", 
          customClass: "code-copy-button-tooltip",
          offset: [0, -8]});
      tooltip.show();    
    }
    setTimeout(function() {
      if (tooltip) {
        tooltip.hide();
        button.removeAttribute("data-bs-title");
        button.removeAttribute("data-bs-toggle");
        button.removeAttribute("data-bs-placement");
      }
      button.setAttribute("title", currentTitle);
      button.classList.remove('code-copy-button-checked');
    }, 1000);
    // clear code selection
    e.clearSelection();
  });
  function tippyHover(el, contentFn, onTriggerFn, onUntriggerFn) {
    const config = {
      allowHTML: true,
      maxWidth: 500,
      delay: 100,
      arrow: false,
      appendTo: function(el) {
          return el.parentElement;
      },
      interactive: true,
      interactiveBorder: 10,
      theme: 'quarto',
      placement: 'bottom-start',
    };
    if (contentFn) {
      config.content = contentFn;
    }
    if (onTriggerFn) {
      config.onTrigger = onTriggerFn;
    }
    if (onUntriggerFn) {
      config.onUntrigger = onUntriggerFn;
    }
    window.tippy(el, config); 
  }
  const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
  for (var i=0; i<noterefs.length; i++) {
    const ref = noterefs[i];
    tippyHover(ref, function() {
      // use id or data attribute instead here
      let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
      try { href = new URL(href).hash; } catch {}
      const id = href.replace(/^#\/?/, "");
      const note = window.document.getElementById(id);
      return note.innerHTML;
    });
  }
  const xrefs = window.document.querySelectorAll('a.quarto-xref');
  const processXRef = (id, note) => {
    // Strip column container classes
    const stripColumnClz = (el) => {
      el.classList.remove("page-full", "page-columns");
      if (el.children) {
        for (const child of el.children) {
          stripColumnClz(child);
        }
      }
    }
    stripColumnClz(note)
    if (id === null || id.startsWith('sec-')) {
      // Special case sections, only their first couple elements
      const container = document.createElement("div");
      if (note.children && note.children.length > 2) {
        container.appendChild(note.children[0].cloneNode(true));
        for (let i = 1; i < note.children.length; i++) {
          const child = note.children[i];
          if (child.tagName === "P" && child.innerText === "") {
            continue;
          } else {
            container.appendChild(child.cloneNode(true));
            break;
          }
        }
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(container);
        }
        return container.innerHTML
      } else {
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(note);
        }
        return note.innerHTML;
      }
    } else {
      // Remove any anchor links if they are present
      const anchorLink = note.querySelector('a.anchorjs-link');
      if (anchorLink) {
        anchorLink.remove();
      }
      if (window.Quarto?.typesetMath) {
        window.Quarto.typesetMath(note);
      }
      // TODO in 1.5, we should make sure this works without a callout special case
      if (note.classList.contains("callout")) {
        return note.outerHTML;
      } else {
        return note.innerHTML;
      }
    }
  }
  for (var i=0; i<xrefs.length; i++) {
    const xref = xrefs[i];
    tippyHover(xref, undefined, function(instance) {
      instance.disable();
      let url = xref.getAttribute('href');
      let hash = undefined; 
      if (url.startsWith('#')) {
        hash = url;
      } else {
        try { hash = new URL(url).hash; } catch {}
      }
      if (hash) {
        const id = hash.replace(/^#\/?/, "");
        const note = window.document.getElementById(id);
        if (note !== null) {
          try {
            const html = processXRef(id, note.cloneNode(true));
            instance.setContent(html);
          } finally {
            instance.enable();
            instance.show();
          }
        } else {
          // See if we can fetch this
          fetch(url.split('#')[0])
          .then(res => res.text())
          .then(html => {
            const parser = new DOMParser();
            const htmlDoc = parser.parseFromString(html, "text/html");
            const note = htmlDoc.getElementById(id);
            if (note !== null) {
              const html = processXRef(id, note);
              instance.setContent(html);
            } 
          }).finally(() => {
            instance.enable();
            instance.show();
          });
        }
      } else {
        // See if we can fetch a full url (with no hash to target)
        // This is a special case and we should probably do some content thinning / targeting
        fetch(url)
        .then(res => res.text())
        .then(html => {
          const parser = new DOMParser();
          const htmlDoc = parser.parseFromString(html, "text/html");
          const note = htmlDoc.querySelector('main.content');
          if (note !== null) {
            // This should only happen for chapter cross references
            // (since there is no id in the URL)
            // remove the first header
            if (note.children.length > 0 && note.children[0].tagName === "HEADER") {
              note.children[0].remove();
            }
            const html = processXRef(null, note);
            instance.setContent(html);
          } 
        }).finally(() => {
          instance.enable();
          instance.show();
        });
      }
    }, function(instance) {
    });
  }
      let selectedAnnoteEl;
      const selectorForAnnotation = ( cell, annotation) => {
        let cellAttr = 'data-code-cell="' + cell + '"';
        let lineAttr = 'data-code-annotation="' +  annotation + '"';
        const selector = 'span[' + cellAttr + '][' + lineAttr + ']';
        return selector;
      }
      const selectCodeLines = (annoteEl) => {
        const doc = window.document;
        const targetCell = annoteEl.getAttribute("data-target-cell");
        const targetAnnotation = annoteEl.getAttribute("data-target-annotation");
        const annoteSpan = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
        const lines = annoteSpan.getAttribute("data-code-lines").split(",");
        const lineIds = lines.map((line) => {
          return targetCell + "-" + line;
        })
        let top = null;
        let height = null;
        let parent = null;
        if (lineIds.length > 0) {
            //compute the position of the single el (top and bottom and make a div)
            const el = window.document.getElementById(lineIds[0]);
            top = el.offsetTop;
            height = el.offsetHeight;
            parent = el.parentElement.parentElement;
          if (lineIds.length > 1) {
            const lastEl = window.document.getElementById(lineIds[lineIds.length - 1]);
            const bottom = lastEl.offsetTop + lastEl.offsetHeight;
            height = bottom - top;
          }
          if (top !== null && height !== null && parent !== null) {
            // cook up a div (if necessary) and position it 
            let div = window.document.getElementById("code-annotation-line-highlight");
            if (div === null) {
              div = window.document.createElement("div");
              div.setAttribute("id", "code-annotation-line-highlight");
              div.style.position = 'absolute';
              parent.appendChild(div);
            }
            div.style.top = top - 2 + "px";
            div.style.height = height + 4 + "px";
            div.style.left = 0;
            let gutterDiv = window.document.getElementById("code-annotation-line-highlight-gutter");
            if (gutterDiv === null) {
              gutterDiv = window.document.createElement("div");
              gutterDiv.setAttribute("id", "code-annotation-line-highlight-gutter");
              gutterDiv.style.position = 'absolute';
              const codeCell = window.document.getElementById(targetCell);
              const gutter = codeCell.querySelector('.code-annotation-gutter');
              gutter.appendChild(gutterDiv);
            }
            gutterDiv.style.top = top - 2 + "px";
            gutterDiv.style.height = height + 4 + "px";
          }
          selectedAnnoteEl = annoteEl;
        }
      };
      const unselectCodeLines = () => {
        const elementsIds = ["code-annotation-line-highlight", "code-annotation-line-highlight-gutter"];
        elementsIds.forEach((elId) => {
          const div = window.document.getElementById(elId);
          if (div) {
            div.remove();
          }
        });
        selectedAnnoteEl = undefined;
      };
        // Handle positioning of the toggle
    window.addEventListener(
      "resize",
      throttle(() => {
        elRect = undefined;
        if (selectedAnnoteEl) {
          selectCodeLines(selectedAnnoteEl);
        }
      }, 10)
    );
    function throttle(fn, ms) {
    let throttle = false;
    let timer;
      return (...args) => {
        if(!throttle) { // first call gets through
            fn.apply(this, args);
            throttle = true;
        } else { // all the others get throttled
            if(timer) clearTimeout(timer); // cancel #2
            timer = setTimeout(() => {
              fn.apply(this, args);
              timer = throttle = false;
            }, ms);
        }
      };
    }
      // Attach click handler to the DT
      const annoteDls = window.document.querySelectorAll('dt[data-target-cell]');
      for (const annoteDlNode of annoteDls) {
        annoteDlNode.addEventListener('click', (event) => {
          const clickedEl = event.target;
          if (clickedEl !== selectedAnnoteEl) {
            unselectCodeLines();
            const activeEl = window.document.querySelector('dt[data-target-cell].code-annotation-active');
            if (activeEl) {
              activeEl.classList.remove('code-annotation-active');
            }
            selectCodeLines(clickedEl);
            clickedEl.classList.add('code-annotation-active');
          } else {
            // Unselect the line
            unselectCodeLines();
            clickedEl.classList.remove('code-annotation-active');
          }
        });
      }
  const findCites = (el) => {
    const parentEl = el.parentElement;
    if (parentEl) {
      const cites = parentEl.dataset.cites;
      if (cites) {
        return {
          el,
          cites: cites.split(' ')
        };
      } else {
        return findCites(el.parentElement)
      }
    } else {
      return undefined;
    }
  };
  var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
  for (var i=0; i<bibliorefs.length; i++) {
    const ref = bibliorefs[i];
    const citeInfo = findCites(ref);
    if (citeInfo) {
      tippyHover(citeInfo.el, function() {
        var popup = window.document.createElement('div');
        citeInfo.cites.forEach(function(cite) {
          var citeDiv = window.document.createElement('div');
          citeDiv.classList.add('hanging-indent');
          citeDiv.classList.add('csl-entry');
          var biblioDiv = window.document.getElementById('ref-' + cite);
          if (biblioDiv) {
            citeDiv.innerHTML = biblioDiv.innerHTML;
          }
          popup.appendChild(citeDiv);
        });
        return popup.innerHTML;
      });
    }
  }
});
</script>
</div> <!-- /content -->




</body></html>